```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=TRUE, cache=TRUE,
                      fig.height=4, fig.width=6, 
#                      fig.path='/figs',   DO NOT USE
                      fig.pos="ht!")
```

# Get started


## What you need
- **Genotype data** for a couple hundred SNPs (<1000); ideally with high minor allele frequency, low missingness, low genotyping error rate, and in low LD, but real-world data usually works well too. The data format is 1 row per individual, 1 column per SNP, coded as 0/1/2 copies of the reference allele. Conversion from a few other formats is supported.
- **Sex** and **birth year** information for as many individuals as possible ('birth year': discrete time unit of birth/hatching/germination/...). If the exact year is unknown, minimum and/or maximum possible years can be provided.
- **R**: can be downloaded from [CRAN](https://cran.r-project.org/). Sequoia definitely works with R versions 3.5 and up; I am unsure about older versions. 
- (optional) [Rstudio](https://rstudio.com/products/rstudio/): for a more user-friendly R experience


## Download & installation

### Current version
To install the most recent thoroughly checked version of `sequoia`, simply open R and run the command 
```{r install-1, eval=FALSE}
install.packages("sequoia") 
```


### Other versions
The bulk of `sequoia` is written in Fortran, and requires compilation (turn human-readable text files into an executable). A Fortran compiler is not by default installed on all computers, and therefore [CRAN](https://CRAN.R-project.org/package=sequoia) provides pre-compiled versions for windows (.zip binary) and macOS (.tgz binary), in addition to hosting the source package (.tar.gz).  

The source files from previous versions are archived on [CRAN](https://cran.r-project.org/src/contrib/Archive/sequoia/), but the binary files are not; from version 1.3.3 onwards these can be found in [my github archive](https://github.com/JiscaH/sequoia_archives). There you can also find the in-between versions that never made it to CRAN as source + windows binary (the latter compiled under 64-bit windows 8 and a then-recent version of R). 

You can install these by downloading them to your hard disk and then in R (for windows, no compiler needed):
```{r install-2, eval=FALSE}
install.packages("C:/path/to/file/sequoia_2.3.3.zip", repos=NULL)
```
Sometimes it seems necessary to first rename the file to 'sequoia.zip', if R can't find the package back after installation, and sometimes you need to turn R off & on before it works. 

On github you can also find a [browsable directory](https://github.com/JiscaH/sequoia) of the latest development version. It can be installed with   
```{r , eval=FALSE}
devtools::install_github("JiscaH/sequoia")
# or
remotes::install_github("JiscaH/sequoia")
```
This requires a Fortran compiler (which I think is included in the devtools package). 



## Reconstruct
The function to perform pedigree reconstruction is also called `sequoia()`:
```
# load the package
library(sequoia)  

# load example data 
data(SimGeno_example, LH_HSg5, package="sequoia")  

# run pedigree reconstruction
SeqOUT <- sequoia(GenoM = SimGeno_example, 
                  LifeHistData = LH_HSg5, 
                  Err = 0.005,   # genotyping error rate
                  Module="ped", 
                  quiet="verbose", 
                  Plot=TRUE)
# the result is a list with the pedigree, run parameters, 
# and various other elements.                 

# graphical summary of results
SummarySeq(SeqOUT)
```


## Access help files
In R, help for each function is available via `?functionname`, e.g. `?sequoia`. An overview of all help files and other documentation in a package is available via `help(package="sequoia")`.

The 'See Also' section near the end of the `sequoia()` help file has a list with the main functions in the package, for those who tend to forget function names (like me).

