<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Simulate Genotyping Errors — MkGenoErrors • sequoia</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Simulate Genotyping Errors — MkGenoErrors" />
<meta property="og:description" content="Generate errors and missing values in a (simulated) genotype
  matrix." />
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png" />
<meta property="og:image:alt" content="sequoia hexlogo" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_main/_book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/_book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/_performance_stats.html">Stats</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Simulate Genotyping Errors</h1>
    
    <div class="hidden name"><code>MkGenoErrors.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate errors and missing values in a (simulated) genotype
  matrix.</p>
    </div>

    <pre class="usage"><span class='fu'>MkGenoErrors</span><span class='op'>(</span>
  <span class='va'>SGeno</span>,
  CallRate <span class='op'>=</span> <span class='fl'>0.99</span>,
  SnpError <span class='op'>=</span> <span class='fl'>5e-04</span>,
  ErrorFM <span class='op'>=</span> <span class='kw'>function</span><span class='op'>(</span><span class='va'>E</span><span class='op'>)</span> <span class='op'>{</span>     <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span> <span class='op'>-</span> <span class='va'>E</span> <span class='op'>-</span> <span class='op'>(</span><span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span>, <span class='va'>E</span>, <span class='op'>(</span><span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span>, <span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span>, <span class='fl'>1</span> <span class='op'>-</span> <span class='va'>E</span>, <span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span>,
    <span class='op'>(</span><span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span>, <span class='va'>E</span>, <span class='fl'>1</span> <span class='op'>-</span> <span class='va'>E</span> <span class='op'>-</span> <span class='op'>(</span><span class='va'>E</span><span class='op'>/</span><span class='fl'>2</span><span class='op'>)</span><span class='op'>^</span><span class='fl'>2</span><span class='op'>)</span>, <span class='fl'>3</span>, <span class='fl'>3</span>, byrow <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span> <span class='op'>}</span>,
  Error.shape <span class='op'>=</span> <span class='fl'>0.5</span>,
  CallRate.shape <span class='op'>=</span> <span class='fl'>1</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>SGeno</th>
      <td><p>matrix with genotype data in Sequoia's format: 1 row per
individual, 1 column per SNP, and genotypes coded as 0/1/2.</p></td>
    </tr>
    <tr>
      <th>CallRate</th>
      <td><p>either a single number for the mean call rate (genotyping
success), OR a vector with the call rate at each SNP, OR a named vector
with the call rate for each individual. In the third case, ParMis is
ignored, and individuals in the pedigree (as id or parent) not included in
this vector are presumed non-genotyped.</p></td>
    </tr>
    <tr>
      <th>SnpError</th>
      <td><p>mean per-locus genotyping error rate across SNPs, and a
beta-distribution will be used to simulate the number of missing cases per
SNP, OR a vector with the genotyping error for each SNP.</p></td>
    </tr>
    <tr>
      <th>ErrorFM</th>
      <td><p>function taking the error rate (scalar) as argument and
returning a 4x4 or 3x3 matrix with probabilities that actual genotype i
(rows) is observed as genotype j (columns).</p></td>
    </tr>
    <tr>
      <th>Error.shape</th>
      <td><p>first shape parameter (alpha) of beta-distribution of
per-SNP error rates. A higher value results in a flatter distribution.</p></td>
    </tr>
    <tr>
      <th>CallRate.shape</th>
      <td><p>as Error.shape, for per-SNP call rates.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>The input genotype matrix, with some genotypes replaced, and some
  set to missing (-9).</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>Ped_HSg5</span><span class='op'>)</span>
<span class='va'>GenoM</span> <span class='op'>&lt;-</span> <span class='fu'><a href='SimGeno.html'>SimGeno</a></span><span class='op'>(</span>Ped <span class='op'>=</span> <span class='va'>Ped_HSg5</span>, nSnp <span class='op'>=</span> <span class='fl'>100</span>, ParMis <span class='op'>=</span> <span class='fl'>0.2</span>,
                 SnpError<span class='op'>=</span><span class='fl'>0</span>, CallRate<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span>
<span class='va'>GenoM.actual</span> <span class='op'>&lt;-</span> <span class='va'>GenoM</span>
<span class='va'>LowQ</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/sample.html'>sample.int</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>GenoM</span><span class='op'>)</span>, <span class='fl'>42</span><span class='op'>)</span>  <span class='co'># low-quality samples</span>
<span class='va'>GenoM</span><span class='op'>[</span><span class='va'>LowQ</span>, <span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'>MkGenoErrors</span><span class='op'>(</span><span class='va'>GenoM</span><span class='op'>[</span><span class='va'>LowQ</span>, <span class='op'>]</span>, SnpError <span class='op'>=</span> <span class='fl'>0.05</span><span class='op'>)</span>
<span class='va'>GenoM</span><span class='op'>[</span><span class='op'>-</span><span class='va'>LowQ</span>, <span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'>MkGenoErrors</span><span class='op'>(</span><span class='va'>GenoM</span><span class='op'>[</span><span class='op'>-</span><span class='va'>LowQ</span>, <span class='op'>]</span>, SnpError <span class='op'>=</span> <span class='fl'>0.001</span><span class='op'>)</span>
<span class='va'>ErrorCount</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span><span class='op'>(</span><span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>GenoM</span><span class='op'>)</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>i</span><span class='op'>)</span> <span class='op'>{</span>
  <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span><span class='op'>(</span><span class='va'>GenoM.actual</span><span class='op'>[</span><span class='va'>i</span>,<span class='op'>]</span> <span class='op'>!=</span> <span class='va'>GenoM</span><span class='op'>[</span><span class='va'>i</span>,<span class='op'>]</span> <span class='op'>&amp;</span> <span class='va'>GenoM</span><span class='op'>[</span><span class='va'>i</span>,<span class='op'>]</span> <span class='op'>!=</span> <span class='op'>-</span><span class='fl'>9</span><span class='op'>)</span> <span class='op'>}</span> <span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>ErrorCount</span><span class='op'>[</span><span class='va'>LowQ</span><span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] 4.547619</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span><span class='op'>(</span><span class='va'>ErrorCount</span><span class='op'>[</span><span class='op'>-</span><span class='va'>LowQ</span><span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] 0.1013072</div><div class='input'>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


