<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 FAQ | Sequoia User Guide</title>
  <meta name="description" content="8 FAQ | Sequoia User Guide" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="8 FAQ | Sequoia User Guide" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="/../../sequoia_logo.png" />
  
  <meta name="github-repo" content="JiscaH/sequoia" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 FAQ | Sequoia User Guide" />
  
  
  <meta name="twitter:image" content="/../../sequoia_logo.png" />

<meta name="author" content="Jisca Huisman" />


<meta name="date" content="2021-05-02" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../../apple-touch-icon-120x120.png" />
  <link rel="shortcut icon" href="../../favicon.ico" type="image/x-icon" />
<link rel="prev" href="hermaphrodites.html"/>
<link rel="next" href="references.html"/>
<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./../../../index.html"><img src="../../home-solid-silver.svg" height="40" alt="home"/></a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="sec-Background.html"><a href="sec-Background.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="key-points.html"><a href="key-points.html"><i class="fa fa-check"></i><b>2</b> Key Points</a>
<ul>
<li class="chapter" data-level="2.1" data-path="key-points.html"><a href="key-points.html#sec:MendelErrors"><i class="fa fa-check"></i><b>2.1</b> Mendelian Errors</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="key-points.html"><a href="key-points.html#parent-pairs"><i class="fa fa-check"></i><b>2.1.1</b> Parent-pairs</a></li>
<li class="chapter" data-level="2.1.2" data-path="key-points.html"><a href="key-points.html#maximum-mismatches"><i class="fa fa-check"></i><b>2.1.2</b> Maximum mismatches</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="key-points.html"><a href="key-points.html#likelihoods"><i class="fa fa-check"></i><b>2.2</b> Likelihoods</a></li>
<li class="chapter" data-level="2.3" data-path="key-points.html"><a href="key-points.html#sec:GenoErrors"><i class="fa fa-check"></i><b>2.3</b> Genotyping errors</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="key-points.html"><a href="key-points.html#effect"><i class="fa fa-check"></i><b>2.3.1</b> Effect</a></li>
<li class="chapter" data-level="2.3.2" data-path="key-points.html"><a href="key-points.html#assumed-genotyping-error-rate-err"><i class="fa fa-check"></i><b>2.3.2</b> Assumed genotyping error rate <code>Err</code></a></li>
<li class="chapter" data-level="2.3.3" data-path="key-points.html"><a href="key-points.html#performance-with-high-genotyping-error-rates"><i class="fa fa-check"></i><b>2.3.3</b> Performance with high genotyping error rates</a></li>
<li class="chapter" data-level="2.3.4" data-path="key-points.html"><a href="key-points.html#effect-on-runtime"><i class="fa fa-check"></i><b>2.3.4</b> Effect on runtime</a></li>
<li class="chapter" data-level="2.3.5" data-path="key-points.html"><a href="key-points.html#error-matrix"><i class="fa fa-check"></i><b>2.3.5</b> Error matrix</a></li>
<li class="chapter" data-level="2.3.6" data-path="key-points.html"><a href="key-points.html#estimate"><i class="fa fa-check"></i><b>2.3.6</b> Estimate</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="key-points.html"><a href="key-points.html#sec:BY-AP"><i class="fa fa-check"></i><b>2.4</b> Birth years &amp; Ageprior</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="key-points.html"><a href="key-points.html#definition-interpretation"><i class="fa fa-check"></i><b>2.4.1</b> Definition &amp; interpretation</a></li>
<li class="chapter" data-level="2.4.2" data-path="key-points.html"><a href="key-points.html#implementation"><i class="fa fa-check"></i><b>2.4.2</b> Implementation</a></li>
<li class="chapter" data-level="2.4.3" data-path="key-points.html"><a href="key-points.html#smooth-flatten"><i class="fa fa-check"></i><b>2.4.3</b> Smooth &amp; Flatten</a></li>
<li class="chapter" data-level="2.4.4" data-path="key-points.html"><a href="key-points.html#age-based-non-assignment"><i class="fa fa-check"></i><b>2.4.4</b> Age-based non-assignment</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="key-points.html"><a href="key-points.html#parent-llr"><i class="fa fa-check"></i><b>2.5</b> Parent LLR</a></li>
<li class="chapter" data-level="2.6" data-path="key-points.html"><a href="key-points.html#sec:EstConf"><i class="fa fa-check"></i><b>2.6</b> Confidence</a></li>
<li class="chapter" data-level="2.7" data-path="key-points.html"><a href="key-points.html#sec:lowAR"><i class="fa fa-check"></i><b>2.7</b> Low assignment rate</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="key-points.html"><a href="key-points.html#causes"><i class="fa fa-check"></i><b>2.7.1</b> Causes</a></li>
<li class="chapter" data-level="2.7.2" data-path="key-points.html"><a href="key-points.html#remedies"><i class="fa fa-check"></i><b>2.7.2</b> Remedies</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="function-overview.html"><a href="function-overview.html"><i class="fa fa-check"></i><b>3</b> Function overview</a>
<ul>
<li class="chapter" data-level="3.1" data-path="function-overview.html"><a href="function-overview.html#input"><i class="fa fa-check"></i><b>3.1</b> Input</a></li>
<li class="chapter" data-level="3.2" data-path="function-overview.html"><a href="function-overview.html#simulate"><i class="fa fa-check"></i><b>3.2</b> Simulate</a></li>
<li class="chapter" data-level="3.3" data-path="function-overview.html"><a href="function-overview.html#ageprior"><i class="fa fa-check"></i><b>3.3</b> Ageprior</a></li>
<li class="chapter" data-level="3.4" data-path="function-overview.html"><a href="function-overview.html#pedigree-reconstruction"><i class="fa fa-check"></i><b>3.4</b> Pedigree reconstruction</a></li>
<li class="chapter" data-level="3.5" data-path="function-overview.html"><a href="function-overview.html#pedigree-check"><i class="fa fa-check"></i><b>3.5</b> Pedigree check</a></li>
<li class="chapter" data-level="3.6" data-path="function-overview.html"><a href="function-overview.html#pairwise-relationships"><i class="fa fa-check"></i><b>3.6</b> Pairwise relationships</a></li>
<li class="chapter" data-level="3.7" data-path="function-overview.html"><a href="function-overview.html#miscellaneous"><i class="fa fa-check"></i><b>3.7</b> Miscellaneous</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequoia-input.html"><a href="sequoia-input.html"><i class="fa fa-check"></i><b>4</b> <code>sequoia()</code> input</a>
<ul>
<li class="chapter" data-level="4.1" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:GenoFormat"><i class="fa fa-check"></i><b>4.1</b> Genotypes</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="sequoia-input.html"><a href="sequoia-input.html#subset-snps"><i class="fa fa-check"></i><b>4.1.1</b> Subset SNPs</a></li>
<li class="chapter" data-level="4.1.2" data-path="sequoia-input.html"><a href="sequoia-input.html#plink"><i class="fa fa-check"></i><b>4.1.2</b> PLINK</a></li>
<li class="chapter" data-level="4.1.3" data-path="sequoia-input.html"><a href="sequoia-input.html#low-call-rate"><i class="fa fa-check"></i><b>4.1.3</b> Low call rate</a></li>
<li class="chapter" data-level="4.1.4" data-path="sequoia-input.html"><a href="sequoia-input.html#very-large-datasets"><i class="fa fa-check"></i><b>4.1.4</b> Very large datasets</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="sequoia-input.html"><a href="sequoia-input.html#birth-years-sex"><i class="fa fa-check"></i><b>4.2</b> Birth years &amp; sex</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="sequoia-input.html"><a href="sequoia-input.html#lifehistdata"><i class="fa fa-check"></i><b>4.2.1</b> LifeHistData</a></li>
<li class="chapter" data-level="4.2.2" data-path="sequoia-input.html"><a href="sequoia-input.html#unknown-birth-years"><i class="fa fa-check"></i><b>4.2.2</b> Unknown birth years</a></li>
<li class="chapter" data-level="4.2.3" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:custom-AP"><i class="fa fa-check"></i><b>4.2.3</b> <code>args.AP</code> (Customising the ageprior)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="sequoia-input.html"><a href="sequoia-input.html#tfilter-tassign"><i class="fa fa-check"></i><b>4.3</b> <code>Tfilter</code> &amp; <code>Tassign</code></a></li>
<li class="chapter" data-level="4.4" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:Complx"><i class="fa fa-check"></i><b>4.4</b> <code>Complex</code></a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:WeirdRels"><i class="fa fa-check"></i><b>4.4.1</b> Unusual relationships</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="sequoia-input.html"><a href="sequoia-input.html#herm"><i class="fa fa-check"></i><b>4.5</b> <code>Herm</code></a></li>
<li class="chapter" data-level="4.6" data-path="sequoia-input.html"><a href="sequoia-input.html#useage"><i class="fa fa-check"></i><b>4.6</b> <code>UseAge</code></a></li>
<li class="chapter" data-level="4.7" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:reuse"><i class="fa fa-check"></i><b>4.7</b> <code>SeqList</code></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="running-sequoia.html"><a href="running-sequoia.html"><i class="fa fa-check"></i><b>5</b> Running <code>sequoia()</code></a>
<ul>
<li class="chapter" data-level="5.1" data-path="running-sequoia.html"><a href="running-sequoia.html#module-maxsibiter"><i class="fa fa-check"></i><b>5.1</b> <code>Module</code> (<code>MaxSibIter</code>)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="running-sequoia.html"><a href="running-sequoia.html#sec:MaxSibIter"><i class="fa fa-check"></i><b>5.1.1</b> <code>MaxSibIter</code></a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="running-sequoia.html"><a href="running-sequoia.html#input-check"><i class="fa fa-check"></i><b>5.2</b> Input check</a></li>
<li class="chapter" data-level="5.3" data-path="running-sequoia.html"><a href="running-sequoia.html#check-for-duplicates"><i class="fa fa-check"></i><b>5.3</b> Check for duplicates</a></li>
<li class="chapter" data-level="5.4" data-path="running-sequoia.html"><a href="running-sequoia.html#parentage-assignment"><i class="fa fa-check"></i><b>5.4</b> Parentage assignment</a></li>
<li class="chapter" data-level="5.5" data-path="running-sequoia.html"><a href="running-sequoia.html#ageprior-1"><i class="fa fa-check"></i><b>5.5</b> AgePrior</a></li>
<li class="chapter" data-level="5.6" data-path="running-sequoia.html"><a href="running-sequoia.html#full-reconstruction"><i class="fa fa-check"></i><b>5.6</b> Full reconstruction</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="running-sequoia.html"><a href="running-sequoia.html#dummy-individuals"><i class="fa fa-check"></i><b>5.6.1</b> Dummy Individuals</a></li>
<li class="chapter" data-level="5.6.2" data-path="running-sequoia.html"><a href="running-sequoia.html#total-likelihood"><i class="fa fa-check"></i><b>5.6.2</b> Total likelihood</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="running-sequoia.html"><a href="running-sequoia.html#save-output"><i class="fa fa-check"></i><b>5.7</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="output-check.html"><a href="output-check.html"><i class="fa fa-check"></i><b>6</b> Output check</a>
<ul>
<li class="chapter" data-level="6.1" data-path="output-check.html"><a href="output-check.html#pedigree-stats-plots"><i class="fa fa-check"></i><b>6.1</b> Pedigree stats &amp; plots</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="output-check.html"><a href="output-check.html#negative-llrs"><i class="fa fa-check"></i><b>6.1.1</b> Negative LLR’s</a></li>
<li class="chapter" data-level="6.1.2" data-path="output-check.html"><a href="output-check.html#dyad-plot"><i class="fa fa-check"></i><b>6.1.2</b> Dyad plot</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="output-check.html"><a href="output-check.html#sec:compPed"><i class="fa fa-check"></i><b>6.2</b> Compare pedigrees</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="output-check.html"><a href="output-check.html#dummy-matching"><i class="fa fa-check"></i><b>6.2.1</b> Dummy matching</a></li>
<li class="chapter" data-level="6.2.2" data-path="output-check.html"><a href="output-check.html#example"><i class="fa fa-check"></i><b>6.2.2</b> Example</a></li>
<li class="chapter" data-level="6.2.3" data-path="output-check.html"><a href="output-check.html#colony"><i class="fa fa-check"></i><b>6.2.3</b> Colony</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="output-check.html"><a href="output-check.html#sec:compR"><i class="fa fa-check"></i><b>6.3</b> Relatedness</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="output-check.html"><a href="output-check.html#sec:Rped"><i class="fa fa-check"></i><b>6.3.1</b> Pedigree relatedness</a></li>
<li class="chapter" data-level="6.3.2" data-path="output-check.html"><a href="output-check.html#genomic-relatedness"><i class="fa fa-check"></i><b>6.3.2</b> Genomic relatedness</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="output-check.html"><a href="output-check.html#cluster-families"><i class="fa fa-check"></i><b>6.4</b> Cluster families</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="hermaphrodites.html"><a href="hermaphrodites.html"><i class="fa fa-check"></i><b>7</b> Hermaphrodites</a>
<ul>
<li class="chapter" data-level="7.1" data-path="hermaphrodites.html"><a href="hermaphrodites.html#herm-a-pedigree-prior"><i class="fa fa-check"></i><b>7.1</b> Herm ‘A’: Pedigree prior</a></li>
<li class="chapter" data-level="7.2" data-path="hermaphrodites.html"><a href="hermaphrodites.html#herm-b-ignore-sex-role"><i class="fa fa-check"></i><b>7.2</b> Herm ‘B’: Ignore sex role</a></li>
<li class="chapter" data-level="7.3" data-path="hermaphrodites.html"><a href="hermaphrodites.html#hermaphrodite-dummies"><i class="fa fa-check"></i><b>7.3</b> Hermaphrodite dummies</a></li>
<li class="chapter" data-level="7.4" data-path="hermaphrodites.html"><a href="hermaphrodites.html#assignment-errors"><i class="fa fa-check"></i><b>7.4</b> Assignment errors</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="faq.html"><a href="faq.html"><i class="fa fa-check"></i><b>8</b> FAQ</a>
<ul>
<li class="chapter" data-level="8.1" data-path="faq.html"><a href="faq.html#error-aboutdue-to-input-data"><i class="fa fa-check"></i><b>8.1</b> Error about/due to input data</a></li>
<li class="chapter" data-level="8.2" data-path="faq.html"><a href="faq.html#sec:FAQlowAR"><i class="fa fa-check"></i><b>8.2</b> Low assignment rate</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="faq.html"><a href="faq.html#not-enough-snps"><i class="fa fa-check"></i><b>8.2.1</b> Not enough SNPs</a></li>
<li class="chapter" data-level="8.2.2" data-path="faq.html"><a href="faq.html#genotyping-errors"><i class="fa fa-check"></i><b>8.2.2</b> Genotyping errors</a></li>
<li class="chapter" data-level="8.2.3" data-path="faq.html"><a href="faq.html#lacking-sex-information"><i class="fa fa-check"></i><b>8.2.3</b> Lacking sex information</a></li>
<li class="chapter" data-level="8.2.4" data-path="faq.html"><a href="faq.html#insufficient-age-information"><i class="fa fa-check"></i><b>8.2.4</b> Insufficient age information</a></li>
<li class="chapter" data-level="8.2.5" data-path="faq.html"><a href="faq.html#uninformative-age-prior"><i class="fa fa-check"></i><b>8.2.5</b> Uninformative age prior</a></li>
<li class="chapter" data-level="8.2.6" data-path="faq.html"><a href="faq.html#mating-system"><i class="fa fa-check"></i><b>8.2.6</b> Mating system</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="faq.html"><a href="faq.html#death-year"><i class="fa fa-check"></i><b>8.3</b> Death year</a></li>
<li class="chapter" data-level="8.4" data-path="faq.html"><a href="faq.html#accuracy"><i class="fa fa-check"></i><b>8.4</b> Accuracy</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="faq.html"><a href="faq.html#field-pedigree"><i class="fa fa-check"></i><b>8.4.1</b> Field pedigree</a></li>
<li class="chapter" data-level="8.4.2" data-path="faq.html"><a href="faq.html#genomic-relatedness-1"><i class="fa fa-check"></i><b>8.4.2</b> Genomic relatedness</a></li>
<li class="chapter" data-level="8.4.3" data-path="faq.html"><a href="faq.html#data-simulation"><i class="fa fa-check"></i><b>8.4.3</b> Data simulation</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="faq.html"><a href="faq.html#add-new-individuals"><i class="fa fa-check"></i><b>8.5</b> Add new individuals</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>9</b> References</a></li>
<li class="divider"></li>
<li><a href="./articles/vignette_age/book/index.html">Ageprior Vignette</a></li>
<li><a href="./../../index.html">Other Manuals</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Sequoia User Guide</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="faq" class="section level1" number="8">
<h1><span class="header-section-number">8</span> FAQ</h1>
<hr />
<div id="error-aboutdue-to-input-data" class="section level2" number="8.1">
<h2><span class="header-section-number">8.1</span> Error about/due to input data</h2>
<p>For smaller datasets (say up to 1000 individuals) it may be useful to read in the data with <code>read.table()</code> (using <code>header=TRUE</code> or <code>FALSE</code>, as appropriate, and typically <code>row.names=1</code>), and inspect the data with <code>table(as.matrix(mydata))</code> for odd entries that weren’t caught by <code>CheckGeno()</code> to give an informative error message. For very large datasets, specialist tools may be required to get the data in a standardised format.</p>
</div>
<div id="sec:FAQlowAR" class="section level2" number="8.2">
<h2><span class="header-section-number">8.2</span> Low assignment rate</h2>
<p>See also <a href="sec:lowAR">key points</a>.</p>
<p>In many cases, assignment rate can be boosted without increasing the number of SNPs: by assuming a different genotyping error rate, providing sex or birth year information on more individuals, or fine-tuning the ageprior. Sometimes <em>reducing</em> the number of SNPs may increase assignment rate, by removing those with the highest error rate and/or SNPs in close LD.</p>
<div id="not-enough-snps" class="section level3" number="8.2.1">
<h3><span class="header-section-number">8.2.1</span> Not enough SNPs</h3>
<p>As opposed to most other pedigree assignment programs, Sequoia does not rely on MCMC to explore many different pedigree possibilities, but instead sequentially assigns highly likely relationships, and expands the pedigree step by step. For a relationship to be highly likely, a substantial number of SNPs is necessary, larger than for MCMC methods: at least 100-200 for parentage assignment, or full sibling clustering in a monogamous population, and about 400-500 otherwise. More is not always better: with tens of thousands of SNPs many will unavoidably be in high LD (unless you work on a species with a very large genome and very high <span class="math inline">\(N_e\)</span>), and the signal will get lost in the noise. Using more than 1000 SNPs is never necessary, and for a typical vertebrate genome there is little benefit of using more than 500 good SNPs.</p>
<p>This unfortunately also means that sequoia is unsuitable for species with a very small genome.</p>
</div>
<div id="genotyping-errors" class="section level3" number="8.2.2">
<h3><span class="header-section-number">8.2.2</span> Genotyping errors</h3>
<p>A few SNPs with a high (apparent) error rate may throw off pedigree reconstruction if their erroneous signal is not sufficiently offset by a large number of more accurate SNPs. Such SNPs can be identified with <code>SnpStats()</code>, using an existing pedigree or one from an preliminary round of parentage assignment. If there are clear outliers with regards to estimated error rate, it may be worth exploring pedigree reconstruction without these SNPs.</p>
</div>
<div id="lacking-sex-information" class="section level3" number="8.2.3">
<h3><span class="header-section-number">8.2.3</span> Lacking sex information</h3>
<p>Currently Sequoia cannot handle sex-linked markers, and therefore cannot distinguish between maternal versus paternal relatives. Sometimes the sex of an individual can be inferred, if it forms a complementary parent-pair with an individual of known sex. This is also the manner in which the sex of dummy parents is determined; half-sibships sharing a parent of unknown sex are not currently implemented.</p>
<p>Pairs of relatives for which it is unclear whether they are maternal or paternal relatives, can be identified with <code>GetMaybeRel()</code>.</p>
</div>
<div id="insufficient-age-information" class="section level3" number="8.2.4">
<h3><span class="header-section-number">8.2.4</span> Insufficient age information</h3>
<p>A parent will only be assigned if it is known to be older than the individual with which it genetically forms a parent–offspring pair, or if a complementary co-parent is identified. Purely genetically, it is impossible to tell who is the parent in a parent–offspring pair. But once an individual has been assigned parents, any remaining individuals with which it forms a parent-offspring pair must be its offspring, and will be assigned as such. Thus, a high proportion of sampled parents may somewhat compensate for unknown birth years.</p>
<p>Providing minimum and maximum possible birth years can substantially increase assignment rate. The risk is that when intervals are taken too narrowly, parent–offspring pairs are flipped the wrong way around, which in turn may lead to other wrong assignments. This may especially be a problem in long-lived species which start breeding at an early age.</p>
<p>Age information will also help to distinguish between the three different types of second degree relatives (half siblings, grandparent – grand-offspring and full avuncular (aunt/uncle – niece/nephew). These are genetically indistinguishable, unless both individuals of the pair already have a parent assigned. In most species, there is limited overlap between the age-difference of half siblings versus grandparents, and only partial overlap of either with the age distribution of avuncular relationships.</p>
</div>
<div id="uninformative-age-prior" class="section level3" number="8.2.5">
<h3><span class="header-section-number">8.2.5</span> Uninformative age prior</h3>
<p>The ageprior used for full pedigree reconstruction is estimated from the birth year differences of parent-offspring pairs assigned during the initial parentage assignment. When only few parents are assigned, or when many birth years are unknown, this ageprior may not be very informative. If a large pedigree is available from the same or a similar population (e.g. based on observations and microsatellite paternity assignments), it can be useful to estimate the agepriors from that pedigree. For details, see the <a href="./articles/vignette_age/book/index.html">age vignette</a>.</p>
</div>
<div id="mating-system" class="section level3" number="8.2.6">
<h3><span class="header-section-number">8.2.6</span> Mating system</h3>
<p>When the population has a complex mating system, with overlapping generations and many double relatives, a large number of SNPs is needed to distinguish between various plausible alternatives. When the power of the SNP panel is insufficient to make the distinction, no assignment will be made.</p>
<p>When inbreeding and complex relationships (e.g. paternal half-sibling as well as maternal half-aunt) are rare, ignoring these typically increases assignment rate (<code>Complex="simp"</code>). Similarly, when polygamy is rare and not of particular interest, assuming a monogamous mating system typically increases assignment rate (<code>Complex="mono"</code>). However, these choices will risk erroneous assignments when complex relationships or polygamy, respectively, do occur.</p>
</div>
</div>
<div id="death-year" class="section level2" number="8.3">
<h2><span class="header-section-number">8.3</span> Death year</h2>
<p>The year of death forms an upper limit to when an individual could have reproduced, and is used by e.g. <a href="https://www.bioinf.uni-leipzig.de/Software/FRANz/About.html">FRANZ</a> during parentage assignment. It is not used by sequoia, because it is mainly focussed on populations of wild animals, where it is often impossible to tell whether an individual has emigrated or died, and identification of dead individuals is not always reliable. Emigrants are still candidate parents, as they may reside just outside the study area boundaries, or return briefly and unseen during the breeding season.</p>
<p>The only way in which a known death date can inform pedigree reconstruction by sequoia, is as an upper limit to the estimated birth year, when the exact birth year is unknown.</p>
</div>
<div id="accuracy" class="section level2" number="8.4">
<h2><span class="header-section-number">8.4</span> Accuracy</h2>
<p>The real, true pedigree is normally unknown, but there are a few ways to infer the accuracy of specific pedigree links, or to estimate the average accuracy of assignments.</p>
<div id="field-pedigree" class="section level3" number="8.4.1">
<h3><span class="header-section-number">8.4.1</span> Field pedigree</h3>
<p>If the genetically assigned mother matches the mother caring for the individual, or the plant from which the seed was collected, there is little reason to doubt the assignment. Similarly, when the genetically assigned father matches (one of) the observed mates of the mother, the assignment is most likely correct.</p>
<p>In rare other cases, the genetically assigned parent is impossible, for example because the assigned parent was not alive at the time of birth (for mothers) or conception (for fathers). The blame for such an erroneous assignment may be the pedigree reconstruction software (due to genotyping errors, or a bug in the code), but may also be due to sample mislabelling in the lab, or a case of mistaken identity in the field.</p>
<p>Tracking down the underlying cause of discrepancies between the field and genetic pedigree will be time consuming, but can be a valuable part of data quality control.</p>
</div>
<div id="genomic-relatedness-1" class="section level3" number="8.4.2">
<h3><span class="header-section-number">8.4.2</span> Genomic relatedness</h3>
<p>When many thousands of SNPs are typed, it is possible to calculate the genomic relatedness (<span class="math inline">\(R_{grm}\)</span>) between all pairs of individuals (see <a href="output-check.html#sec:compR">Relatedness</a>). Due to the random nature of Mendelian inheritance there is always considerable scatter of genomic relatedness around pedigree relatedness, but when the pedigree relatedness is considerably higher (say <span class="math inline">\(R_{ped} - R_{grm} &gt;0.2\)</span>), this is often indicative of a pedigree error. Note however that most estimators of <span class="math inline">\(R_{grm}\)</span> assume a large, panmictic, non-inbred population, and deviations from these assumptions may contribute to differences between <span class="math inline">\(R_{ped}\)</span> and <span class="math inline">\(R_{grm}\)</span>.</p>
</div>
<div id="data-simulation" class="section level3" number="8.4.3">
<h3><span class="header-section-number">8.4.3</span> Data simulation</h3>
<p>Simulations as performed by ‘EstConf()’ do not tell which assignments may be incorrect, but do give an estimate of the overall number of incorrect assignments. The simulations are done presuming the inferred pedigree (or an existing pedigree) is the true pedigree, i.e. for a pedigree that is (hopefully) very close to the actual true pedigree. It relies on several simplifying assumptions, and will therefore always be an optimistic estimate of the actual accuracy.</p>
</div>
</div>
<div id="add-new-individuals" class="section level2" number="8.5">
<h2><span class="header-section-number">8.5</span> Add new individuals</h2>
<p>When new individuals have been genotyped, such as a new cohort of offspring, it is best to re-run the pedigree reconstruction for all genotyped individuals. This ensures that older siblings of the new offspring are identified, as well as any grandparents. A possible exception is when all candidate parents have been genotyped, although even then inclusion of their parents (the new individuals’ grandparents) may correct for any genotyping errors they may have.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="hermaphrodites.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"info": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
