<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 Function overview | Sequoia User Guide</title>
  <meta name="description" content="3 Function overview | Sequoia User Guide" />
  <meta name="generator" content="bookdown 0.22 and GitBook 2.6.7" />

  <meta property="og:title" content="3 Function overview | Sequoia User Guide" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="/../../sequoia_logo.png" />
  
  <meta name="github-repo" content="JiscaH/sequoia" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 Function overview | Sequoia User Guide" />
  
  
  <meta name="twitter:image" content="/../../sequoia_logo.png" />

<meta name="author" content="Jisca Huisman" />


<meta name="date" content="2021-05-02" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  <link rel="apple-touch-icon-precomposed" sizes="120x120" href="../../apple-touch-icon-120x120.png" />
  <link rel="shortcut icon" href="../../favicon.ico" type="image/x-icon" />
<link rel="prev" href="key-points.html"/>
<link rel="next" href="sequoia-input.html"/>
<script src="libs/header-attrs-2.8/header-attrs.js"></script>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />









<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./../../../index.html"><img src="../../home-solid-silver.svg" height="40" alt="home"/></a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="1" data-path="sec-Background.html"><a href="sec-Background.html"><i class="fa fa-check"></i><b>1</b> Background</a></li>
<li class="chapter" data-level="2" data-path="key-points.html"><a href="key-points.html"><i class="fa fa-check"></i><b>2</b> Key Points</a>
<ul>
<li class="chapter" data-level="2.1" data-path="key-points.html"><a href="key-points.html#sec:MendelErrors"><i class="fa fa-check"></i><b>2.1</b> Mendelian Errors</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="key-points.html"><a href="key-points.html#parent-pairs"><i class="fa fa-check"></i><b>2.1.1</b> Parent-pairs</a></li>
<li class="chapter" data-level="2.1.2" data-path="key-points.html"><a href="key-points.html#maximum-mismatches"><i class="fa fa-check"></i><b>2.1.2</b> Maximum mismatches</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="key-points.html"><a href="key-points.html#likelihoods"><i class="fa fa-check"></i><b>2.2</b> Likelihoods</a></li>
<li class="chapter" data-level="2.3" data-path="key-points.html"><a href="key-points.html#sec:GenoErrors"><i class="fa fa-check"></i><b>2.3</b> Genotyping errors</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="key-points.html"><a href="key-points.html#effect"><i class="fa fa-check"></i><b>2.3.1</b> Effect</a></li>
<li class="chapter" data-level="2.3.2" data-path="key-points.html"><a href="key-points.html#assumed-genotyping-error-rate-err"><i class="fa fa-check"></i><b>2.3.2</b> Assumed genotyping error rate <code>Err</code></a></li>
<li class="chapter" data-level="2.3.3" data-path="key-points.html"><a href="key-points.html#performance-with-high-genotyping-error-rates"><i class="fa fa-check"></i><b>2.3.3</b> Performance with high genotyping error rates</a></li>
<li class="chapter" data-level="2.3.4" data-path="key-points.html"><a href="key-points.html#effect-on-runtime"><i class="fa fa-check"></i><b>2.3.4</b> Effect on runtime</a></li>
<li class="chapter" data-level="2.3.5" data-path="key-points.html"><a href="key-points.html#error-matrix"><i class="fa fa-check"></i><b>2.3.5</b> Error matrix</a></li>
<li class="chapter" data-level="2.3.6" data-path="key-points.html"><a href="key-points.html#estimate"><i class="fa fa-check"></i><b>2.3.6</b> Estimate</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="key-points.html"><a href="key-points.html#sec:BY-AP"><i class="fa fa-check"></i><b>2.4</b> Birth years &amp; Ageprior</a>
<ul>
<li class="chapter" data-level="2.4.1" data-path="key-points.html"><a href="key-points.html#definition-interpretation"><i class="fa fa-check"></i><b>2.4.1</b> Definition &amp; interpretation</a></li>
<li class="chapter" data-level="2.4.2" data-path="key-points.html"><a href="key-points.html#implementation"><i class="fa fa-check"></i><b>2.4.2</b> Implementation</a></li>
<li class="chapter" data-level="2.4.3" data-path="key-points.html"><a href="key-points.html#smooth-flatten"><i class="fa fa-check"></i><b>2.4.3</b> Smooth &amp; Flatten</a></li>
<li class="chapter" data-level="2.4.4" data-path="key-points.html"><a href="key-points.html#age-based-non-assignment"><i class="fa fa-check"></i><b>2.4.4</b> Age-based non-assignment</a></li>
</ul></li>
<li class="chapter" data-level="2.5" data-path="key-points.html"><a href="key-points.html#parent-llr"><i class="fa fa-check"></i><b>2.5</b> Parent LLR</a></li>
<li class="chapter" data-level="2.6" data-path="key-points.html"><a href="key-points.html#sec:EstConf"><i class="fa fa-check"></i><b>2.6</b> Confidence</a></li>
<li class="chapter" data-level="2.7" data-path="key-points.html"><a href="key-points.html#sec:lowAR"><i class="fa fa-check"></i><b>2.7</b> Low assignment rate</a>
<ul>
<li class="chapter" data-level="2.7.1" data-path="key-points.html"><a href="key-points.html#causes"><i class="fa fa-check"></i><b>2.7.1</b> Causes</a></li>
<li class="chapter" data-level="2.7.2" data-path="key-points.html"><a href="key-points.html#remedies"><i class="fa fa-check"></i><b>2.7.2</b> Remedies</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="3" data-path="function-overview.html"><a href="function-overview.html"><i class="fa fa-check"></i><b>3</b> Function overview</a>
<ul>
<li class="chapter" data-level="3.1" data-path="function-overview.html"><a href="function-overview.html#input"><i class="fa fa-check"></i><b>3.1</b> Input</a></li>
<li class="chapter" data-level="3.2" data-path="function-overview.html"><a href="function-overview.html#simulate"><i class="fa fa-check"></i><b>3.2</b> Simulate</a></li>
<li class="chapter" data-level="3.3" data-path="function-overview.html"><a href="function-overview.html#ageprior"><i class="fa fa-check"></i><b>3.3</b> Ageprior</a></li>
<li class="chapter" data-level="3.4" data-path="function-overview.html"><a href="function-overview.html#pedigree-reconstruction"><i class="fa fa-check"></i><b>3.4</b> Pedigree reconstruction</a></li>
<li class="chapter" data-level="3.5" data-path="function-overview.html"><a href="function-overview.html#pedigree-check"><i class="fa fa-check"></i><b>3.5</b> Pedigree check</a></li>
<li class="chapter" data-level="3.6" data-path="function-overview.html"><a href="function-overview.html#pairwise-relationships"><i class="fa fa-check"></i><b>3.6</b> Pairwise relationships</a></li>
<li class="chapter" data-level="3.7" data-path="function-overview.html"><a href="function-overview.html#miscellaneous"><i class="fa fa-check"></i><b>3.7</b> Miscellaneous</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="sequoia-input.html"><a href="sequoia-input.html"><i class="fa fa-check"></i><b>4</b> <code>sequoia()</code> input</a>
<ul>
<li class="chapter" data-level="4.1" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:GenoFormat"><i class="fa fa-check"></i><b>4.1</b> Genotypes</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="sequoia-input.html"><a href="sequoia-input.html#subset-snps"><i class="fa fa-check"></i><b>4.1.1</b> Subset SNPs</a></li>
<li class="chapter" data-level="4.1.2" data-path="sequoia-input.html"><a href="sequoia-input.html#plink"><i class="fa fa-check"></i><b>4.1.2</b> PLINK</a></li>
<li class="chapter" data-level="4.1.3" data-path="sequoia-input.html"><a href="sequoia-input.html#low-call-rate"><i class="fa fa-check"></i><b>4.1.3</b> Low call rate</a></li>
<li class="chapter" data-level="4.1.4" data-path="sequoia-input.html"><a href="sequoia-input.html#very-large-datasets"><i class="fa fa-check"></i><b>4.1.4</b> Very large datasets</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="sequoia-input.html"><a href="sequoia-input.html#birth-years-sex"><i class="fa fa-check"></i><b>4.2</b> Birth years &amp; sex</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="sequoia-input.html"><a href="sequoia-input.html#lifehistdata"><i class="fa fa-check"></i><b>4.2.1</b> LifeHistData</a></li>
<li class="chapter" data-level="4.2.2" data-path="sequoia-input.html"><a href="sequoia-input.html#unknown-birth-years"><i class="fa fa-check"></i><b>4.2.2</b> Unknown birth years</a></li>
<li class="chapter" data-level="4.2.3" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:custom-AP"><i class="fa fa-check"></i><b>4.2.3</b> <code>args.AP</code> (Customising the ageprior)</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="sequoia-input.html"><a href="sequoia-input.html#tfilter-tassign"><i class="fa fa-check"></i><b>4.3</b> <code>Tfilter</code> &amp; <code>Tassign</code></a></li>
<li class="chapter" data-level="4.4" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:Complx"><i class="fa fa-check"></i><b>4.4</b> <code>Complex</code></a>
<ul>
<li class="chapter" data-level="4.4.1" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:WeirdRels"><i class="fa fa-check"></i><b>4.4.1</b> Unusual relationships</a></li>
</ul></li>
<li class="chapter" data-level="4.5" data-path="sequoia-input.html"><a href="sequoia-input.html#herm"><i class="fa fa-check"></i><b>4.5</b> <code>Herm</code></a></li>
<li class="chapter" data-level="4.6" data-path="sequoia-input.html"><a href="sequoia-input.html#useage"><i class="fa fa-check"></i><b>4.6</b> <code>UseAge</code></a></li>
<li class="chapter" data-level="4.7" data-path="sequoia-input.html"><a href="sequoia-input.html#sec:reuse"><i class="fa fa-check"></i><b>4.7</b> <code>SeqList</code></a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="running-sequoia.html"><a href="running-sequoia.html"><i class="fa fa-check"></i><b>5</b> Running <code>sequoia()</code></a>
<ul>
<li class="chapter" data-level="5.1" data-path="running-sequoia.html"><a href="running-sequoia.html#module-maxsibiter"><i class="fa fa-check"></i><b>5.1</b> <code>Module</code> (<code>MaxSibIter</code>)</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="running-sequoia.html"><a href="running-sequoia.html#sec:MaxSibIter"><i class="fa fa-check"></i><b>5.1.1</b> <code>MaxSibIter</code></a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="running-sequoia.html"><a href="running-sequoia.html#input-check"><i class="fa fa-check"></i><b>5.2</b> Input check</a></li>
<li class="chapter" data-level="5.3" data-path="running-sequoia.html"><a href="running-sequoia.html#check-for-duplicates"><i class="fa fa-check"></i><b>5.3</b> Check for duplicates</a></li>
<li class="chapter" data-level="5.4" data-path="running-sequoia.html"><a href="running-sequoia.html#parentage-assignment"><i class="fa fa-check"></i><b>5.4</b> Parentage assignment</a></li>
<li class="chapter" data-level="5.5" data-path="running-sequoia.html"><a href="running-sequoia.html#ageprior-1"><i class="fa fa-check"></i><b>5.5</b> AgePrior</a></li>
<li class="chapter" data-level="5.6" data-path="running-sequoia.html"><a href="running-sequoia.html#full-reconstruction"><i class="fa fa-check"></i><b>5.6</b> Full reconstruction</a>
<ul>
<li class="chapter" data-level="5.6.1" data-path="running-sequoia.html"><a href="running-sequoia.html#dummy-individuals"><i class="fa fa-check"></i><b>5.6.1</b> Dummy Individuals</a></li>
<li class="chapter" data-level="5.6.2" data-path="running-sequoia.html"><a href="running-sequoia.html#total-likelihood"><i class="fa fa-check"></i><b>5.6.2</b> Total likelihood</a></li>
</ul></li>
<li class="chapter" data-level="5.7" data-path="running-sequoia.html"><a href="running-sequoia.html#save-output"><i class="fa fa-check"></i><b>5.7</b> Save output</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="output-check.html"><a href="output-check.html"><i class="fa fa-check"></i><b>6</b> Output check</a>
<ul>
<li class="chapter" data-level="6.1" data-path="output-check.html"><a href="output-check.html#pedigree-stats-plots"><i class="fa fa-check"></i><b>6.1</b> Pedigree stats &amp; plots</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="output-check.html"><a href="output-check.html#negative-llrs"><i class="fa fa-check"></i><b>6.1.1</b> Negative LLR’s</a></li>
<li class="chapter" data-level="6.1.2" data-path="output-check.html"><a href="output-check.html#dyad-plot"><i class="fa fa-check"></i><b>6.1.2</b> Dyad plot</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="output-check.html"><a href="output-check.html#sec:compPed"><i class="fa fa-check"></i><b>6.2</b> Compare pedigrees</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="output-check.html"><a href="output-check.html#dummy-matching"><i class="fa fa-check"></i><b>6.2.1</b> Dummy matching</a></li>
<li class="chapter" data-level="6.2.2" data-path="output-check.html"><a href="output-check.html#example"><i class="fa fa-check"></i><b>6.2.2</b> Example</a></li>
<li class="chapter" data-level="6.2.3" data-path="output-check.html"><a href="output-check.html#colony"><i class="fa fa-check"></i><b>6.2.3</b> Colony</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="output-check.html"><a href="output-check.html#sec:compR"><i class="fa fa-check"></i><b>6.3</b> Relatedness</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="output-check.html"><a href="output-check.html#sec:Rped"><i class="fa fa-check"></i><b>6.3.1</b> Pedigree relatedness</a></li>
<li class="chapter" data-level="6.3.2" data-path="output-check.html"><a href="output-check.html#genomic-relatedness"><i class="fa fa-check"></i><b>6.3.2</b> Genomic relatedness</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="output-check.html"><a href="output-check.html#cluster-families"><i class="fa fa-check"></i><b>6.4</b> Cluster families</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="hermaphrodites.html"><a href="hermaphrodites.html"><i class="fa fa-check"></i><b>7</b> Hermaphrodites</a>
<ul>
<li class="chapter" data-level="7.1" data-path="hermaphrodites.html"><a href="hermaphrodites.html#herm-a-pedigree-prior"><i class="fa fa-check"></i><b>7.1</b> Herm ‘A’: Pedigree prior</a></li>
<li class="chapter" data-level="7.2" data-path="hermaphrodites.html"><a href="hermaphrodites.html#herm-b-ignore-sex-role"><i class="fa fa-check"></i><b>7.2</b> Herm ‘B’: Ignore sex role</a></li>
<li class="chapter" data-level="7.3" data-path="hermaphrodites.html"><a href="hermaphrodites.html#hermaphrodite-dummies"><i class="fa fa-check"></i><b>7.3</b> Hermaphrodite dummies</a></li>
<li class="chapter" data-level="7.4" data-path="hermaphrodites.html"><a href="hermaphrodites.html#assignment-errors"><i class="fa fa-check"></i><b>7.4</b> Assignment errors</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="faq.html"><a href="faq.html"><i class="fa fa-check"></i><b>8</b> FAQ</a>
<ul>
<li class="chapter" data-level="8.1" data-path="faq.html"><a href="faq.html#error-aboutdue-to-input-data"><i class="fa fa-check"></i><b>8.1</b> Error about/due to input data</a></li>
<li class="chapter" data-level="8.2" data-path="faq.html"><a href="faq.html#sec:FAQlowAR"><i class="fa fa-check"></i><b>8.2</b> Low assignment rate</a>
<ul>
<li class="chapter" data-level="8.2.1" data-path="faq.html"><a href="faq.html#not-enough-snps"><i class="fa fa-check"></i><b>8.2.1</b> Not enough SNPs</a></li>
<li class="chapter" data-level="8.2.2" data-path="faq.html"><a href="faq.html#genotyping-errors"><i class="fa fa-check"></i><b>8.2.2</b> Genotyping errors</a></li>
<li class="chapter" data-level="8.2.3" data-path="faq.html"><a href="faq.html#lacking-sex-information"><i class="fa fa-check"></i><b>8.2.3</b> Lacking sex information</a></li>
<li class="chapter" data-level="8.2.4" data-path="faq.html"><a href="faq.html#insufficient-age-information"><i class="fa fa-check"></i><b>8.2.4</b> Insufficient age information</a></li>
<li class="chapter" data-level="8.2.5" data-path="faq.html"><a href="faq.html#uninformative-age-prior"><i class="fa fa-check"></i><b>8.2.5</b> Uninformative age prior</a></li>
<li class="chapter" data-level="8.2.6" data-path="faq.html"><a href="faq.html#mating-system"><i class="fa fa-check"></i><b>8.2.6</b> Mating system</a></li>
</ul></li>
<li class="chapter" data-level="8.3" data-path="faq.html"><a href="faq.html#death-year"><i class="fa fa-check"></i><b>8.3</b> Death year</a></li>
<li class="chapter" data-level="8.4" data-path="faq.html"><a href="faq.html#accuracy"><i class="fa fa-check"></i><b>8.4</b> Accuracy</a>
<ul>
<li class="chapter" data-level="8.4.1" data-path="faq.html"><a href="faq.html#field-pedigree"><i class="fa fa-check"></i><b>8.4.1</b> Field pedigree</a></li>
<li class="chapter" data-level="8.4.2" data-path="faq.html"><a href="faq.html#genomic-relatedness-1"><i class="fa fa-check"></i><b>8.4.2</b> Genomic relatedness</a></li>
<li class="chapter" data-level="8.4.3" data-path="faq.html"><a href="faq.html#data-simulation"><i class="fa fa-check"></i><b>8.4.3</b> Data simulation</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="faq.html"><a href="faq.html#add-new-individuals"><i class="fa fa-check"></i><b>8.5</b> Add new individuals</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i><b>9</b> References</a></li>
<li class="divider"></li>
<li><a href="./articles/vignette_age/book/index.html">Ageprior Vignette</a></li>
<li><a href="./../../index.html">Other Manuals</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Sequoia User Guide</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="function-overview" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Function overview</h1>
<hr />
<table>
<caption><span id="tab:CoreFun">Table 3.1: </span>Core functions with selected subset of their required (Y) and optional (+) input, and their output that can be re-used as input in other functions.</caption>
<thead>
<tr class="header">
<th align="left">Function</th>
<th align="center">GenoM</th>
<th align="center">Pedigree</th>
<th align="center">Pairs</th>
<th align="center">AgePrior</th>
<th align="center">SeqList</th>
<th align="left">Reusable output</th>
<th align="left">Plot function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">sequoia</td>
<td align="center">Y</td>
<td align="center"></td>
<td align="center"></td>
<td align="center">+</td>
<td align="center">+</td>
<td align="left">Pedigree, SeqList</td>
<td align="left">SeqListSummary</td>
</tr>
<tr class="even">
<td align="left">GetMaybeRel</td>
<td align="center">Y</td>
<td align="center"></td>
<td align="center"></td>
<td align="center">+</td>
<td align="center">+</td>
<td align="left">Pairs</td>
<td align="left">PlotRelPairs</td>
</tr>
<tr class="odd">
<td align="left">CalcOHLLR</td>
<td align="center">Y</td>
<td align="center">Y</td>
<td align="center"></td>
<td align="center">+</td>
<td align="center">+</td>
<td align="left"></td>
<td align="left">SeqListSummary</td>
</tr>
<tr class="even">
<td align="left">CalcPairLL</td>
<td align="center">Y</td>
<td align="center">+</td>
<td align="center">Y</td>
<td align="center">+</td>
<td align="center">+</td>
<td align="left"></td>
<td align="left">PlotPairLL</td>
</tr>
<tr class="odd">
<td align="left">MakeAgePrior</td>
<td align="center"></td>
<td align="center">+</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">AgePrior</td>
<td align="left">PlotAgePrior</td>
</tr>
<tr class="even">
<td align="left">SimGeno</td>
<td align="center"></td>
<td align="center">Y</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="left">GenoM</td>
<td align="left">SnpStats</td>
</tr>
<tr class="odd">
<td align="left">PedCompare</td>
<td align="center"></td>
<td align="center">YY</td>
<td align="center"></td>
<td align="center"></td>
<td align="center"></td>
<td align="left"></td>
<td align="left">PlotPedComp</td>
</tr>
<tr class="even">
<td align="left">ComparePairs</td>
<td align="center"></td>
<td align="center">Y+</td>
<td align="center">+</td>
<td align="center"></td>
<td align="center"></td>
<td align="left">Pairs</td>
<td align="left">PlotRelPairs</td>
</tr>
</tbody>
</table>
<div id="input" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Input</h2>
<ul>
<li><p><strong>GenoConvert</strong><br />
Read in genotype data from PLINK file, Colony file, or many user-specified formats, and return a matrix in sequoia’s (or Colony) format. <em>Does not support vcf yet</em>.</p></li>
<li><p><strong>LHConvert</strong><br />
Extract sex and birth year from PLINK file; optionally recode sex to 1=female, 2=male, check consistency with other LifeHistData, or combine family ID and individual ID into FID__IID.</p></li>
<li><p><strong>CheckGeno</strong><br />
Check that the provided genotype matrix is in the correct format, and check for low call rate samples and SNPs</p></li>
<li><p><strong>SnpStats</strong><br />
Calculate per-SNP allele frequency, missingness, and, if a pedigree provided, number of Mendelian errors.</p></li>
<li><p><strong>CalcMaxMismatch</strong><br />
Calculate the maximum expected number of mismatches for duplicate samples, and Mendelian errors for parent-offspring pairs and parent-parent-offspring trios.</p></li>
</ul>
</div>
<div id="simulate" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Simulate</h2>
<ul>
<li><p><strong>SimGeno</strong><br />
Simulate genotype data for independent SNPs. Specify pedigree, founder MAF, call rate, proportion of non-genotyped parents, genotyping error &amp; error model.</p></li>
<li><p><strong>MkGenoErrors</strong><br />
Add genotyping errors and missingness to genotype data; more fine-scale control than with SimGeno.</p></li>
</ul>
</div>
<div id="ageprior" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Ageprior</h2>
<ul>
<li><p><strong>MakeAgePrior</strong><br />
For various categories of pairwise relatives (R), calculate age-difference (A) based probability ratios <span class="math inline">\(P(A|R) / P(A) = P(R|A) / P(R)\)</span>, or how much likelier a relationship is given the age difference. It applies corrections when the skeleton-pedigree contains few/no pairs with known age difference for some relationships.</p></li>
<li><p><strong>PlotAgePrior</strong><br />
Visualise the age-difference based prior probability ratios as a heatmap.</p></li>
</ul>
</div>
<div id="pedigree-reconstruction" class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Pedigree reconstruction</h2>
<ul>
<li><p><strong>sequoia</strong><br />
Main function to run parentage assignment and full pedigree reconstruction, calls many of the other functions.</p></li>
<li><p><strong>GetMaybeRel</strong><br />
Identify pairs of individuals likely to be related, but not assigned as such in the provided pedigree. Either search only for potential parent-offspring pairs, or for all 1st and 2nd degree relatives.</p></li>
</ul>
</div>
<div id="pedigree-check" class="section level2" number="3.5">
<h2><span class="header-section-number">3.5</span> Pedigree check</h2>
<p><em>These functions can be applied to any pedigree, not just pedigrees reconstructed by sequoia. Required input between brackets</em></p>
<ul>
<li><p><strong>SummarySeq</strong> (1 pedigree)<br />
Graphical overview of the assignment rate, the proportion dummy parents, sibship sizes, parental LLR distributions, and Mendelian errors, <span class="math inline">\(+\)</span> tables with pedigree summary statistics.</p></li>
<li><p><strong>CalcOHLLR</strong> (pedigree + genotypes)<br />
Count opposite homozygous (OH) loci between parent-offspring pairs and Mendelian errors (ME) between parent-parent-offspring trios, and calculate the parental log-likelihood ratios (LLR).</p></li>
<li><p><strong>EstConf</strong> (pedigree + genotypes)<br />
Estimate assignment error rate (false positives &amp; false negatives). Using a reference pedigree, repeatedly simulate genotype data, run sequoia, and compare inferred to reference pedigree.</p></li>
<li><p><strong>getAssignCat</strong> (1 pedigree)<br />
Identify which individuals are genotyped, and which can potentially be substituted by a dummy individual. ‘Dummifiable’ are those non-genotyped individuals with at least 2 genotyped offspring, or at least 1 genotyped offspring and 1 genotyped parent.</p></li>
<li><p><strong>PedCompare</strong> (2 pedigrees)<br />
Compare 2 pedigrees, e.g. field and genetically inferred, or reference and inferred-from-simulated-data. Matches dummy parents to non-genotyped parents.</p></li>
<li><p><strong>CalcRped</strong> (1 pedigree)
This is a wrapper for <code>kinship()</code> in package <code>kinship2</code>.</p></li>
</ul>
</div>
<div id="pairwise-relationships" class="section level2" number="3.6">
<h2><span class="header-section-number">3.6</span> Pairwise relationships</h2>
<ul>
<li><p><strong>CalcPairLL</strong> (pairs + genotypes)
For each pair, calculate the log10-likelihoods of being various different types of relative, or unrelated.</p></li>
<li><p><strong>GetRelCat</strong> (1 pedigree)<br />
Determine the relationship between individual X and all other individuals in the pedigree, going up to 1 or 2 generations back.</p></li>
<li><p><strong>GetRelM</strong> (pedigree or pairs)
Generate a matrix with all pairwise relationships from a pedigree or dataframe with pairs</p></li>
<li><p><strong>ComparePairs</strong> (1 or 2 pedigrees)<br />
Compare, count and identify different types of relative pairs between two pedigrees, or within one pedigree.<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a></p></li>
</ul>
<ul>
<li><strong>PlotRelPairs</strong> (matrix)
plot pairwise relationships between all individuals, as by Colony.</li>
</ul>
</div>
<div id="miscellaneous" class="section level2" number="3.7">
<h2><span class="header-section-number">3.7</span> Miscellaneous</h2>
<ul>
<li><p><strong>PedPolish</strong><br />
Ensure all parents &amp; all genotyped individuals are included, remove duplicates, rename columns, and replace 0 by NA or v.v. Can also generate ‘filler’ parents for software that requires individuals to have either 0 or 2 parents, never 1.</p></li>
<li><p><strong>getGenerations</strong>
For each individual in a pedigree, count the number of generations since its most distant pedigree founder.</p></li>
<li><p><strong>ErrToM</strong><br />
Generate a matrix with the probabilities of observed genotypes (columns) conditional on actual genotypes (rows), or return a function to generate such matrices. The error matrix can be used as input for <code>sequoia</code> and <code>CalcOHLLR</code>, the error function as input for <code>SimGeno</code></p></li>
<li><p><strong>writeSeq</strong><br />
Write the list with sequoia output in human-readable format, either as a folder with .txt files, or as a many-tabbed excel file. The latter uses R package <code>xlsx</code>, which requires java and can (therefore) be cumbersome to install.</p></li>
<li><p><strong>writeColumns</strong><br />
write data.frame or matrix to a text file, using white space padding to keep columns aligned.</p></li>
<li><p><strong>FindFamilies</strong><br />
Add a column with family IDs (FIDs) to a pedigree, with each number denoting a cluster of connected individuals.</p></li>
<li><p><strong>PedStripFID</strong><br />
Reverse the joining of FID and IID in GenoConvert and LHConvert</p></li>
<li><p>** CalcBYprobs**
Estimate the probability that individual i with unknown birth year is born in year y, based on the birthyears of its parents and offspring.</p></li>
</ul>

</div>
</div>
<div class="footnotes">
<hr />
<ol start="2">
<li id="fn2"><p>The matrix returned by <code>DyadCompare</code>  is a subset of the matrix returned here using default settings.<a href="function-overview.html#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="key-points.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="sequoia-input.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
},
"info": true
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
