<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genotyping errors • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genotyping errors">
<meta name="description" content="Error patterns; per-locus vs per-allele error rate
">
<meta property="og:description" content="Error patterns; per-locus vs per-allele error rate
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="genotyping_errors_files/kePrint-0.0.1/kePrint.js"></script><link href="genotyping_errors_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Genotyping errors</h1>
            
      

      <div class="d-none name"><code>genotyping_errors.Rmd</code></div>
    </div>

    
    
<p>No genotyping method is 100% error free. For some methods, such as
SNP arrays, the genotyping error rate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>
is very low (say 0.01%). In those cases, the assumed error pattern will
have no noticeable effect. For other methods, such as RADseq, the error
rate is typically much higher (1%, or even 5%), and then the assumed
pattern <em>can</em> have a noticeable effect.</p>
<div class="section level2">
<h2 id="per-locus-vs-per-allele-error-rate">Per-locus vs per-allele error rate<a class="anchor" aria-label="anchor" href="#per-locus-vs-per-allele-error-rate"></a>
</h2>
<p>The error rate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>
(<code>Err</code>) throughout the <code>sequoia</code> documentation
represents the per-locus error rate; the per-allele error rate is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">E/2</annotation></semantics></math>.
There appears to be no consistency in the literature on whether the
per-locus or per-allele error rate is used, so it is worth making sure
it is defined consistently throughout your analysis pipeline.</p>
</div>
<div class="section level2">
<h2 id="error-patterns">Error patterns<a class="anchor" aria-label="anchor" href="#error-patterns"></a>
</h2>
<p>With ‘error pattern’ I mean the relative chances of a true homozygote
<span style="color: grey;"><strong>AA</strong></span> being erroneously
scored as a heterozygote <strong>Aa</strong> versus the other homozygote
<strong>aa</strong>, versus a true <span style="color: grey;"><strong>Aa</strong></span> being observed as
<strong>AA</strong> or <strong>aa</strong>.</p>
<p>The pattern of genotyping errors will depend on the genotyping method
and the software used to call SNPs. For SNP arrays, the clustering based
on the intensity of the <strong>A</strong> vs <strong>a</strong> probe
results in the chance of a true <span style="color: grey;"><strong>AA</strong></span> being scored as
<strong>Aa</strong> being about as likely as the reverse. In contrast,
for sequencing-based methods scoring a true <span style="color: grey;"><strong>Aa</strong></span> erroneously as a
homozygote is much more likely than the reverse, especially with low
coverage.</p>
<p>Because there is thus no one-size-fits-all error model,
<code>sequoia</code> offers full flexibility in specifying the error
pattern. This can be done either</p>
<ul>
<li>via a length 3 vector, if the two alleles are interchangeable with
respect to error rate (so that e.g. $P(<strong>aa</strong>|<span style="color: grey;"><strong>Aa</strong></span>) ==
P(<strong>AA</strong>|<span style="color: grey;"><strong>Aa</strong></span>)), or</li>
<li>via a 3x3 matrix, which only assumes that the heterozygotes <span style="color: grey;"><strong>Aa</strong></span> and <span style="color: grey;"><strong>aA</strong></span> are
interchangeable.</li>
</ul>
</div>
<div class="section level2">
<h2 id="default-model">Default model<a class="anchor" aria-label="anchor" href="#default-model"></a>
</h2>
<p>The default error model is based on SNP array genotyping, and from
version 2.9 onwards is defined as:</p>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
Probability of observed genotype (columns) conditional on actual
genotype (rows) and per-locus error rate E
</caption>
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
aa
</th>
<th style="text-align:left;">
Aa
</th>
<th style="text-align:left;">
AA
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;font-weight: bold;color: grey !important;">
aa
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(1-E/2)^2</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">E(1-E/2)</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(E/2)^2</annotation></semantics></math>
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: grey !important;">
Aa
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">E/2</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>E</mi></mrow><annotation encoding="application/x-tex">1-E</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">E/2</annotation></semantics></math>
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;color: grey !important;">
AA
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(E/2)^2</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">E(1-E/2)</annotation></semantics></math>
</td>
<td style="text-align:left;">
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mn>1</mn><mo>−</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(1-E/2)^2</annotation></semantics></math>
</td>
</tr>
</tbody>
</table>
<p>Which for an error rate of 5% translates to:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sequoia</span><span class="fu">::</span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act      0      1      2</span></span>
<span><span class="co">##   0 0.9500 0.0494 0.0006</span></span>
<span><span class="co">##   1 0.0250 0.9500 0.0250</span></span>
<span><span class="co">##   2 0.0006 0.0494 0.9500</span></span></code></pre>
<p>where 0, 1 or 2 refers to the number of copies of the reference
allele <strong>A</strong>.</p>
<p>Under this model, the probability that the genotype is observed
correctly is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>E</mi></mrow><annotation encoding="application/x-tex">1-E</annotation></semantics></math>,
irrespective of the true genotype. The probability that a genotype is
correct, <em>does</em> differ between observed heterozygotes and
observed homozygotes:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErM</span> <span class="op">&lt;-</span> <span class="fu">sequoia</span><span class="fu">::</span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span> <span class="fl">0.95</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">ErM</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     0     1     2 </span></span>
<span><span class="co">## 0.974 0.906 0.974</span></span></code></pre>
<p>So an observed heterozygote has a 90.6% chance to be correct, while a
observed homozygote a 97.4% chance.</p>
<div class="section level3">
<h3 id="error-model-for-sequence-based-methods">Error model for sequence-based methods<a class="anchor" aria-label="anchor" href="#error-model-for-sequence-based-methods"></a>
</h3>
<p>For RADseq and similar methods, any observed homozygote may be a
heterozygote with relatively high probability, especially with low
coverage, while any heterozygote is much less likely to be
incorrect.</p>
<p>We can specify this for example as follows [^3] (numbers are
completely random!!):</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrV_RAD</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.002</span><span class="op">^</span><span class="fl">2</span>, <span class="co"># hom|hom: true hom observed as other hom</span></span>
<span>              <span class="fl">0.002</span>,   <span class="co"># het|hom: true hom observed as het</span></span>
<span>              <span class="fl">0.05</span><span class="op">)</span>    <span class="co"># hom|het: true het observed as hom </span></span></code></pre></div>
<p>for an allelic dropout rate of 5%, and an error rate of 0.2% per
allele, giving a hom-to-other-hom error rate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mn>0.001</mn><mn>2</mn></msup><mo>=</mo><mn>1</mn><mi>e</mi><mo>−</mo><mn>6</mn></mrow><annotation encoding="application/x-tex">0.001^2 = 1e-6</annotation></semantics></math>.
This can be turned into a error matrix for double checking:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErM_RAD</span> <span class="op">&lt;-</span> <span class="fu">sequoia</span><span class="fu">::</span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err <span class="op">=</span> <span class="va">ErrV_RAD</span>, Return<span class="op">=</span><span class="st">'matrix'</span><span class="op">)</span></span>
<span><span class="va">ErM_RAD</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act        0     1        2</span></span>
<span><span class="co">##   0 0.997996 0.002 0.000004</span></span>
<span><span class="co">##   1 0.050000 0.900 0.050000</span></span>
<span><span class="co">##   2 0.000004 0.002 0.997996</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/diag.html" class="external-link">diag</a></span><span class="op">(</span><span class="va">ErM_RAD</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">ErM_RAD</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     0     1     2 </span></span>
<span><span class="co">## 0.952 0.996 0.952</span></span></code></pre>
<p>so now we have an error model where any observed heterozygote has a
0.4% probability to be incorrect, and any observed homozygote a 4.8%
probability.</p>
</div>
</div>
<div class="section level2">
<h2 id="specify-error-matrix">Specify error matrix<a class="anchor" aria-label="anchor" href="#specify-error-matrix"></a>
</h2>
<p>You can pass a 3x3 matrix to <code>Err</code>, with the only
requirement that rows must sum to 1 (each actual genotype is observed as
either <strong>aa</strong>, <strong>Aa</strong>, or
<strong>AA</strong>). You can use this if for example the homozygote for
the reference allele may be more likely to be mis-typed than the
homozygote for the alternative allele, the heterozygote may be mis-typed
with different probabilities for the two homozygotes, or whatever else
is appropriate for your dataset.</p>
<p>For example, you may have the following counts from genotyping some
samples several times:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">14</span>, <span class="fl">3</span>, <span class="fl">5</span>,</span>
<span>                     <span class="fl">4</span>, <span class="fl">28</span>, <span class="fl">9</span>,</span>
<span>                     <span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">72</span><span class="op">)</span>,</span>
<span>                   nrow<span class="op">=</span><span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ErrCount</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3]</span></span>
<span><span class="co">## [1,]   14    3    5</span></span>
<span><span class="co">## [2,]    4   28    9</span></span>
<span><span class="co">## [3,]    0    4   72</span></span></code></pre>
<p>A problem here is the cell with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>:
by leaving this as is, it would imply that true <span style="color: grey;"><strong>AA</strong></span> can never be observed as
<strong>aa</strong>, even if we would genotype thousands of samples.
That seems unlikely, so as a quick fix let’s pretend that if we had
genotyped twice as many, we would have found 1 error:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrCount</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span></code></pre></div>
<p>Then divide by rowsums, e.g. using <code><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep()</a></code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrProbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">ErrCount</span>, MARGIN<span class="op">=</span><span class="fl">1</span>, STATS<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">ErrCount</span><span class="op">)</span>, FUN<span class="op">=</span><span class="st">"/"</span><span class="op">)</span></span>
<span><span class="va">ErrProbs</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1]  [,2]  [,3]</span></span>
<span><span class="co">## [1,] 0.636 0.136 0.227</span></span>
<span><span class="co">## [2,] 0.098 0.683 0.220</span></span>
<span><span class="co">## [3,] 0.007 0.052 0.941</span></span></code></pre>
<p>and then you can use this as input for the <code><a href="../reference/sequoia.html">sequoia()</a></code>
analysis:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoData</span>, <span class="va">LifeHistData</span>,</span>
<span>                  Err <span class="op">=</span> <span class="va">ErrProbs</span><span class="op">)</span></span></code></pre></div>
<p>In reality you probably want something a bit more sophisticated,
e.g. relying on the known error pattern for your genotyping
platform.</p>
<div class="section level3">
<h3 id="do-not-use-0">Do not use 0<a class="anchor" aria-label="anchor" href="#do-not-use-0"></a>
</h3>
<p>Using a value of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>
in the error matrix is not advised, as it puts an infinite amount of
weight on the locus if it does happen after all (see also <a href="https://en.wikipedia.org/wiki/Cromwell%27s_rule" class="external-link">Cromwell’s
rule</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="specify-a-function-errflavour">Specify a function (<code>ErrFlavour</code>)<a class="anchor" aria-label="anchor" href="#specify-a-function-errflavour"></a>
</h2>
<p>If you want to run <code><a href="../reference/sequoia.html">sequoia()</a></code> with several presumed error
rates but with the same pattern, you can create a function that takes
the value(s) as input, and returns either a length 3 vector (hom|hom,
het|hom, hom|het), or a 3x3 matrix with rows summing to 1. If this
function takes a single value as input, you can pass it directly to
<code><a href="../reference/sequoia.html">sequoia()</a></code> or various other functions. Those will internally
call <code><a href="../reference/ErrToM.html">ErrToM()</a></code>, which takes care of the conversion to a 3x3
matrix.</p>
<p>For example,</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RAD_flavour</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">E</span><span class="op">)</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">E</span><span class="op">^</span><span class="fl">2</span>, <span class="va">E</span>, <span class="fl">.05</span><span class="op">+</span><span class="va">E</span><span class="op">)</span></span>
<span><span class="va">test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">Geno_griffin</span>, <span class="va">LH_griffin</span>, Module<span class="op">=</span><span class="st">'dup'</span>, quiet<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                Err<span class="op">=</span><span class="fl">0.01</span>, ErrFlavour<span class="op">=</span><span class="va">RAD_flavour</span><span class="op">)</span></span></code></pre></div>
<p>you can inspect the error matrix with</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test</span><span class="op">$</span><span class="va">ErrM</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act      0    1      2</span></span>
<span><span class="co">##   0 0.9899 0.01 0.0001</span></span>
<span><span class="co">##   1 0.0600 0.88 0.0600</span></span>
<span><span class="co">##   2 0.0001 0.01 0.9899</span></span></code></pre>
<div class="section level3">
<h3 id="per-allele-error-rate">Per-allele error rate<a class="anchor" aria-label="anchor" href="#per-allele-error-rate"></a>
</h3>
<p>If you would want to keep the default error pattern, but specify
<code>Err</code> as the per-allele rather than per-locus error rate, you
may do this as follows:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErFunc_per_allele</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ea</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Ea</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">*</span><span class="va">Ea</span> <span class="op">-</span> <span class="va">Ea</span><span class="op">^</span><span class="fl">2</span>, <span class="va">Ea</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">ErFunc_per_allele</span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0001 0.0199 0.0100</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err<span class="op">=</span><span class="fl">0.01</span>, flavour <span class="op">=</span> <span class="va">ErFunc_per_allele</span>, Return <span class="op">=</span> <span class="st">'matrix'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act      0      1      2</span></span>
<span><span class="co">##   0 0.9800 0.0199 0.0001</span></span>
<span><span class="co">##   1 0.0100 0.9800 0.0100</span></span>
<span><span class="co">##   2 0.0001 0.0199 0.9800</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># and for comparison:</span></span>
<span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err<span class="op">=</span><span class="fl">0.02</span>, Return <span class="op">=</span> <span class="st">'vector'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.0001 0.0199 0.0100</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="inbuilt-functions">Inbuilt functions<a class="anchor" aria-label="anchor" href="#inbuilt-functions"></a>
</h3>
<p>Currently the only inbuilt functions are ones that return the error
pattern in earlier versions of <code>sequoia</code>. These are described
in <code><a href="../reference/ErrToM.html">?ErrToM</a></code>, and can be used by specifying
e.g. <code>ErrFlavour = 'version1.3'</code>.</p>
<p>If you have a good reference for error patterns for RADseq etc,
please let me know as I’d be more than happy to build those in.</p>
</div>
</div>
<div class="section level2">
<h2 id="variation-between-snps">Variation between SNPs<a class="anchor" aria-label="anchor" href="#variation-between-snps"></a>
</h2>
<p>Currently a constant error rate across all SNPs is assumed. It is
unclear whether setting this presumed rate to the arithmetic mean rate
(the regular average) across SNPs gives the best performance for real
world data, or if e.g. the harmonic mean or median may be more
appropriate.</p>
<p>Anecdotal evidence suggests that using a somewhat higher presumed
error rate increases performance (more correct assignments and/or fewer
incorrect assignments), but this has not been thoroughly explored. It
may partly be due to confusion between the per-locus and per-allele
error rate, but also due to the very skewed nature of most distributions
(many alleles with low error rate, and some with high error rate),
resulting in a arithmetic mean that is lower than the median.</p>
</div>
<div class="section level2">
<h2 id="effect-of-err-on-pedigree-reconstruction">Effect of <code>Err</code> on pedigree reconstruction<a class="anchor" aria-label="anchor" href="#effect-of-err-on-pedigree-reconstruction"></a>
</h2>
<p>For each individual at each SNP the probability that the actual
genotype is 0, 1 or 2 copies of the reference allele is estimated, based
on among others its own observed genotype and the genotyping error rate
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>.
If
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>
is close to zero, the actual genotype is assumed to be known with high
certainty, and the likelihood differences between various relationships
are large. The higher
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>E</mi><annotation encoding="application/x-tex">E</annotation></semantics></math>,
the fuzzier the differences between the relationships become, and the
more challenging pedigree reconstruction is. This can be explored by
running <code>CalcPairLL</code> a few times on the same pedigree with
the same genetic data, only varying <code>Err</code>.</p>
<p>All details of calculation of likelihoods incorporating genotyping
errors are described in <span class="citation">Huisman (<a href="#ref-huisman17">2017</a>)</span>.</p>
<p>Noteworthy is that the probabilistic estimate of an individual’s true
genotype depends on its own observed genotype and its parents’ estimated
true genotypes, but <em>not on the genotypes of its offspring</em>.
Doing so would potentially create an unstable seesaw situation, where
assignment of a parent is both affected by and does affect that parent’s
estimated genotype.</p>
<p>For dummy individuals, which do not have an observed genotype, the
offspring genotypes do necessarily affect its estimated genotype. To
prevent the algorithm from seesawing, marginal probabilities are
calculated, excluding grandparents and/or one, two or all offspring from
the calculation, as appropriate for a particular relationship likelihood
(most importantly avoiding implicitly conditioning on oneself).</p>
<p>The main reason those marginal probabilities are not implemented for
genotyped individuals, is because of their relatively large
computational cost, with negligible effect on the pedigree outcome for
moderately high quality SNP data (error rate &lt; 1% and low
missingness).<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;If you are interested in a change in this, e.g. as an
optional feature, please do contact me to discuss possibilities. I am
fairly sure it would improve performance for low quality SNP data.&lt;/p&gt;"><sup>1</sup></a></p>
</div>
<div class="section level2">
<h2 id="try-out-different-values">Try out different values<a class="anchor" aria-label="anchor" href="#try-out-different-values"></a>
</h2>
<p>Due to both the variation in error rate between SNPs and how
genotyping errors are implemented, it is usually a good idea to run
<code><a href="../reference/sequoia.html">sequoia()</a></code> with a few different presumed genotyping error
rates. This allows you to make sure the results are robust against
slight mis-specifications of this parameter, and/or to optimise the
false negative and false positive assignment rates.</p>
<div class="section level3">
<h3 id="err-0">
<code>Err = 0</code><a class="anchor" aria-label="anchor" href="#err-0"></a>
</h3>
<p><code><a href="../reference/sequoia.html">sequoia()</a></code> is not guaranteed to work with an assumed
error rate of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0</mn><annotation encoding="application/x-tex">0</annotation></semantics></math>,
especially not for more complex pedigrees. The problem is that a new
assignment considers only the direct vicinity of the putative
parent-offspring pair, not the full pedigree. An assignment will have
cascading effects on the probabilities of the true, underlying genotypes
of dummy or partially genotyped individuals downstream in the pedigree,
which may result in impossibilities when <code>Err=0</code> but
genotyping errors are present.</p>
<hr>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-huisman17" class="csl-entry">
Huisman, Jisca. 2017. <span>“Pedigree Reconstruction from
<span>SNP</span> Data: Parentage Assignment, Sibship Clustering and
Beyond.”</span> <em>Molecular Ecology Resources</em> 17 (5): 1009–24.
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
