<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Error patterns; per-locus vs per-allele error rate
">
<title>Genotyping errors • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Genotyping errors">
<meta property="og:description" content="Error patterns; per-locus vs per-allele error rate
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo_s.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.5.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href=".././index.html">
    <span class="fas fa fas fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/sequoia.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/vignette_main/book/sec-Background.html">Main vignette: User guide</a>
    <a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    <a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a>
    <a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/help.html">Help</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/performance_stats.html">Stats</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa fas fa-envelope"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JiscaH/sequoia">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca">
    <span class="fab fa fab fa-mastodon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/HuismanJisca">
    <span class="fab fa fab fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Genotyping errors</h1>
            
      
      
      <div class="d-none name"><code>genotyping_errors.Rmd</code></div>
    </div>

    
    
<hr>
<div class="section level2">
<h2 id="error-patterns">Error patterns<a class="anchor" aria-label="anchor" href="#error-patterns"></a>
</h2>
<p>The pattern of genotyping errors will depend on the genotyping method
and the software used to call SNPs. For SNP arrays, the clustering based
on the intensity of the <strong>A</strong> vs <strong>a</strong> probe
results in the chance of a true homozygote <strong>AA</strong> being
erroneously scored as a heterozygote <strong>Aa</strong> being about as
likely as the reverse. In contrast, for sequencing-based methods scoring
a true heterozygote erroneously as a homozygote is much more likely than
the reverse, especially with low coverage.</p>
<p>Because there is no one-size-fits-all error model for SNP genotypes,
<code>sequoia</code> offers full flexibility in specifying the error
pattern. This is done via a 3x3 matrix (heterozygotes
<strong>Aa</strong> and <strong>aA</strong> are interchangeable) with
the probabilities of observing genotype <span class="math inline">\(y\)</span> given true genotype <span class="math inline">\(x\)</span>.</p>
</div>
<div class="section level2">
<h2 id="default-model">Default model<a class="anchor" aria-label="anchor" href="#default-model"></a>
</h2>
<p>The default error model is based on SNP array genotyping, and from
version 2.0 onwards is defined as:</p>
<table class="table">
<caption>Probability of observed genotype (columns) conditional on
actual genotype (rows) and per-locus error rate E</caption>
<thead><tr class="header">
<th align="left"></th>
<th align="left"><strong>aa</strong></th>
<th align="left"><strong>Aa</strong></th>
<th align="left"><strong>AA</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>aa</strong></td>
<td align="left"><span class="math inline">\((1-E/2)^2\)</span></td>
<td align="left"><span class="math inline">\(E(1-E/2)\)</span></td>
<td align="left"><span class="math inline">\((E/2)^2\)</span></td>
</tr>
<tr class="even">
<td align="left"><strong>Aa</strong></td>
<td align="left"><span class="math inline">\(E/2\)</span></td>
<td align="left"><span class="math inline">\(1-E\)</span></td>
<td align="left"><span class="math inline">\(E/2\)</span></td>
</tr>
<tr class="odd">
<td align="left"><strong>AA</strong></td>
<td align="left"><span class="math inline">\((E/2)^2\)</span></td>
<td align="left"><span class="math inline">\(E(1-E/2)\)</span></td>
<td align="left"><span class="math inline">\((1-E/2)^2\)</span></td>
</tr>
</tbody>
</table>
<p>Which for an error rate of 1% translates to:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sequoia</span><span class="fu">::</span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       obs-0|act obs-1|act obs-2|act</span></span>
<span><span class="co">## act-0  0.990025   0.00995  0.000025</span></span>
<span><span class="co">## act-1  0.005000   0.99000  0.005000</span></span>
<span><span class="co">## act-2  0.000025   0.00995  0.990025</span></span></code></pre>
<p>where 0, 1 or 2 refers to the number of copies of the reference
allele <strong>A</strong>.</p>
<p>When the true genotype is homozygote <strong>aa</strong> (top row),
the probabilities are derived as follows:</p>
<ul>
<li><p><strong>aa</strong>: the probability that neither allele is
observed incorrectly, i.e. <span class="math inline">\((1-E/2) \times
(1-E/2)\)</span></p></li>
<li><p><strong>Aa</strong>: the probability that the first allele is
mis-observed (<span class="math inline">\(E/2\)</span>) and the second
allele is not (<span class="math inline">\(1-E/2\)</span>), plus
<em>vice versa</em>. Thus, <span class="math inline">\(P=2 \times (E/2)
\times (1-E/2) = E (1-E/2)\)</span>.</p></li>
<li><p><strong>AA</strong>: the probability that both alleles are
observed incorrectly</p></li>
</ul>
<div class="section level3">
<h3 id="error-model-in-previous-versions">Error model in previous versions<a class="anchor" aria-label="anchor" href="#error-model-in-previous-versions"></a>
</h3>
<p><span class="math inline">\((E/2)^2 \approx 0\)</span> and <span class="math inline">\((1-E/2)^2 \approx 1-E\)</span>, as illustrated in
the example for <code>Err=0.01</code>. This implies that the current
default error model does not differ as much from the simpler error model
in version 0.9 as may seem at first glance:</p>
<table class="table">
<caption>As previous table, for sequoia 0.9</caption>
<thead><tr class="header">
<th align="left"></th>
<th align="left"><strong>aa</strong></th>
<th align="left"><strong>Aa</strong></th>
<th align="left"><strong>AA</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>aa</strong></td>
<td align="left"><span class="math inline">\(1-E\)</span></td>
<td align="left"><span class="math inline">\(E\)</span></td>
<td align="left"><span class="math inline">\(0\)</span></td>
</tr>
<tr class="even">
<td align="left"><strong>Aa</strong></td>
<td align="left"><span class="math inline">\(E/2\)</span></td>
<td align="left"><span class="math inline">\(1-E\)</span></td>
<td align="left"><span class="math inline">\(E/2\)</span></td>
</tr>
<tr class="odd">
<td align="left"><strong>AA</strong></td>
<td align="left"><span class="math inline">\(0\)</span></td>
<td align="left"><span class="math inline">\(E\)</span></td>
<td align="left"><span class="math inline">\(1-E\)</span></td>
</tr>
</tbody>
</table>
<p>Using a value of <span class="math inline">\(0\)</span> in the error
matrix is not advised, as it puts an infinite amount of weight on the
locus if it does happen after all (see also <a href="https://en.wikipedia.org/wiki/Cromwell%27s_rule" class="external-link">Cromwell’s
rule</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="per-locus-vs-per-allele-error-rate">Per-locus vs per-allele error rate<a class="anchor" aria-label="anchor" href="#per-locus-vs-per-allele-error-rate"></a>
</h2>
<p>The error rate <span class="math inline">\(E\)</span>
(<code>Err</code>) in the previous part represents the per-locus error
rate; the per-allele error rate is <span class="math inline">\(E/2\)</span>. There appears to be no consistency
in the literature on whether the per-locus or per-allele error rate is
used, so it is worth making sure it is defined consistently throughout
your analysis pipeline. For pedigree reconstruction with
<code>sequoia</code>, being off by a factor 2 should generally not have
large consequences.</p>
<p>If you prefer, you could easily re-define the error matrix in terms
of a per-allele error rate (say <span class="math inline">\(\epsilon\)</span>):</p>
<table class="table">
<caption>Error matrix redefined with per-allele error rate <span class="math inline">\(\epsilon\)</span>
</caption>
<thead><tr class="header">
<th align="left"></th>
<th align="left"><strong>aa</strong></th>
<th align="left"><strong>Aa</strong></th>
<th align="left"><strong>AA</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left"><strong>aa</strong></td>
<td align="left"><span class="math inline">\((1-\epsilon)^2\)</span></td>
<td align="left"><span class="math inline">\(2\epsilon(1-\epsilon)\)</span></td>
<td align="left"><span class="math inline">\(\epsilon^2\)</span></td>
</tr>
<tr class="even">
<td align="left"><strong>Aa</strong></td>
<td align="left"><span class="math inline">\(\epsilon\)</span></td>
<td align="left"><span class="math inline">\(1-2\epsilon\)</span></td>
<td align="left"><span class="math inline">\(\epsilon\)</span></td>
</tr>
<tr class="odd">
<td align="left"><strong>AA</strong></td>
<td align="left"><span class="math inline">\(\epsilon^2\)</span></td>
<td align="left"><span class="math inline">\(2\epsilon(1-\epsilon)\)</span></td>
<td align="left"><span class="math inline">\((1-\epsilon)^2\)</span></td>
</tr>
</tbody>
</table>
<p>This can be done via <code><a href="../reference/sequoia.html">sequoia()</a></code> argument <code>Err</code>
or <code>ErrFlavour</code>, as described next.</p>
</div>
<div class="section level2">
<h2 id="changing-the-error-model">Changing the error model<a class="anchor" aria-label="anchor" href="#changing-the-error-model"></a>
</h2>
<div class="section level3">
<h3 id="via-err">via <code>Err</code><a class="anchor" aria-label="anchor" href="#via-err"></a>
</h3>
<p>You can pass a 3x3 matrix to <code>Err</code>, for which the only
requirement is that rows must sum to 1. The homozygote for the reference
allele may be more likely to be mis-typed than the homozygote for the
alternative allele, the heterozygote may be mis-typed with different
probabilities for the two homozygotes, or whatever else is appropriate
for your dataset.</p>
<p>For example, you have the following counts from genotyping some
samples several times:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">14</span>, <span class="fl">3</span>, <span class="fl">5</span>,</span>
<span>                     <span class="fl">4</span>, <span class="fl">28</span>, <span class="fl">9</span>,</span>
<span>                     <span class="fl">0</span>, <span class="fl">4</span>, <span class="fl">72</span><span class="op">)</span>,</span>
<span>                   nrow<span class="op">=</span><span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ErrCount</span></span></code></pre></div>
<pre><code><span><span class="co">##      [,1] [,2] [,3]</span></span>
<span><span class="co">## [1,]   14    3    5</span></span>
<span><span class="co">## [2,]    4   28    9</span></span>
<span><span class="co">## [3,]    0    4   72</span></span></code></pre>
<p>A problem here is the cell with <span class="math inline">\(0\)</span>: by leaving this as is, it would imply
that true <strong>A</strong> can never be observed as
<strong>aa</strong>, even if we would genotype thousands of samples.
That seems unlikely, so as a quick fix let’s pretend that if we had
genotyped twice as many, we would have found 1 error:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrCount</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span></code></pre></div>
<p>Then divide by rowsums, e.g. using <code><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep()</a></code>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErrProbs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html" class="external-link">sweep</a></span><span class="op">(</span><span class="va">ErrCount</span>, MARGIN<span class="op">=</span><span class="fl">1</span>, STATS<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">ErrCount</span><span class="op">)</span>, FUN<span class="op">=</span><span class="st">"/"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">ErrProbs</span>, <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       [,1]  [,2]  [,3]</span></span>
<span><span class="co">## [1,] 0.636 0.136 0.227</span></span>
<span><span class="co">## [2,] 0.098 0.683 0.220</span></span>
<span><span class="co">## [3,] 0.007 0.052 0.941</span></span></code></pre>
<p>and use this as input for the <code><a href="../reference/sequoia.html">sequoia()</a></code> analysis:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoData</span>, <span class="va">LifeHistData</span>,</span>
<span>                  Err <span class="op">=</span> <span class="va">ErrProbs</span><span class="op">)</span></span></code></pre></div>
<p>In reality you probably want something a bit more sophisticated,
e.g. relying on the known error pattern for your genotyping
platform.</p>
</div>
<div class="section level3">
<h3 id="via-errflavour">via <code>ErrFlavour</code><a class="anchor" aria-label="anchor" href="#via-errflavour"></a>
</h3>
<p>If you want to run <code><a href="../reference/sequoia.html">sequoia()</a></code> with several presumed error
rates, you would need to re-calculate the 3x3 error matrix several
times. You can do this yourself, or you can provide argument
<code>ErrFlavour</code> with a function that turns a single number into
a 3x3 matrix.</p>
<p>For example, to re-define the error matrix in terms of a per-allele
error rate:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ErFunc_allele</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">Ea</span><span class="op">)</span> <span class="op">{</span></span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">Ea</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">*</span><span class="va">Ea</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">Ea</span><span class="op">)</span>, <span class="op">(</span><span class="va">Ea</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>,</span>
<span>              <span class="va">Ea</span>, <span class="fl">1</span><span class="op">-</span><span class="fl">2</span><span class="op">*</span><span class="va">Ea</span>, <span class="va">Ea</span>,</span>
<span>              <span class="op">(</span><span class="va">Ea</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">*</span><span class="va">Ea</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">Ea</span><span class="op">)</span>, <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">Ea</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>,</span>
<span>             <span class="fl">3</span>,<span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">ErFunc_allele</span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        [,1]   [,2]   [,3]</span></span>
<span><span class="co">## [1,] 0.9801 0.0198 0.0001</span></span>
<span><span class="co">## [2,] 0.0100 0.9800 0.0100</span></span>
<span><span class="co">## [3,] 0.0001 0.0198 0.9801</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ErFunc_allele</span><span class="op">(</span><span class="fl">0.005</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          [,1]    [,2]     [,3]</span></span>
<span><span class="co">## [1,] 0.990025 0.00995 0.000025</span></span>
<span><span class="co">## [2,] 0.005000 0.99000 0.005000</span></span>
<span><span class="co">## [3,] 0.000025 0.00995 0.990025</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoData</span>, <span class="va">LifeHistData</span>,</span>
<span>                  Err <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>                  ErrFlavour <span class="op">=</span> <span class="va">ErFunc_allele</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="variation-between-snps">Variation between SNPs<a class="anchor" aria-label="anchor" href="#variation-between-snps"></a>
</h2>
<p>Currently a constant error rate across all SNPs is assumed. It is
unclear whether setting this presumed rate to the arithmetic mean rate
(the regular average) across SNPs gives the best performance for real
world data. Anecdotal evidence suggests that using a somewhat higher
presumed error rate increases performance (more correct assignments
and/or fewer incorrect assignments), but this has not been thoroughly
explored. It may partly be due to confusion between the per-locus and
per-allele error rate, but also due to the very skewed nature of most
distributions (many alleles with low error rate, and some with high
error rate), resulting in a arithmetic mean that is lower than the
median. For rates, the harmonic mean may be more appropriate.</p>
</div>
<div class="section level2">
<h2 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h2>
<p>The calculation of likelihoods incorporating genotyping errors is
described in <span class="citation">Huisman (<a href="#ref-huisman17" role="doc-biblioref">2017</a>)</span>, and can be explored using
<code>CalcPairLL</code>.</p>
<p>Noteworthy is that the probabilistic estimate of an individual’s true
genotype depends on its own observed genotype and its parents’ estimated
true genotypes, but <em>not on the genotypes of its offspring</em>.
Doing so would potentially create an unstable seesaw situation, where
assignment of a parent is both affected by and does affect that parent’s
estimated true genotype.</p>
<p>For dummy individuals, which do not have an observed genotype, the
offspring genotypes do necessarily affect its estimated true genotype.
To prevent the algorithm from seesawing marginal probabilities are
calculated, excluding grandparents and/or one, two or all offspring from
the calculation, as appropriate for a particular relationship likelihood
(most importantly avoiding implicitly conditioning on oneself).</p>
<p>The main reason those marginal probabilities are not implemented for
genotyped individuals, is because of their relatively large
computational cost, with negligible effect on the pedigree outcome for
moderately high quality SNP data (error rate &lt; 1% and low
missingness).<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;If you are interested in a change in this, e.g. as an
optional feature, please do contact me to discuss possibilities. I am
fairly sure it would improve performance for low quality SNP data.&lt;/p&gt;"><sup>1</sup></a></p>
</div>
<div class="section level2">
<h2 id="try-out-different-values">Try out different values<a class="anchor" aria-label="anchor" href="#try-out-different-values"></a>
</h2>
<p>Due to both the variation in error rate between SNPs and how
genotyping errors are implemented, it is usually a good idea to run
<code><a href="../reference/sequoia.html">sequoia()</a></code> with a few different presumed genotyping error
rates. This allows you to make sure the results are robust against
slight mis-specifications of this parameter, and/or to optimise the
false negative and false positive assignment rates.</p>
<div class="section level3">
<h3 id="err-0">
<code>Err = 0</code><a class="anchor" aria-label="anchor" href="#err-0"></a>
</h3>
<p><code><a href="../reference/sequoia.html">sequoia()</a></code> is not guaranteed to work with an assumed
error rate of <span class="math inline">\(0\)</span>, especially for
more complex pedigrees. The problem is that a new assignment considers
only the direct vicinity of the putative parent-offspring pair, not the
full pedigree. An assignment will have cascading effects on the
probabilities of the true, underlying genotypes of dummy or partially
genotyped individuals downstream in the pedigree, which may result in
impossibilities when <code>Err=0</code> but genotyping errors are
present.</p>
<hr>
</div>
<div class="section level3 unnumbered">
<h3 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-huisman17" class="csl-entry">
Huisman, Jisca. 2017. <span>“Pedigree Reconstruction from
<span>SNP</span> Data: Parentage Assignment, Sibship Clustering and
Beyond.”</span> <em>Molecular Ecology Resources</em> 17 (5): 1009–24.
</div>
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
