<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>How to: RADseq data • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="How to: RADseq data">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>How to: RADseq data</h1>
                        <h4 data-toc-skip class="author">Jisca Huisman (
jisca.huisman @ gmail.com )</h4>
            
            <h4 data-toc-skip class="date">16 August 2025</h4>
      

      <div class="d-none name"><code>howto_RADseq.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="background-caveats">Background &amp; Caveats<a class="anchor" aria-label="anchor" href="#background-caveats"></a>
</h2>
<p><code>Sequoia</code> was initially developed for SNP array data,
which after typical quality control has a low rate of missing data
(&lt;1%) and a very low genotyping error rate (&lt;0.1%). This is quite
different from RADseq and similar sequence data, which typically have a
high rate of missingness (often &gt;10%) and a high rate of genotyping
errors (&gt;5%, see e.g. <span class="citation">(<a href="#ref-bresadola20">Bresadola et al. 2020</a>)</span>).</p>
<p>Because of the difference in data accuracy and completeness, the
optimal strategy for pedigree reconstruction differs. With SNP array
data it is often very obvious if and how a sample pair are related,
which allows the use of the fast hill-climbing algorithm implemented in
<code>sequoia</code>, that relies on (almost) all assignments made in
previous steps being correct. In contrast, for RADseq data the
relationship between samples is often a lot less obvious. In such
situations, algorithms that fairly randomly explore the huge number of
possible pedigree configurations are more powerful. However, this is
very time consuming, and (at least some years ago) only implemented for
parentage assignment (<a href="https://CRAN.R-project.org/package=MasterBayes" class="external-link uri">https://CRAN.R-project.org/package=MasterBayes</a> ;
archived), for sibship clustering and parentage assignment within a
single cohort (<a href="https://www.zsl.org/about-zsl/resources/software/colony" class="external-link uri">https://www.zsl.org/about-zsl/resources/software/colony</a>),
or for idealised datasets (various theoretical papers; assuming no
inbreeding and/or no missing data and/or discrete generations, etc).</p>
<p>Substantial power can be gained by making use of information from
neighbouring SNPs, rather than treating each SNP as independent as
<code>sequoia</code>, <code>Colony</code>, and most other programs do.
However, this added complexity is likely to come with a substantial
increase in computation time.</p>
<p><em>Do please let me know if you a good general pedigree
reconstruction program for RADseq data, so that I can recommend it to
users for which <code>sequoia</code> does not work well!</em></p>
<p>So, while <code>sequoia</code> is thus not an ideal tool to use with
RADseq data, it is often sufficient, and may sometimes be the only
suitable tool available.</p>
<p>Keeping that in mind, there are various tips &amp; tricks to get the
most out of your data, listed below.</p>
</div>
<div class="section level2">
<h2 id="data-filtering">Data filtering<a class="anchor" aria-label="anchor" href="#data-filtering"></a>
</h2>
<p>This step is crucial; simply using all genetic data as input for
<code>sequoia</code> is almost guaranteed to give you very poor
results.</p>
<p>Filter data based on:</p>
<ul>
<li><p><strong>Autosomal SNPs only</strong>. Remove any SNPs that are
heterozygous in the heterogametic sex (females in mammals, males in
birds and some reptiles), but always homozygous in the other
sex.</p></li>
<li><p><strong>Minor Allele Frequency (MAF)</strong>. SNPs with very low
MAF are not informative, as almost all individuals will be identical
homozygotes. Moreover, with typical RADseq error rates, any
heterozygotes or rare-allele-homozygotes are almost as likely to be
incorrect as true carriers of the rare allele. Therefore, inclusion of
these SNPs may obfuscate the true pedigree signal, rather than
strengthen it.</p></li>
<li><p><strong>Missingness</strong>. If a SNP is only scored for say 20%
of individuals, this means that only
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mn>0.2</mn><mn>2</mn></msup><mo>=</mo><mn>0.04</mn></mrow><annotation encoding="application/x-tex">0.2^2=0.04</annotation></semantics></math>
or 4% of sample pairs are both scored, so for 96% of pairs the SNP
carries no information at all.</p></li>
<li><p><strong>Linkage Disequilibrium (LD)</strong>. By chance, pairs of
individuals will be more related in some regions of their genomes than
you’d expect from the pedigree, and less related in other regions. As an
extreme example, if you have 50 SNPs in very close LD, and 30 other SNPs
scattered throughout the genome, the signal from those 50 SNPs can
overwhelm the signal from the rest of the genome, and bias the inference
of how the samples are related. It is possible to filter SNPs based on
LD even if the genetic map of the species is unknown.</p></li>
<li><p><strong>Genotyping error rate</strong>. If you have some estimate
of the genotyping error rate, for example from samples genotyped twice,
consider removing SNPs with the very worst error rates. There is no need
to filter extremely harshly: even if all SNPs had exactly the same
underlying rate, you would see a Poisson distribution of the error count
per SNP with a long tail to the right.</p></li>
<li><p><strong>Deviation from HWE</strong> Strong deviations from
genotype proportions under Hardy-Weinberg Equilibrium suggest null
alleles or other genotyping issues. Again no need to filter extremely
harshly, as variation in this metric is expected, especially in small
inbred populations and/or due to selection.</p></li>
</ul>
<p>There is no one-size-fits-all recommendation on thresholds to use, or
the relative emphasis on the different filtering criteria – it depends
on the number of SNPs you start with, the data quality, the genome size
and chromosome number, … . It can be useful to first remove the very
worst SNPs, then remove the very worst samples, and only then do the
final filtering.</p>
<p>Aim for at least 100-200 SNPs for parentage assignment, and more for
sibship clustering (say 400-500). Generally, the more complicated the
pedigree, in terms of inbreeding, overlapping generations, unknown birth
years, etc., the more SNPs you will need. But, genome size and
chromosome number put an upper limit on the number of semi-independent
SNPs. Consequently, <code>sequoia</code> is unlikely to work well on
inbred populations of species with a small genome.</p>
</div>
<div class="section level2">
<h2 id="genotyping-error-pattern">Genotyping error pattern<a class="anchor" aria-label="anchor" href="#genotyping-error-pattern"></a>
</h2>
<div class="section level3">
<h3 id="snp-array-vs-radseq">SNP array vs RADseq<a class="anchor" aria-label="anchor" href="#snp-array-vs-radseq"></a>
</h3>
<p>By default, <code>sequoia</code> assumes that the chance that an
allele is observed incorrectly is independent from whether the true
genotype is homozygote or heterozygote. This is a reasonable assumption
for SNP array data, where most SNPs that are strongly affected by
allelic dropout are removed during quality control. Remaining errors
largely arise from the clustering software used to translate the signal
strength for each of the two alleles into three genotype classes. For
each genotype, the probability to be correctly observed is
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>1</mn><mo>−</mo><mi>E</mi></mrow><annotation encoding="application/x-tex">1-E</annotation></semantics></math>,
with for the heterozygote an equal chance to be either homozygote (each
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>/</mi><mn>2</mn></mrow><annotation encoding="application/x-tex">E/2</annotation></semantics></math>),
and for the homozygotes a much smaller chance to be observed as the
other homozygote
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup><annotation encoding="application/x-tex">(E/2)^2</annotation></semantics></math>)
than as heterozygote
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mo>−</mo><msup><mrow><mo stretchy="true" form="prefix">(</mo><mi>E</mi><mi>/</mi><mn>2</mn><mo stretchy="true" form="postfix">)</mo></mrow><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">E - (E/2)^2</annotation></semantics></math>).</p>
<p>In contrast, for RADseq the chance that a true heterozygote is
observed as homozygote is considerably higher than the reverse, due to
much higher allelic dropout rates than for SNP arrays. In <span class="citation">(<a href="#ref-bresadola20">Bresadola et al.
2020</a>)</span>, the per-allele genotyping error rate at homozygous
sites
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϵ</mi><mn>0</mn></msub><annotation encoding="application/x-tex">\epsilon_0</annotation></semantics></math>
is mentioned as being typically &lt;1%, while the rate at heterozygous
sites
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϵ</mi><mn>1</mn></msub><annotation encoding="application/x-tex">\epsilon_1</annotation></semantics></math>
is around 5-10%, and can be much higher.</p>
</div>
<div class="section level3">
<h3 id="implementation">Implementation<a class="anchor" aria-label="anchor" href="#implementation"></a>
</h3>
<p>All relevant functions in the package allow full flexibility with
respect to the genotyping error pattern by letting you specify a vector
with three genotyping error rates:</p>
<ul>
<li>
<strong>hom|hom</strong>: the probability that a true homozygote is
observed as the other homozygote;</li>
<li>
<strong>het|hom</strong>: the probability that a true homozygote is
observed as heterozygote;</li>
<li>
<strong>hom|het</strong>: the probability that a true heterozygote
is observed as the minor homozygote, which is equal to the probability
it is observed as the major homozygote.</li>
</ul>
<p>So, the only assumption made is that the two homozygotes are
interchangeable with respect to genotyping error chances.</p>
<p>There are two functions to help with creating these vectors,
<code><a href="../reference/ErrToM.html">ErrToM()</a></code> (introduced in version 2.0 in 2020) and
<code><a href="../reference/Err_RADseq.html">Err_RADseq()</a></code> (introduced in version 3.0 in 2025). Both can
return both a length-3 vector and a 3x3 matrix; both of which can be
used as input to other functions. See the <code><a href="../reference/ErrToM.html">?ErrToM</a></code> helpfile
for further details.</p>
</div>
</div>
<div class="section level2">
<h2 id="code-examples">Code Examples<a class="anchor" aria-label="anchor" href="#code-examples"></a>
</h2>
<p>With a per-allele genotyping error rate at homozygous sites of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>E</mi><mn>0</mn></msub><mo>=</mo><mn>0.005</mn></mrow><annotation encoding="application/x-tex">E_0=0.005</annotation></semantics></math>
and at heterozygous sites of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>E</mi><mn>1</mn></msub><mo>=</mo><mn>0.10</mn></mrow><annotation encoding="application/x-tex">E_1=0.10</annotation></semantics></math>,
the error matrix would be</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/Err_RADseq.html">Err_RADseq</a></span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.005</span>, E1<span class="op">=</span><span class="fl">0.10</span>, Return<span class="op">=</span><span class="st">'matrix'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act        0       1        2</span></span>
<span><span class="co">##   0 0.990025 0.00995 0.000025</span></span>
<span><span class="co">##   1 0.090000 0.82000 0.090000</span></span>
<span><span class="co">##   2 0.000025 0.00995 0.990025</span></span></code></pre>
<p>where 0, 1 and 2 are the number of copies of the reference (minor)
allele.</p>
<p>For comparison, the default SNP-array based pattern with the same
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>h</mi><mi>o</mi><mi>m</mi><mo stretchy="false" form="prefix">|</mo><mi>h</mi><mi>e</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mn>0.09</mn></mrow><annotation encoding="application/x-tex">P(hom|het)=0.09</annotation></semantics></math>
would be</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span><span class="fl">0.09</span><span class="op">*</span><span class="fl">2</span>, Return<span class="op">=</span><span class="st">'matrix'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    obs</span></span>
<span><span class="co">## act      0      1      2</span></span>
<span><span class="co">##   0 0.8200 0.1719 0.0081</span></span>
<span><span class="co">##   1 0.0900 0.8200 0.0900</span></span>
<span><span class="co">##   2 0.0081 0.1719 0.8200</span></span></code></pre>
<p>and thus has a MUCH higher
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>h</mi><mi>e</mi><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>h</mi><mi>o</mi><mi>m</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(het|hom)</annotation></semantics></math>.
Basically, with the RADseq pattern, any observed heterozygote is
probably correct
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>.82</mn><mi>/</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>.82</mn><mo>+</mo><mn>2</mn><mo>*</mo><mn>.00995</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mn>0.976</mn></mrow><annotation encoding="application/x-tex">.82/(.82+2*.00995)= 0.976</annotation></semantics></math>),
but with the default pattern
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>h</mi><mi>e</mi><mi>t</mi><mo stretchy="false" form="prefix">|</mo><mi>h</mi><mi>o</mi><mi>m</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(het|hom)</annotation></semantics></math>
is always about twice as large as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>h</mi><mi>o</mi><mi>m</mi><mo stretchy="false" form="prefix">|</mo><mi>h</mi><mi>e</mi><mi>t</mi><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">P(hom|het)</annotation></semantics></math>,
and any heterozygote is quite likely to be wrong
(<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>.82</mn><mi>/</mi><mrow><mo stretchy="true" form="prefix">(</mo><mn>.82</mn><mo>+</mo><mn>2</mn><mo>*</mo><mn>.1719</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>=</mo><mn>0.705</mn></mrow><annotation encoding="application/x-tex">.82/(.82+2*.1719)= 0.705</annotation></semantics></math>).</p>
<p>You can try running <code><a href="../reference/sequoia.html">sequoia()</a></code> (or any other function)
with several different error vectors, to explore the effect it has on
assignments. Hopefully, the majority of assignments is not affected by
the assumed genotyping error rate or any other input parameter, forming
a solid core of your pedigree.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Err_RAD_low</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Err_RADseq.html">Err_RADseq</a></span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.002</span>, E1<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">Err_RAD_high</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Err_RADseq.html">Err_RADseq</a></span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.01</span>, E1<span class="op">=</span><span class="fl">0.15</span><span class="op">)</span></span>
<span><span class="va">Err_chip</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span><span class="fl">0.15</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fl">0.15</span><span class="op">)</span><span class="op">*</span><span class="fl">2</span>, Return<span class="op">=</span><span class="st">'vector'</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT_lowErr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoM</span>, <span class="va">LHdata</span>, Err<span class="op">=</span><span class="va">Err_RAD_low</span><span class="op">)</span></span>
<span><span class="va">SeqOUT_highErr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoM</span>, <span class="va">LHdata</span>, Err<span class="op">=</span><span class="va">Err_RAD_high</span><span class="op">)</span></span>
<span><span class="va">SeqOUT_chipErr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoM</span>, <span class="va">LHdata</span>, Err<span class="op">=</span><span class="va">Err_chip</span><span class="op">)</span></span></code></pre></div>
<p>You can also use <code><a href="../reference/EstConf.html">EstConf()</a></code> to explore potential
consequences of the actual (i.e., simulated) genotyping error rate being
much higher/lower than assumed (during pedigree reconstruction), or of
assuming the wrong pattern:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EC_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstConf.html">EstConf</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, <span class="va">LH_griffin</span>, </span>
<span>                args.sim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SnpError<span class="op">=</span><span class="va">Err_RAD_high</span>, nSnp<span class="op">=</span><span class="fl">200</span>, CallRate<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span>,</span>
<span>                args.seq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Err<span class="op">=</span><span class="va">Err_RAD_low</span><span class="op">)</span>,</span>
<span>                nSim <span class="op">=</span> <span class="fl">10</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EC_B</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstConf.html">EstConf</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, <span class="va">LH_griffin</span>, </span>
<span>                args.sim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SnpError<span class="op">=</span><span class="va">Err_RAD_high</span>, nSnp<span class="op">=</span><span class="fl">200</span>, CallRate<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span>,</span>
<span>                args.seq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Err<span class="op">=</span><span class="va">Err_chip</span><span class="op">)</span>,</span>
<span>                nSim <span class="op">=</span> <span class="fl">10</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EC_C</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/EstConf.html">EstConf</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, <span class="va">LH_griffin</span>, </span>
<span>                args.sim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SnpError<span class="op">=</span><span class="va">Err_RAD_high</span>, nSnp<span class="op">=</span><span class="fl">200</span>, CallRate<span class="op">=</span><span class="fl">0.8</span><span class="op">)</span>,</span>
<span>                args.seq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Err<span class="op">=</span><span class="va">Err_RAD_high</span><span class="op">)</span>,</span>
<span>                nSim <span class="op">=</span> <span class="fl">10</span>, nCores <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Calculating the number of correct and incorrect assignments, averaged
across replicates, and summed over dams and sires:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count_ARER</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">EC</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>AR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">EC</span><span class="op">$</span><span class="va">PedComp.fwd</span><span class="op">[</span>,<span class="st">'TT'</span>,<span class="st">'Match'</span>,<span class="op">]</span>,<span class="fl">1</span>,<span class="va">sum</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       ER <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">EC</span><span class="op">$</span><span class="va">PedComp.fwd</span><span class="op">[</span>,<span class="st">'TT'</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Mismatch'</span>,<span class="st">'P2only'</span><span class="op">)</span>,<span class="op">]</span>, <span class="fl">1</span>, <span class="va">sum</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">count_ARER</span><span class="op">(</span><span class="va">EC_A</span><span class="op">)</span>  <span class="co"># RAD pattern, presumed rate lower than simulated</span></span>
<span><span class="co"># $AR</span></span>
<span><span class="co"># [1] 71.3</span></span>
<span><span class="co"># $ER</span></span>
<span><span class="co"># [1] 27.2</span></span>
<span></span>
<span><span class="fu">count_ARER</span><span class="op">(</span><span class="va">EC_B</span><span class="op">)</span>  <span class="co"># SNP chip pattern, similar/higher rate presumed than sim</span></span>
<span><span class="co"># $AR</span></span>
<span><span class="co"># [1] 250.8</span></span>
<span><span class="co"># $ER</span></span>
<span><span class="co"># [1] 47.8</span></span>
<span></span>
<span><span class="fu">count_ARER</span><span class="op">(</span><span class="va">EC_C</span><span class="op">)</span>  <span class="co"># presumed rate and pattern identical to simulated</span></span>
<span><span class="co"># $AR</span></span>
<span><span class="co"># [1] 266.3</span></span>
<span><span class="co"># $ER</span></span>
<span><span class="co"># [1] 8.2</span></span></code></pre></div>
<p>Of course, in reality you likely will not know the actual genotyping
error rates very accurately (if at all), but trying out different values
in <code><a href="../reference/sequoia.html">sequoia()</a></code>, <code><a href="../reference/EstConf.html">EstConf()</a></code>,
<code><a href="../reference/GetMaybeRel.html">GetMaybeRel()</a></code>, <code><a href="../reference/CalcOHLLR.html">CalcOHLLR()</a></code> and/or
<code><a href="../reference/CalcPairLL.html">CalcPairLL()</a></code> will get you a feeling of what might and might
not be possible with your dataset.</p>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="literature">Literature<a class="anchor" aria-label="anchor" href="#literature"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-bresadola20" class="csl-entry">
Bresadola, L, V Link, C A Buerkle, C Lexer, and D Wegmann. 2020.
<span>“Estimating and Accounting for Genotyping Errors in RAD-Seq
Experiments.”</span> <em>Molecular Ecology Resources</em> 20 (4):
856–70.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
