<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Mendelian Inheritance • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Mendelian Inheritance">
<meta property="og:description" content="Mendelian inheritance probability, counting opposing homozygous (OH) loci, and calculating threshold OH value for parentage assignment. 
">
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href=".././index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_main/book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/performance_stats.html">Stats</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="mendelian_inheritance_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Mendelian Inheritance</h1>
            
      
      
      <div class="hidden name"><code>mendelian_inheritance.Rmd</code></div>

    </div>

    
    
<hr>
<div id="general" class="section level2">
<h2 class="hasAnchor">
<a href="#general" class="anchor"></a>General</h2>
<p>Genetic pedigree reconstruction relies critically on the pattern of Mendelian inheritance: the probability that an offspring inherits a specific genotype at a locus, given the genotypes of its parents. For a locus with two alleles (only biallelic SNPs are allowed in <code>sequoia</code>), the inheritance probabilities take the values of <span class="math inline">\(0\)</span>, <span class="math inline">\(1/4\)</span>, <span class="math inline">\(1/2\)</span>, or <span class="math inline">\(1\)</span>.</p>
<p>They can be depicted in many different ways, which is mostly a matter of personal preference, but I tend to think of them as this kind of table:</p>
<p>Table 1: Possible offspring genotypes conditional on two parental genotypes</p>
<p><img src="figs/Mendelian_inheritance_table.png" width="70%"></p>
<p>which shows the possible offspring genotypes that may result from each possible combination of parental genotypes. For example, if the mother is <strong>Aa</strong> and father <strong>AA</strong>, there is a 50% chance the offspring is <strong>AA</strong>, and a 50% chance the offspring is <strong>Aa</strong>.</p>
<p>In this table the allele inherited from the mother and the one inherited from the father are considered interchangeable, so that <strong>Aa</strong> = <strong>aA</strong>.</p>
<p>Note also that these are the <em>true</em> genotypes, which may differ from the <em>observed</em> genotypes due to genotyping errors.<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a></p>
<div id="single-parent" class="section level3">
<h3 class="hasAnchor">
<a href="#single-parent" class="anchor"></a>Single parent</h3>
<p>A similar table can also be constructed conditional on one instead of two parental genotypes. The other allele is then assumed to be drawn from the population allele pool at random, which contains the <strong>A</strong> allele at frequency <span class="math inline">\(q\)</span> and the <strong>a</strong> allele at frequency <span class="math inline">\(1-q\)</span>. This is equivalent to the assumption that the genotype of the other parent is a random draw from the genotypes in the population under Hardy-Weinberg equilibrium (genotype frequencies of <span class="math inline">\(q^2\)</span>, <span class="math inline">\(2q(1-q)\)</span> and <span class="math inline">\((1-q)^2\)</span> for <strong>AA</strong>, <strong>Aa</strong> and <strong>aa</strong>, respectively).</p>
<p>Table 2: Possible offspring genotypes conditional on one parental genotype and the allele inherited from the other parent</p>
<p><img src="figs/Mendelian_table_1parent.png" width="40%"></p>
<p>Table 3: Probability of offspring genotype (columns) conditional on one parental genotype and population allele frequency q for the ‘A’ allele</p>
<p><img src="figs/Mendelian_prob_1parent.png" width="60%"></p>
</div>
</div>
<div id="parentage-assignment-based-on-mendelian-incompatibilities" class="section level2">
<h2 class="hasAnchor">
<a href="#parentage-assignment-based-on-mendelian-incompatibilities" class="anchor"></a>Parentage assignment based on Mendelian incompatibilities</h2>
<div id="no-genotyping-errors" class="section level3">
<h3 class="hasAnchor">
<a href="#no-genotyping-errors" class="anchor"></a>No genotyping errors</h3>
<p>For an individual with <strong>AA</strong> at a locus, the true parent will not be <strong>aa</strong> at that locus (bottom-left in Table 3), and thus all <strong>aa</strong> individuals can be discarded as potential parents. If for example the <strong>A</strong> allele has a frequency of <span class="math inline">\(q=0.4\)</span>, this will discard about 36% of individuals (<span class="math inline">\((1-0.4)^2=0.36\)</span>), from just considering a single SNP.</p>
<p>With an increasing number of SNPs considered, an increasing proportion of the population can be discarded as potential parent based on this opposing homozygosity test alone.</p>
</div>
<div id="with-genotyping-errors" class="section level3">
<h3 class="hasAnchor">
<a href="#with-genotyping-errors" class="anchor"></a>With genotyping errors</h3>
<p>In real data, there is a small chance that an individual that was scored as <strong>AA</strong> truly has a parent that was scored as <strong>aa</strong>, as either one might be mis-genotyped and actually be heterozygous<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> (see also the background info on <a href="genotyping_errors.html">genotyping errors</a>. To allow for this possibility, a non-zero maximum threshold is set for the number of loci at which an individual and candidate parent may be opposing homozygotes (OH), and the number of loci at which individual and candidate parent-pair may have Mendelian errors. The latter includes besides OH also cases were the offspring is heterozygous <strong>Aa</strong> and both parents are identical homozygotes (both <strong>AA</strong> or both <strong>aa</strong>; see Table 1).</p>
<p>If genotyping error is low and the number of SNPs is large (high hundreds/thousands), counting Mendelian errors is usually sufficient to reduce the number of candidate parents to a single pair or individual. It is widely used as a reliable method of parentage assignment, since with an abundance of high-quality SNP data the distinction between parent and not-parent is typically very clear (<span class="citation"><a href="#ref-hayes11" role="doc-biblioref">Hayes</a> (<a href="#ref-hayes11" role="doc-biblioref">2011</a>)</span>, <span class="citation"><a href="#ref-calus11" role="doc-biblioref">Calus, Mulder, and Bastiaansen</a> (<a href="#ref-calus11" role="doc-biblioref">2011</a>)</span>, <span class="citation"><a href="#ref-strucken15" role="doc-biblioref">Strucken et al.</a> (<a href="#ref-strucken15" role="doc-biblioref">2016</a>)</span>).</p>
</div>
<div id="limitations" class="section level3">
<h3 class="hasAnchor">
<a href="#limitations" class="anchor"></a>Limitations</h3>
<p>Even with high quality data, close relatives other than the true parents will also be opposing homozygotes at a very low number of SNPs. This is especially true for full siblings of the focal individual and full siblings of the true parent. In some settings it is straight forward to exclude full siblings as potential parents (e.g. in animal husbandry or with discrete generations), but full aunts and uncles can rarely be excluded <em>a priori</em>.</p>
<p>Moreover, many datasets have too few SNPs and/or a too high genotyping error rate for this method to work reliably – see the <a href="#sec:Example">example</a></p>
</div>
</div>
<div id="expected-number-of-mendelian-errors" class="section level2">
<h2 class="hasAnchor">
<a href="#expected-number-of-mendelian-errors" class="anchor"></a>Expected number of Mendelian errors</h2>
<p>The theoretical distribution of the number of Mendelian errors for true parent-offspring pairs and true parent-parent-offspring trios can be calculated from the number of SNPs, their genotyping error rate, and the allele frequencies. This is implemented in the R package as function <code><a href="../reference/CalcMaxMismatch.html">CalcMaxMismatch()</a></code>.</p>
<p>For parent-offspring pairs the calculations and R code are worked out here in detail, the calculations for parent-parent-offspring trios are similar.</p>
<div id="mismatch-probability" class="section level3">
<h3 class="hasAnchor">
<a href="#mismatch-probability" class="anchor"></a>Mismatch probability</h3>
<p>The probability that an offspring is observed as <strong>aa</strong> and its parent as <strong>AA</strong>, if the other parent is unknown, is a function of:</p>
<ol style="list-style-type: decimal">
<li>the probability that the parent’s true genotype is <span class="math inline">\(g_{par}\)</span>, given the population allele frequency <span class="math inline">\(q\)</span> and assuming Hardy-Weinberg equilibrium;</li>
<li>the probability that the parent is observed as <strong>AA</strong> if its true genotype is <span class="math inline">\(g_{par}\)</span>; this depends on the genotyping error rate <span class="math inline">\(E\)</span> and <a href="genotyping_errors.html">error structure</a>;</li>
<li>the probability that the offspring inherits true genotype <span class="math inline">\(g_O\)</span>, given its parents true genotype <span class="math inline">\(g_{par}\)</span> and the other parent’s genotype is random;</li>
<li>the probability that the offspring is observed as <strong>aa</strong> if its true genotype is <span class="math inline">\(g_{off}\)</span>;</li>
</ol>
<p>Each of these probabilities is fairly straight forward, as is multiplying them and summing across all <span class="math inline">\(3\times3\)</span> possible true genotypes for the parent and the offspring. For example, for the case where <span class="math inline">\(g_{par}=Aa\)</span> and <span class="math inline">\(g_{off}=aa\)</span>:</p>
<ol style="list-style-type: decimal">
<li><p><span class="math inline">\(P(g_{par}=Aa) = 2q(1-q)\)</span></p></li>
<li><p><span class="math inline">\(P(obs\ AA | act\ Aa) = E/2\)</span> (for the default error matrix)</p></li>
<li><p><span class="math inline">\(P_{inherit} = (1-q)/2\)</span> (Table 3)</p></li>
<li><p><span class="math inline">\(P(obs\ aa | act\ aa) = (1-E/2)^2\)</span></p></li>
</ol>
<p>We could write out and simplify the expression summing over all nine genotype combinations, or we could let R do the hard work:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example values</span>
<span class="va">q</span> <span class="op">&lt;-</span> <span class="fl">0.4</span>   <span class="co"># allele frequency</span>
<span class="va">E</span> <span class="op">&lt;-</span> <span class="fl">0.01</span>  <span class="co"># genotyping error rate</span>

<span class="co"># 1. probabilities of actual genotypes under HWE (0, 1 or 2 copies of reference allele)</span>
<span class="va">AHWE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">q</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">*</span><span class="va">q</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">q</span><span class="op">)</span>, <span class="va">q</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>

<span class="co"># 2 + 4. probability of observed genotype (columns) conditional on actual genotype (rows) </span>
<span class="va">ErrM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ErrToM.html">ErrToM</a></span><span class="op">(</span>Err <span class="op">=</span> <span class="va">E</span>, flavour<span class="op">=</span><span class="st">"version2.0"</span>, Return<span class="op">=</span><span class="st">"matrix"</span><span class="op">)</span>
<span class="va">ErrM</span></code></pre></div>
<pre><code>##       obs-0|act obs-1|act obs-2|act
## act-0  0.990025   0.00995  0.000025
## act-1  0.005000   0.99000  0.005000
## act-2  0.000025   0.00995  0.990025</code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># 3. probability that offspring actual genotype is 0/1/2 (rows) given </span>
<span class="co">#    parent actual genotype is 0/1/2 (columns). </span>
<span class="co"># (Transposed relative to Table 3)</span>
<span class="va">AKAP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">q</span>, <span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">q</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">0</span>,
                 <span class="va">q</span>,   <span class="fl">1</span><span class="op">/</span><span class="fl">2</span>,     <span class="fl">1</span><span class="op">-</span><span class="va">q</span>,
                 <span class="fl">0</span>,   <span class="va">q</span><span class="op">/</span><span class="fl">2</span>,     <span class="va">q</span><span class="op">)</span>, 
                 nrow<span class="op">=</span><span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>
                 
<span class="co"># joined probability actual genotypes: (1. x 3.)</span>
<span class="va">P.AA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sweep.html">sweep</a></span><span class="op">(</span><span class="va">AKAP</span>, <span class="fl">2</span>, <span class="va">AHWE</span>, <span class="st">"*"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span><span class="op">(</span><span class="va">P.AA</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">0</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>
<span class="va">P.AA</span></code></pre></div>
<pre><code>##       0     1     2
## 0 0.216 0.144 0.000
## 1 0.144 0.240 0.096
## 2 0.000 0.096 0.064</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># joined probability observed genotypes: (2. x (1. x 3.) x 4.)</span>
<span class="va">P.OO</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">ErrM</span><span class="op">)</span> <span class="op">%*%</span> <span class="va">P.AA</span> <span class="op">%*%</span> <span class="va">ErrM</span>
<span class="fu"><a href="https://rdrr.io/r/base/dimnames.html">dimnames</a></span><span class="op">(</span><span class="va">P.OO</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">2</span>, <span class="fl">0</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">P.OO</span>, <span class="fl">4</span><span class="op">)</span></code></pre></div>
<pre><code>##        0      1      2
## 0 0.2131 0.1445 0.0012
## 1 0.1445 0.2400 0.0959
## 2 0.0012 0.0959 0.0637</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">P.Mismatch</span> <span class="op">&lt;-</span> <span class="va">P.OO</span><span class="op">[</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">]</span> <span class="op">+</span> <span class="va">P.OO</span><span class="op">[</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">]</span>
<span class="va">P.Mismatch</span></code></pre></div>
<pre><code>## [1] 0.00240198</code></pre>
<p>This is the probability that a single parent-offspring pair at this single locus are opposing homozygotes (OH).</p>
<p>The expected number of loci at which a parent-offspring pair are OH follows a binomial distribution. For simplicity we assume in this example that all SNPs have the same allele frequency and same genotyping error rate (this assumption is not made by <code><a href="../reference/CalcMaxMismatch.html">CalcMaxMismatch()</a></code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># using same no. SNPs as in low quality example data below</span>
<span class="va">nSnp</span> <span class="op">&lt;-</span> <span class="fl">100</span>   

<span class="fu"><a href="https://rdrr.io/r/graphics/curve.html">curve</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">dbinom</a></span><span class="op">(</span><span class="va">x</span>, size <span class="op">=</span> <span class="va">nSnp</span>, prob <span class="op">=</span> <span class="va">P.Mismatch</span><span class="op">)</span>,
      from<span class="op">=</span><span class="fl">0</span>, to<span class="op">=</span><span class="fl">10</span>, n<span class="op">=</span><span class="fl">11</span>, type<span class="op">=</span><span class="st">"b"</span>,
      xlab<span class="op">=</span><span class="st">"No. opposing homozygous loci"</span>, ylab<span class="op">=</span><span class="st">"Probability"</span><span class="op">)</span></code></pre></div>
<p><img src="mendelian_inheritance_files/figure-html/unnamed-chunk-4-1.png" width="576"></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span>OHcount <span class="op">=</span> <span class="fl">0</span><span class="op">:</span><span class="fl">5</span>,
      prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">dbinom</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">5</span>, size <span class="op">=</span> <span class="va">nSnp</span>, prob <span class="op">=</span> <span class="va">P.Mismatch</span><span class="op">)</span>,<span class="fl">4</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##      OHcount   prob
## [1,]       0 0.7862
## [2,]       1 0.1893
## [3,]       2 0.0226
## [4,]       3 0.0018
## [5,]       4 0.0001
## [6,]       5 0.0000</code></pre>
</div>
<div id="threshold" class="section level3">
<h3 class="hasAnchor">
<a href="#threshold" class="anchor"></a>Threshold</h3>
<p>During parentage assignment, a threshold OH count is used to filter candidate parents. This threshold is set such that is highly unlikely that any true parent-offspring pairs are discarded. Assignments are made based on likelihoods calculated after the filtering step, and thus any not-parents that pass the filter will not necessarily result in false positive assignments.<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a></p>
<p>The threshold must be such that the expected OH count for <em>all</em> true parent-offspring pairs will be below this value, not just for a random pair. For example, the probability that a random parent-offspring pair has a OH count of 3 or lower is</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">pbinom</a></span><span class="op">(</span><span class="fl">3</span>, size<span class="op">=</span><span class="va">nSnp</span>, prob<span class="op">=</span><span class="va">P.Mismatch</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0.9998914</code></pre>
<p>which makes 3 seem like a good value for a threshold. However, the probability that all say 3000 parent-offspring pairs are at or below this value (pretending they are all independent) is</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">pbinom</a></span><span class="op">(</span><span class="fl">3</span>, size<span class="op">=</span><span class="va">nSnp</span>, prob<span class="op">=</span><span class="va">P.Mismatch</span><span class="op">)</span><span class="op">^</span><span class="fl">3000</span></code></pre></div>
<pre><code>## [1] 0.7219437</code></pre>
<p>which is not so great.</p>
<p>To set a threshold <code>maxOH</code> that with a certain probability (say <span class="math inline">\(x=0.999\)</span>) will not exclude any of the <span class="math inline">\(Z\)</span> true parent-offspring pairs, we must solve <code>pbinom(maxOH, size=nSnp, prob=P.Mismatch)^Z</code> <span class="math inline">\(= x\)</span>:</p>
<p><span class="math display">\[
\begin{aligned}
 y^Z = x \\
 y = x^{1/Z}
\end{aligned}
\]</span> in other words, set the desired probability to <span class="math inline">\(0.999^{1/Z}\)</span>.</p>
<p>Of course we do not know the number of parent-offspring pairs <span class="math inline">\(Z\)</span> prior to doing parentage assignment, but for most datasets it is reasonable to assume that it is roughly equal to the number of individuals <span class="math inline">\(N\)</span> (or at least of the same order of magnitude).</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">N</span> <span class="op">&lt;-</span> <span class="fl">3000</span>
<span class="fu">stats</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">qbinom</a></span><span class="op">(</span><span class="fl">0.999</span><span class="op">^</span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="va">N</span><span class="op">)</span>, size <span class="op">=</span> <span class="va">nSnp</span>, prob <span class="op">=</span> <span class="va">P.Mismatch</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 5</code></pre>
<p>double check:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">pbinom</a></span><span class="op">(</span><span class="fl">5</span>, size<span class="op">=</span><span class="va">nSnp</span>, prob<span class="op">=</span><span class="va">P.Mismatch</span><span class="op">)</span><span class="op">^</span><span class="va">N</span></code></pre></div>
<pre><code>## [1] 0.9994341</code></pre>
</div>
</div>
<div id="sec:Example" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:Example" class="anchor"></a>Example</h2>
<div id="data" class="section level3">
<h3 class="hasAnchor">
<a href="#data" class="anchor"></a>Data</h3>
<p>The R package includes a griffin example pedigree with 200 individuals, of which 167 have a known dam and 163 a known sire. Here two sets of genotype data are simulated for this pedigree: one of high quality, and one of low quality.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, package<span class="op">=</span><span class="st">"sequoia"</span><span class="op">)</span>

<span class="co"># simulate genotype data</span>
<span class="va">Geno.highQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SimGeno.html">SimGeno</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, nSnp<span class="op">=</span><span class="fl">500</span>, ParMis<span class="op">=</span><span class="fl">0</span>, SnpError<span class="op">=</span><span class="fl">1e-4</span><span class="op">)</span>
<span class="va">Geno.lowQ</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SimGeno.html">SimGeno</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, nSnp<span class="op">=</span><span class="fl">100</span>, ParMis<span class="op">=</span><span class="fl">0</span>, SnpError<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span>
<span class="co"># default parameter values: 1% of SNPs missing, MAF uniform between 0.3 and 0.5.</span>
<span class="co"># ParMis=0 : no non-genotyped parents</span></code></pre></div>
</div>
<div id="distribution-of-opposing-homozygous-loci-count" class="section level3">
<h3 class="hasAnchor">
<a href="#distribution-of-opposing-homozygous-loci-count" class="anchor"></a>Distribution of opposing homozygous loci count</h3>
<p>The number of opposing homozygous loci between all possible pairs of individuals is counted for both SNP datasets:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># define function to count no. opposing homozygous loci in two vectors x and y</span>
<span class="va">CountOH</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span><span class="op">)</span>  <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="op">(</span><span class="va">x</span><span class="op">==</span><span class="fl">2</span> <span class="op">&amp;</span> <span class="va">y</span><span class="op">==</span><span class="fl">0</span><span class="op">)</span> <span class="op">|</span> <span class="op">(</span><span class="va">x</span><span class="op">==</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">y</span><span class="op">==</span><span class="fl">2</span><span class="op">)</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># apply this function all possible pairs </span>
<span class="va">N</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">Geno.highQ</span><span class="op">)</span>  <span class="co"># no. individuals</span>
<span class="va">OH.highQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">N</span>, <span class="va">N</span><span class="op">)</span>  <span class="co"># matrix for output</span>
<span class="va">OH.lowQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="va">N</span>, <span class="va">N</span><span class="op">)</span>
<span class="co"># do upper triangle of matrix only, excluding diagonal (self)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="fl">1</span>, <span class="va">N</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
  <span class="kw">for</span> <span class="op">(</span><span class="va">j</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="va">i</span><span class="op">+</span><span class="fl">1</span>, <span class="va">N</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="va">OH.highQ</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CountOH</span><span class="op">(</span><span class="va">Geno.highQ</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>, <span class="va">Geno.highQ</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">)</span>
    <span class="va">OH.lowQ</span><span class="op">[</span><span class="va">i</span>,<span class="va">j</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">CountOH</span><span class="op">(</span><span class="va">Geno.lowQ</span><span class="op">[</span><span class="va">i</span>,<span class="op">]</span>, <span class="va">Geno.lowQ</span><span class="op">[</span><span class="va">j</span>,<span class="op">]</span><span class="op">)</span>
  <span class="op">}</span>
<span class="op">}</span>

<span class="co"># plot histograms</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfcol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, mai<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.9</span>,<span class="fl">.9</span>,<span class="fl">.5</span>,<span class="fl">.6</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">OH.highQ</span>, breaks<span class="op">=</span><span class="fl">50</span>, xlab<span class="op">=</span><span class="st">"No. SNPs opposing homozygous"</span>,
     main<span class="op">=</span><span class="st">"High quality:\n 500 SNPs, error rate 0.01%"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">OH.lowQ</span>, breaks<span class="op">=</span><span class="fl">25</span>, xlab<span class="op">=</span><span class="st">"No. SNPs opposing homozygous"</span>,
     main<span class="op">=</span><span class="st">"Low quality:\n 100 SNPs, error rate 1%"</span><span class="op">)</span></code></pre></div>
<p><img src="mendelian_inheritance_files/figure-html/unnamed-chunk-10-1.png" width="100%"></p>
<p>For the high quality SNP set there is a clear separation between the first peak (parent-offspring pairs, OH=0) and the second peak (all other pairs). For the low quality SNP set, there is an area where the two peaks overlap. Zooming in shows this more clearly:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/graphics/par.html">par</a></span><span class="op">(</span>mfcol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>, mai<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">.9</span>,<span class="fl">.9</span>,<span class="fl">.5</span>,<span class="fl">.6</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">OH.highQ</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">OH.highQ</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fl">.5</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">20</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">100</span><span class="op">)</span>,
     xlab<span class="op">=</span><span class="st">"No. SNPs opposing homozygous"</span>, main<span class="op">=</span><span class="st">"500 SNPs, error rate 0.01%"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"forestgreen"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/rect.html">rect</a></span><span class="op">(</span>xleft<span class="op">=</span><span class="fl">1.5</span>, ybottom<span class="op">=</span><span class="op">-</span><span class="fl">20</span>, xright<span class="op">=</span><span class="fl">7.5</span>, ytop<span class="op">=</span><span class="fl">200</span>,
     col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/adjustcolor.html">adjustcolor</a></span><span class="op">(</span><span class="st">"forestgreen"</span>, alpha.f<span class="op">=</span><span class="fl">0.4</span><span class="op">)</span>, border<span class="op">=</span><span class="cn">NA</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="va">OH.lowQ</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">OH.lowQ</span>,na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fl">.5</span>, xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>,<span class="fl">10</span><span class="op">)</span>, ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1000</span><span class="op">)</span>, 
     xlab<span class="op">=</span><span class="st">"No. SNPs opposing homozygous"</span>, main<span class="op">=</span><span class="st">"100 SNPs, error rate 1%"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span><span class="op">(</span>v<span class="op">=</span><span class="fl">2</span>, lwd<span class="op">=</span><span class="fl">3</span>, lty<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"forestgreen"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/rect.html">rect</a></span><span class="op">(</span>xleft<span class="op">=</span><span class="fl">0.5</span>, ybottom<span class="op">=</span><span class="op">-</span><span class="fl">50</span>, xright<span class="op">=</span><span class="fl">4.5</span>, ytop<span class="op">=</span><span class="fl">1200</span>,
     col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/grDevices/adjustcolor.html">adjustcolor</a></span><span class="op">(</span><span class="st">"black"</span>, alpha.f<span class="op">=</span><span class="fl">0.4</span><span class="op">)</span>, border<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></code></pre></div>
<p><img src="mendelian_inheritance_files/figure-html/unnamed-chunk-11-1.png" width="100%"></p>
<p>For the high quality SNP set, the shaded area representing 2-7 opposing homozygous loci includes one or two pairs of individuals at most (variable between simulated datasets), so that it does not matter much were exactly the threshold is placed. For the low quality SNP set, there are hundreds of pairs with 1-4 opposing homozygous loci, so that slightly changing the threshold will considerably change the number of assignments (with an unknown proportion of false positives).</p>
</div>
<div id="assignment-with-sequoia" class="section level3">
<h3 class="hasAnchor">
<a href="#assignment-with-sequoia" class="anchor"></a>Assignment with <code>sequoia()</code>
</h3>
<p>Likelihood-based parentage assignment with <code>sequoia</code> is intended for those situations where the SNP data is not overwhelmingly strong, like the low quality genotype data in this example. It checks not only whether or not a parent-offspring genotype combination is possible or not, but also <em>how likely</em> it is.</p>
<p>For example, a mating between an <strong>aa</strong> and <strong>aa</strong> parent pair always results in <strong>aa</strong> offspring, but an <strong>aa</strong> X <strong>Aa</strong> mating only 50% of the time, and an <strong>Aa</strong> X <strong>Aa</strong> mating 25% of the time. Thus, for an <strong>aa</strong> offspring, an <strong>aa</strong> parent is more likely than an <strong>Aa</strong> parent.</p>
<p>For single parents it also considers how likely it is that the ‘other’ allele (the one not inherited from the candidate parent under consideration) is inherited from a random individual. This depends on the allele frequency (Table 3): if a putative parent shares a rare allele with an individual, this affects the likelihood strongly (the individual is unlikely to have inherited from some other individual), while if the allele they share is common it could have been inherited from anyone.<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
<p>For further details, please see <span class="citation"><a href="#ref-huisman17" role="doc-biblioref">Huisman</a> (<a href="#ref-huisman17" role="doc-biblioref">2017</a>)</span>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># parentage assignment using sequoia</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">LH_griffin</span>, package<span class="op">=</span><span class="st">"sequoia"</span><span class="op">)</span>  <span class="co"># sex + birth year information</span>

<span class="va">SeqPAR.lowQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">Geno.lowQ</span>, LifeHistData<span class="op">=</span><span class="va">LH_griffin</span>, 
                               Err<span class="op">=</span><span class="fl">0.01</span>, Module<span class="op">=</span><span class="st">"par"</span>, 
                               quiet<span class="op">=</span><span class="cn">TRUE</span>, Plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>

<span class="va">SeqPAR.lowQ</span><span class="op">$</span><span class="va">Specs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MaxMismatchOH"</span>, <span class="st">"MaxMismatchME"</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<pre><code>##       MaxMismatchOH MaxMismatchME
## Specs             4             7</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># check reconstructed pedigree</span>
<span class="va">PC.lowQ</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, <span class="va">SeqPAR.lowQ</span><span class="op">$</span><span class="va">PedigreePar</span>, Plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">PC.lowQ</span><span class="op">$</span><span class="va">Counts</span><span class="op">[</span><span class="st">"GT"</span>,,<span class="op">]</span></code></pre></div>
<pre><code>##           parent
## class      dam sire
##   Total    167  163
##   Match    165  160
##   Mismatch   0    0
##   P1only     2    3
##   P2only     0    0</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># P1only: false negative (assignment in pedigree 1 (= original) only)</span>
<span class="co"># P2only: false positive (assignment in pedigree 2 (= reconstructed) only)</span></code></pre></div>
</div>
<div id="count-mendelian-errors" class="section level3">
<h3 class="hasAnchor">
<a href="#count-mendelian-errors" class="anchor"></a>Count Mendelian errors</h3>
<p>We can count OH for all true parent-offspring pairs, and check whether the non-assigned parents were particular outliers:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Ped.true.OH</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CalcOHLLR.html">CalcOHLLR</a></span><span class="op">(</span>Pedigree <span class="op">=</span> <span class="va">Ped_griffin</span>,
                                  GenoM <span class="op">=</span> <span class="va">Geno.lowQ</span>,
                                  CalcLLR <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## Counting Mendelian errors ...</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># distribution of OH &amp; trio Mendelian errors:</span>
<span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">Ped.true.OH</span>, Panels<span class="op">=</span><span class="st">"OH"</span><span class="op">)</span></code></pre></div>
<p><img src="mendelian_inheritance_files/figure-html/unnamed-chunk-13-1.png" width="576"></p>
<p>The reason some parents remained unassigned was not because they did not pass this filtering step, but because of the subsequent likelihood calculations. Probably the likelihood to be parent and offspring is very similar to the likelihood to be full siblings, and thus no assignment was made. These likelihoods can be calculated and retrieved with function <code><a href="../reference/CalcPairLL.html">CalcPairLL()</a></code>. For further details, see <a href="vignette_main/book/key-points.html#sec:lowAR">here in the vignette</a> and <a href="./pairLL_classification.html">more info on pairwise likelihoods</a>.</p>
</div>
<div id="notes" class="section level3">
<h3 class="hasAnchor">
<a href="#notes" class="anchor"></a>Notes</h3>
<p>Please note that also <code>sequoia</code> has its limitations, and that parentage assignment with real-world data with 100 SNPs and an estimated average genotyping error rate of 1% probably performs worse, due to higher missingness, lower allele frequencies, LD between SNPs, non-random distribution of genotyping errors, etc.</p>
<p>Moreover, please note that sibship clustering and grandparent assignment require more and/or higher quality SNPs than parentage assignment.</p>
<hr>
</div>
<div id="references" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-calus11" class="csl-entry">
Calus, M P L, H A Mulder, and J W M Bastiaansen. 2011. <span>“Identification of <span>M</span>endelian Inconsistencies Between <span>SNP</span> and Pedigree Information of Sibs.”</span> <em>Genet. Sel. Evol</em> 43: 34.
</div>
<div id="ref-hayes11" class="csl-entry">
Hayes, B J. 2011. <span>“<span><em>Technical Note:</em></span> Efficient Parentage Assignment and Pedigree Reconstruction with Dense Single Nucleotide Polymorphism Data.”</span> <em>Journal of Dairy Science</em> 94 (4): 2114–17.
</div>
<div id="ref-huisman17" class="csl-entry">
Huisman, Jisca. 2017. <span>“Pedigree Reconstruction from <span>SNP</span> Data: Parentage Assignment, Sibship Clustering and Beyond.”</span> <em>Molecular Ecology Resources</em> 17 (5): 1009–24.
</div>
<div id="ref-strucken15" class="csl-entry">
Strucken, E M, S H Lee, H K Lee, K D Song, J P Gibson, and C Gondro. 2016. <span>“How Many Markers Are Enough? <span>F</span>actors Influencing Parentage Testing in Different Livestock Populations.”</span> <em>Journal of Animal Breeding and Genetics</em> 133 (1): 13–23.
</div>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>mutations can also cause deviations from this expected pattern, but are for pedigree reconstruction usually grouped together with genotyping errors; over the time spans of a typical pedigree they will be very rare.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>the chance that both are mistyped, or that they are actually identical homozygotes, is usually much much smaller but rarely zero.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>but letting lots of pairs through the filtering step will increase computational time.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>SNPs with low allele frequencies are not very useful in general for pedigree reconstruction though, because almost all individuals will be homozygous for the common allele.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
