<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Using mitochondrial haplotypes • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using mitochondrial haplotypes">
<meta name="description" content="Use of option `mtSame`
">
<meta property="og:description" content="Use of option `mtSame`
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="mtDNA_files/kePrint-0.0.1/kePrint.js"></script><link href="mtDNA_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Using mitochondrial haplotypes</h1>
            
      

      <div class="d-none name"><code>mtDNA.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="why-use-mitochondrial-data">Why use mitochondrial data<a class="anchor" aria-label="anchor" href="#why-use-mitochondrial-data"></a>
</h2>
<p>Using data on the sharing of mitochondrial (mt) haplotypes makes it
possible to distinguish between maternal/matrilineal and
paternal/patrilineal relatives. For example, half-siblings sharing the
same mother will always have the same mitochondrial haplotype, while
half-siblings sharing the same father may have different haplotypes.
Similarly, an individual always has the same mitochondrial haplotype as
its maternal grandmother, but not (necessarily) as its paternal
grandmother.</p>
<div class="figure">
<img src="figs/pedigree_mtDNA_inheritance.png" alt="Example pedigree with colour coded mitochondrial haplotype inheritance" width="70%"><p class="caption">
Example pedigree with colour coded mitochondrial haplotype inheritance
</p>
</div>
</div>
<div class="section level2">
<h2 id="limitations">Limitations<a class="anchor" aria-label="anchor" href="#limitations"></a>
</h2>
<ul>
<li>Some paternal relatives may share the same mitochondrial haplotype,
especially in small populations or if the population was founded by a
limited number of females;</li>
<li>Mitochondrial haplotype sharing does not help distinguish between
maternal half-siblings, full avuncular, and grand-parental pairs, and so
when generations overlap the assignment rate may still be below
100%;</li>
<li>It can be challenging to set a threshold between ‘same’ versus
‘different’ haplotypes, especially with a non-negligible genotyping
error rate, limited number of polymorphic mitochondrial markers, and/or
low call rate. The most conservative approach as input for
<code>sequoia</code> is to set any uncertain pairs to ‘same’ (see next
section).</li>
</ul>
</div>
<div class="section level2">
<h2 id="input">Input<a class="anchor" aria-label="anchor" href="#input"></a>
</h2>
<p>From package version 2.7 onwards, function <code><a href="../reference/sequoia.html">sequoia()</a></code> has
the input parameter <code>mtSame</code>. This is a matrix with for each
pair of individuals indicating whether they share the same mitochondrial
haplotype (1) or have different haplotypes (0).</p>
<p>The algorithm focusses on the differences: if two individuals
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>i</mi><annotation encoding="application/x-tex">i</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>j</mi><annotation encoding="application/x-tex">j</annotation></semantics></math>
are close relatives according to the autosomal SNPs, and have different
mitochondrial haplotypes (<code>mtsame[i,j]==0</code>), they cannot be
maternal/matrilineal relatives, and thus must be paternal/patrilineal
relatives. In contrast, if two individuals share the same mitochondrial
haplotype (<code>mtsame[i,j]==1</code>), they may be maternal relatives,
but it cannot be excluded that they are paternal relatives.</p>
<p>Therefore, if it is unclear or unknown whether two individuals share
the same mitochondrial haplotype, this should be coded as <code>1</code>
or <code>NA</code>, which is interpreted as ‘may or may not be the
same’, while <code>0</code> is interpreted as ‘definitely different’ (=
definitely not maternal relatives).</p>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead>
<tr>
<th style="empty-cells: hide;border-bottom:hidden;" colspan="1">
</th>
<th style="empty-cells: hide;border-bottom:hidden;" colspan="1">
</th>
<th style="border-bottom:hidden;padding-bottom:0; padding-left:3px;padding-right:3px;text-align: center; " colspan="3">
<div style="border-bottom: 1px solid #ddd; padding-bottom: 5px; ">
Related
</div>
</th>
</tr>
<tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
maternal
</th>
<th style="text-align:left;">
paternal
</th>
<th style="text-align:left;">
unrelated
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;font-weight: bold;">
</td>
<td style="text-align:left;font-weight: bold;">
same
</td>
<td style="text-align:left;color: black !important;">
OK
</td>
<td style="text-align:left;">
OK
</td>
<td style="text-align:left;">
OK
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;">
mt haplotype
</td>
<td style="text-align:left;font-weight: bold;">
different
</td>
<td style="text-align:left;color: red !important;">
X
</td>
<td style="text-align:left;">
OK
</td>
<td style="text-align:left;">
OK
</td>
</tr>
<tr>
<td style="text-align:left;font-weight: bold;">
</td>
<td style="text-align:left;font-weight: bold;">
?
</td>
<td style="text-align:left;color: black !important;">
OK
</td>
<td style="text-align:left;">
OK
</td>
<td style="text-align:left;">
OK
</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h2>
<p>This example uses a single cohort from pedigree
<code>Ped_HSg5</code>, which is included in the package. Both females
and males have multiple mates, resulting in both maternal and paternal
half-siblings as well as full siblings. We pretend here that none of the
parents are genotyped, and that each breeding female has a unique
mitochondrial haplotype.</p>
<p>First, let us create a dataset for this example:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jiscah.github.io/">sequoia</a></span><span class="op">)</span></span>
<span><span class="co"># pedigree subset</span></span>
<span><span class="va">id_subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="va">LH_HSg5</span><span class="op">[</span><span class="va">LH_HSg5</span><span class="op">$</span><span class="va">BirthYear</span> <span class="op">==</span> <span class="fl">2001</span>, <span class="st">'ID'</span><span class="op">]</span> <span class="op">)</span></span>
<span><span class="va">Ped_sub</span> <span class="op">&lt;-</span> <span class="va">Ped_HSg5</span><span class="op">[</span><span class="va">Ped_HSg5</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">id_subset</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># assign a mitochondrial haplotype code to each founding female</span></span>
<span><span class="va">founder_females</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span> <span class="va">LH_HSg5</span><span class="op">[</span><span class="va">LH_HSg5</span><span class="op">$</span><span class="va">BirthYear</span> <span class="op">==</span> <span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">LH_HSg5</span><span class="op">$</span><span class="va">Sex</span><span class="op">==</span><span class="fl">1</span>, <span class="st">'ID'</span><span class="op">]</span> <span class="op">)</span></span>
<span><span class="va">mtHaps_founders</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/setNames.html" class="external-link">setNames</a></span><span class="op">(</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">founder_females</span><span class="op">)</span><span class="op">]</span>, <span class="va">founder_females</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">mtHaps_founders</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## a00001 a00002 a00003 a00004 a00005 a00006 </span></span>
<span><span class="co">##    "A"    "B"    "C"    "D"    "E"    "F"</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># infer mitochondrial haplotypes for all matrilineal descendants of individuals </span></span>
<span><span class="co"># with known haplotypes</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">'https://raw.githubusercontent.com/JiscaH/sequoiaExtra/main/infer_mt_haplo.R'</span><span class="op">)</span> </span>
<span><span class="va">mtHaps_all</span> <span class="op">&lt;-</span> <span class="fu">infer_mt_haplotype</span><span class="op">(</span><span class="va">Ped_sub</span>, <span class="va">mtHaps_founders</span><span class="op">)</span></span>
<span><span class="va">mtHaps_all</span><span class="op">[</span><span class="fl">41</span><span class="op">:</span><span class="fl">50</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## a01036 a01039 a01042 a01043 a01046 a01047 a01048 a01050 a01051 a01053 </span></span>
<span><span class="co">##    "F"    "D"    "R"    "R"    "F"    "F"    "F"    "E"    "E"    "H"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add mitochondrial haplotype code to pedigree</span></span>
<span><span class="va">Ped_sub</span><span class="op">$</span><span class="va">mt</span> <span class="op">&lt;-</span> <span class="va">mtHaps_all</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">Ped_sub</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Ped_sub</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        id    dam   sire mt</span></span>
<span><span class="co">## 41 a01001 a00014 b00011  N</span></span>
<span><span class="co">## 42 a01002 a00014 b00011  N</span></span>
<span><span class="co">## 43 b01003 a00014 b00011  N</span></span>
<span><span class="co">## 44 b01004 a00014 b00011  N</span></span>
<span><span class="co">## 45 a01005 a00013 b00001  M</span></span>
<span><span class="co">## 46 b01006 a00013 b00001  M</span></span></code></pre>
<p>The function <code>infer_mt_haplotype()</code> is not included in the
package, but available from <a href="github.com/JiscaH/sequoiaExtra"></a>. It can also be used on real
data, when mother-offspring relationships are reliably known, e.g. from
behavioural observations, and only a subset of individuals is genotyped
for mitochondrial markers. It infers haplotypes for both ancestors and
descendants of individuals with known haplotypes, and then iterates a
few times to infer haplotypes for all matrilineal relatives.</p>
<p>Then, create the matrix with for each pair of individuals whether
they have the same (1) or different (0) haplotype:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mt_same</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/outer.html" class="external-link">outer</a></span><span class="op">(</span><span class="va">mtHaps_all</span>, <span class="va">mtHaps_all</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">x</span><span class="op">==</span><span class="va">y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mt_same</span><span class="op">[</span><span class="fl">41</span><span class="op">:</span><span class="fl">48</span>, <span class="fl">41</span><span class="op">:</span><span class="fl">48</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##        a01036 a01039 a01042 a01043 a01046 a01047 a01048 a01050</span></span>
<span><span class="co">## a01036      1      0      0      0      1      1      1      0</span></span>
<span><span class="co">## a01039      0      1      0      0      0      0      0      0</span></span>
<span><span class="co">## a01042      0      0      1      1      0      0      0      0</span></span>
<span><span class="co">## a01043      0      0      1      1      0      0      0      0</span></span>
<span><span class="co">## a01046      1      0      0      0      1      1      1      0</span></span>
<span><span class="co">## a01047      1      0      0      0      1      1      1      0</span></span>
<span><span class="co">## a01048      1      0      0      0      1      1      1      0</span></span>
<span><span class="co">## a01050      0      0      0      0      0      0      0      1</span></span></code></pre>
<p>Simulate autosomal SNP data, with none of the parents genotyped:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Geno_A</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SimGeno.html">SimGeno</a></span><span class="op">(</span><span class="va">Ped_sub</span>, nSnp<span class="op">=</span><span class="fl">500</span>, ParMis<span class="op">=</span><span class="fl">1</span>, SnpError<span class="op">=</span><span class="fl">1e-3</span><span class="op">)</span></span></code></pre></div>
<p>Run sequoia with &amp; without <code>mtSame</code> (note that these
can take a couple minutes to run):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT_regular</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">Geno_A</span>, <span class="va">LH_HSg5</span>, quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">SeqOUT_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">Geno_A</span>, <span class="va">LH_HSg5</span>, mtSame <span class="op">=</span> <span class="va">mt_same</span>, quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Count number of half-siblings assigned in each run:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RelM_regular</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetRelM.html">GetRelM</a></span><span class="op">(</span><span class="va">SeqOUT_regular</span><span class="op">$</span><span class="va">Pedigree</span>, patmat<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">RelM_regular</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    FS     M     O     P     S     U </span></span>
<span><span class="co">##   620   188   376   188   284 79000</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">RelM_mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetRelM.html">GetRelM</a></span><span class="op">(</span><span class="va">SeqOUT_mt</span><span class="op">$</span><span class="va">Pedigree</span>, patmat<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">RelM_mt</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##    FS     M   MHS     O     P   PHS     S     U </span></span>
<span><span class="co">##   640   192   704   384   192  1472   232 50008</span></span></code></pre>
<p>Thus, without <code>mtSame</code> not a single half-sibling
relationship is assigned, as it cannot be determined whether they are
maternal or paternal half-siblings. In contrast, with
<code>mtSame</code> hundreds of half-siblings are assigned.</p>
<p>In this example, the effect of <code>mtSame</code> on the number of
individuals to which a dummy parent is assigned is negligible, as every
individual has several full-siblings, and a dummy mother and father are
assigned to each full-sibling cluster. Some additional full siblings are
assigned: with default option <code>Complex = 'full'</code> the
distinction between full siblings versus e.g. ‘half siblings as well as
cousins’ (r=3/8) is not always clear, and the assignment of additional
maternal and/or paternal half-siblings may clarify the full sibling
relationships.</p>
<p>For a graphical representation of these results you can run:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">RelM_regular</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">RelM_mt</span><span class="op">)</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
