<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quick start example 2: Real data • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quick start example 2: Real data">
<meta property="og:description" content="How to prepare data for sequoia.
">
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_main/_book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/_book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/_performance_stats.html">Stats</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="quick_start_example2_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quick start example 2: Real data</h1>
            
      
      
      <div class="hidden name"><code>quick_start_example2.Rmd</code></div>

    </div>

    
    
<hr>
<p>First get a subset of SNPs which are as informative (high MAF, high call rate), reliable (low error rate), and independent (low LD) as possible. Ideally around 400 – 700 for full pedigree reconstruction; fewer are necessary for only parentage assignment, while more may be necessary with high levels of polygamy or inbreeding.</p>
<p>One possible tool to perform such subsetting of SNPs is <a href="https://www.cog-genomics.org/plink2">PLINK</a> using something like this (on the command line), with thresholds adapted to your particular dataset.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> --file mydata --geno 0.1 --maf 0.3 --indep 50 5 2</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">plink</span> --file mydata --extract plink.prune.in --recodeA --out inputfile_for_sequoia</span></code></pre></div>
<p>In addition you need a dataframe with the sex and birth/hatching year of as many individuals as possible, in arbitrary order.</p>
<p>Then, in R start with something similar to this, adapting the code to match your file names and guestimated genotyping error rate.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># read in genotype data if already coded as 0/1/2, with missing=-9:</span>
<span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.csv</a></span><span class="op">(</span><span class="st">"mydata.csv"</span>, header<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/CheckGeno.html">CheckGeno</a></span><span class="op">(</span><span class="va">Geno</span><span class="op">)</span>
<span class="co"># read in many other input formats (not .vcf (yet)):</span>
<span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenoConvert.html">GenoConvert</a></span><span class="op">(</span>InFile <span class="op">=</span> <span class="st">"mydata.ped"</span>, InFormat<span class="op">=</span><span class="st">"ped"</span><span class="op">)</span>
<span class="co">#</span>
<span class="co"># read in lifehistory data: ID-Sex-birthyear, column names ignored</span>
<span class="co"># optional: minimum &amp; maximum birth year, when not exactly known</span>
<span class="va">LH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"LifeHistoryData.txt"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="co">#</span>
<span class="co"># duplicate check &amp; parentage assignment (takes few minutes)</span>
<span class="co"># (maximum number of sibship-clustering iterations = 0)</span>
<span class="va">ParOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,
                  Module<span class="op">=</span><span class="st">"par"</span>, Err<span class="op">=</span><span class="fl">0.005</span>,
                  quiet <span class="op">=</span> <span class="cn">FALSE</span>, Plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#</span>
<span class="co"># inspect duplicates (intentional or accidental)</span>
<span class="va">ParOUT</span><span class="op">$</span><span class="va">DupGenotype</span>
<span class="co">#</span>
<span class="co"># compare assigned parents to field pedigree (check column order!)</span>
<span class="va">FieldPed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">"FieldPed.txt"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span>
<span class="va">PC.par</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">FieldPed</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"dam"</span>, <span class="st">"sire"</span><span class="op">)</span><span class="op">]</span>,
                     Ped2 <span class="op">=</span> <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span>
<span class="va">PC.par</span><span class="op">$</span><span class="va">Counts</span><span class="op">[</span><span class="st">"TT"</span>,,<span class="op">]</span>
<span class="co">#</span>
<span class="co"># calculate Mendelian errors per SNP (works also w field pedigree)</span>
<span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SnpStats.html">SnpStats</a></span><span class="op">(</span><span class="va">Geno</span>, <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span>
<span class="va">MAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">0.5</span>, <span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span>, <span class="fl">1</span><span class="op">-</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span><span class="op">)</span>
<span class="co">#</span>
<span class="co"># ..........................................................</span>
<span class="co"># polish dataset: remove one indiv. from each duplicate pair</span>
<span class="co"># &amp; drop low call rate samples</span>
<span class="co"># &amp; drop SNPs with high error rate and/or low MAF</span>
<span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Geno</span><span class="op">)</span>,<span class="op">]</span> 
<span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno2</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"Err.hat"</span><span class="op">]</span><span class="op">&gt;</span><span class="fl">0.05</span> <span class="op">|</span> <span class="va">MAF</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">]</span>
<span class="co">#</span>
<span class="va">Indiv.Mis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html">apply</a></span><span class="op">(</span><span class="va">Geno2</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="op">-</span><span class="fl">9</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">Geno2</span><span class="op">)</span>
<span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno2</span><span class="op">[</span><span class="va">Indiv.Mis</span> <span class="op">&lt;</span> <span class="fl">0.2</span>, <span class="op">]</span>
<span class="co"># check histograms for sensible thresholds, iterate if necessary</span>
<span class="co">#</span>
<span class="co"># run full pedigree reconstruction (may take up to a few hours)</span>
<span class="co"># including re-run of parentage assignment</span>
<span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno2</span>,
                  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,
                  Module <span class="op">=</span> <span class="st">"ped"</span>,
                  Err <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span>
<span class="co">#</span>
<span class="co"># inspect assigned parents, proportion dummy parents, etc.</span>
<span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">)</span>
<span class="co"># (see Example 1 for saving results)</span></code></pre></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
