<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quick start example 2: Real data • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quick start example 2: Real data">
<meta name="description" content="How to prepare data for sequoia.
">
<meta property="og:description" content="How to prepare data for sequoia.
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help">
<li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quick start example 2: Real data</h1>
            
      

      <div class="d-none name"><code>quick_start_example2.Rmd</code></div>
    </div>

    
    
<hr>
<p>First get a subset of SNPs which are as informative (high MAF, high
call rate), reliable (low error rate), and independent (low LD) as
possible. Ideally around 400 – 700 for full pedigree reconstruction;
fewer are necessary for only parentage assignment, while more may be
necessary with high levels of polygamy or inbreeding.</p>
<p>One possible tool to perform such subsetting of SNPs is <a href="https://www.cog-genomics.org/plink2" class="external-link">PLINK</a> using something
like this (on the command line), with thresholds adapted to your
particular dataset.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--file</span> mydata <span class="at">--geno</span> 0.1 <span class="at">--maf</span> 0.3 <span class="at">--indep</span> 50 5 2</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--file</span> mydata <span class="at">--extract</span> plink.prune.in <span class="at">--recodeA</span> <span class="at">--out</span> inputfile_for_sequoia</span></code></pre></div>
<p>In addition you need a dataframe with the sex and birth/hatching year
of as many individuals as possible, in arbitrary order.</p>
<p>Then, in R start with something similar to this, adapting the code to
match your file names and guestimated genotyping error rate.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read in genotype data if already coded as 0/1/2, with missing=-9:</span></span>
<span><span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="st">"mydata.csv"</span>, header<span class="op">=</span><span class="cn">FALSE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/CheckGeno.html">CheckGeno</a></span><span class="op">(</span><span class="va">Geno</span><span class="op">)</span></span>
<span><span class="co"># read in many other input formats (not .vcf (yet)):</span></span>
<span><span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GenoConvert.html">GenoConvert</a></span><span class="op">(</span>InFile <span class="op">=</span> <span class="st">"mydata.ped"</span>, InFormat<span class="op">=</span><span class="st">"ped"</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># read in lifehistory data: ID-Sex-birthyear, column names ignored</span></span>
<span><span class="co"># optional: minimum &amp; maximum birth year, when not exactly known</span></span>
<span><span class="va">LH</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"LifeHistoryData.txt"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># duplicate check &amp; parentage assignment (takes few minutes)</span></span>
<span><span class="co"># (maximum number of sibship-clustering iterations = 0)</span></span>
<span><span class="va">ParOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,</span>
<span>                  Module<span class="op">=</span><span class="st">"par"</span>, Err<span class="op">=</span><span class="fl">0.005</span>,</span>
<span>                  quiet <span class="op">=</span> <span class="cn">FALSE</span>, Plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># inspect duplicates (intentional or accidental)</span></span>
<span><span class="va">ParOUT</span><span class="op">$</span><span class="va">DupGenotype</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># compare assigned parents to field pedigree (check column order!)</span></span>
<span><span class="va">FieldPed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"FieldPed.txt"</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">PC.par</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">FieldPed</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"dam"</span>, <span class="st">"sire"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                     Ped2 <span class="op">=</span> <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></span>
<span><span class="va">PC.par</span><span class="op">$</span><span class="va">Counts</span><span class="op">[</span><span class="st">"TT"</span>,,<span class="op">]</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># calculate Mendelian errors per SNP (works also w field pedigree)</span></span>
<span><span class="va">stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SnpStats.html">SnpStats</a></span><span class="op">(</span><span class="va">Geno</span>, <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></span>
<span><span class="va">MAF</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">0.5</span>, <span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span>, <span class="fl">1</span><span class="op">-</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"AF"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># ..........................................................</span></span>
<span><span class="co"># polish dataset: remove one indiv. from each duplicate pair</span></span>
<span><span class="co"># &amp; drop low call rate samples</span></span>
<span><span class="co"># &amp; drop SNPs with high error rate and/or low MAF</span></span>
<span><span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">Geno</span><span class="op">)</span>,<span class="op">]</span> </span>
<span><span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno2</span><span class="op">[</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">stats</span><span class="op">[</span>,<span class="st">"Err.hat"</span><span class="op">]</span><span class="op">&gt;</span><span class="fl">0.05</span> <span class="op">|</span> <span class="va">MAF</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">#</span></span>
<span><span class="va">Indiv.Mis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">Geno2</span>, <span class="fl">1</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span> <span class="op">==</span> <span class="op">-</span><span class="fl">9</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Geno2</span><span class="op">)</span></span>
<span><span class="va">Geno2</span> <span class="op">&lt;-</span> <span class="va">Geno2</span><span class="op">[</span><span class="va">Indiv.Mis</span> <span class="op">&lt;</span> <span class="fl">0.2</span>, <span class="op">]</span></span>
<span><span class="co"># check histograms for sensible thresholds, iterate if necessary</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># run full pedigree reconstruction (may take up to a few hours)</span></span>
<span><span class="co"># including re-run of parentage assignment</span></span>
<span><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno2</span>,</span>
<span>                  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,</span>
<span>                  Module <span class="op">=</span> <span class="st">"ped"</span>,</span>
<span>                  Err <span class="op">=</span> <span class="fl">0.001</span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># inspect assigned parents, proportion dummy parents, etc.</span></span>
<span><span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">)</span></span>
<span><span class="co"># (see Example 1 for saving results)</span></span></code></pre></div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
