<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Install sequoia, simulate data, run parentage assignment and full pedigree reconstruction.
">
<title>Quick start example 1: Simulated data • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quick start example 1: Simulated data">
<meta property="og:description" content="Install sequoia, simulate data, run parentage assignment and full pedigree reconstruction.
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.11.4.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href=".././index.html">
    <span class="fas fa fas fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/sequoia.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a>
    <a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    <a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a>
    <a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a>
    <a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a>
    <a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-help">Help</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-help">
    <a class="dropdown-item" href="../articles/help.html">Need help?</a>
    <a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/pedigree_consultancy.html">Consultancy</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa fas fa-envelope"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JiscaH/sequoia">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca">
    <span class="fab fa fab fa-mastodon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Quick start example 1: Simulated data</h1>
                        <h4 data-toc-skip class="author">J Huisman</h4>
            
            <h4 data-toc-skip class="date">15 Oct 2020</h4>
      
      
      <div class="d-none name"><code>quick_start_example1.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start-example-1-simulated-data">Quick start example 1: Simulated data<a class="anchor" aria-label="anchor" href="#quick-start-example-1-simulated-data"></a>
</h2>
<div class="section level3">
<h3 id="get-started">Get started<a class="anchor" aria-label="anchor" href="#get-started"></a>
</h3>
<p><em>For a succinct version of this example where you can copy/paste
all code in one go, see the main vignette</em></p>
<p>In this example genotypes are simulated from a fictional pedigree,
which together with the associated life history data is provided with
the package. This pedigree consists of 5 discrete generations with
interconnected half-sib clusters (Pedigree II in the paper<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;Huisman, Jisca. “Pedigree reconstruction from SNP data:
parentage assignment, sibship clustering and beyond.” Molecular ecology
resources 17.5 (2017): 1009-1024.&lt;/p&gt;"><sup>1</sup></a>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install the package. This is only the required the first time, or if you wish to update</span></span>
<span><span class="co"># install.packages("sequoia")   </span></span>
<span></span>
<span><span class="co"># Load the package. This is required at the start of every new R session.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jiscah.github.io/">sequoia</a></span><span class="op">)</span> </span>
<span></span>
<span><span class="co"># Load the example pedigree and life history data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">Ped_HSg5</span>, <span class="va">LH_HSg5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look at the data structure</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">Ped_HSg5</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">LH_HSg5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or, in Rstudio, view the full dataframe:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">Ped_HSg5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="simulate-genotypes">Simulate genotypes<a class="anchor" aria-label="anchor" href="#simulate-genotypes"></a>
</h3>
<p>Simulate some genotype data to use for this try-out.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SimGeno.html">SimGeno</a></span><span class="op">(</span>Ped <span class="op">=</span> <span class="va">Ped_HSg5</span>, nSnp <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p>Function <code><a href="../reference/SimGeno.html">SimGeno()</a></code> can simulate data with a specific
genotyping error rate, call rate and proportion of non-genotyped
parents, see its helpfile (<code><a href="../reference/SimGeno.html">?SimGeno</a></code>) for details. Here we
use the defaults; 40% of parents are presumed non-genotyped (their
genotypes are removed from the dataset before <code><a href="../reference/SimGeno.html">SimGeno()</a></code>
returns its result).</p>
<p>Alternatively, use the simulated genotype data included in the
package to get identical results:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Geno</span> <span class="op">&lt;-</span> <span class="va">Geno_HSg5</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parentage-assignment">Parentage assignment<a class="anchor" aria-label="anchor" href="#parentage-assignment"></a>
</h3>
<p>Run sequoia with the genotype data, lifehistory data, and otherwise
default values. It is often advisable to first only run parentage
assignment, and check if the results are sensible and/or if any
parameters need adjusting. Full pedigree reconstruction, including
sibship clustering etc., is much more time consuming.</p>
<p>To only run a data check, duplicate check (always performed first)
and parentage assignment, specify <code>Module = 'par'</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ParOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,</span>
<span>                  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,</span>
<span>                  Module <span class="op">=</span> <span class="st">'par'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Checking input data ...</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #BBBB00;">!</span> There are 2 SNPs scored for &lt;50% of individuals</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> There are  920  individuals and  200  SNPs.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## ── Among genotyped individuals: ___</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> There are 467 females, 453 males, 0 of unknown sex, and 0 hermaphrodites.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Exact birth years are from 2000 to 2005</span></span></code></pre>
<pre><code><span><span class="co">## ___</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Calling `MakeAgePrior()` ...</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Ageprior: Flat 0/1, overlapping generations, MaxAgeParent = 6,6</span></span></code></pre>
<p><img src="quick_start_example1_files/figure-html/example1-e-1.png" width="576"></p>
<pre><code><span><span class="co">## </span></span>
<span><span class="co"><span style="color: #00BB00;">## ~~~ Duplicate check ~~~</span></span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> No potential duplicates found</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co"><span style="color: #00BB00;">## ~~~ Parentage assignment ~~~</span></span></span></code></pre>
<pre><code><span><span class="co">##     Time |  R |       Step |   progress |  dams | sires |   Total LL </span></span>
<span><span class="co">## -------- | -- | ---------- | ---------- | ----- | ----- | ----------</span></span>
<span><span class="co">## 20:40:29 |  0 | initial    |            |     0 |     0 |   -70880.7 </span></span>
<span><span class="co">## 20:40:29 |  0 | parents    |            |   514 |   524 |   -57593.7 </span></span>
<span><span class="co">## 20:40:34 | 99 | est byears |            | </span></span>
<span><span class="co">## 20:40:34 | 99 | calc LLR   |            |</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BB00;">✔</span> <span style="color: #00BB00;">assigned 514 dams and 524 sires to 920 individuals </span></span></span>
<span><span class="co"><span style="color: #00BB00;">## </span></span></span></code></pre>
<p><img src="quick_start_example1_files/figure-html/example1-e-2.png" width="576"><img src="quick_start_example1_files/figure-html/example1-e-3.png" width="576"></p>
<!--- explain about ageprior messages & plots, and LL --->
<p>You will see several plots appearing:</p>
<ul>
<li>Histograms of the minor allele frequency and missingness, produced
by <code><a href="../reference/SnpStats.html">SnpStats()</a></code>, which is called by <code><a href="../reference/CheckGeno.html">CheckGeno()</a></code>
to make sure there are no monomorphic SNPs or SNPs with extremely high
missingness in the dataset<br>
</li>
<li>Heatmap of the ageprior before parentage assignment, showing that an
age difference of 0 years is disallowed (<span class="math inline">\(0\)</span>, black) for mothers (<em>M</em>) and
fathers (<em>P</em>), but all other age differences are allowed (<span class="math inline">\(1\)</span>, pale green) for all relationships. The
initial maximum age of parents is set to the maximum age difference
found in the lifehistory data.<br>
</li>
<li>Barplots of assignment rate (produced by <code><a href="../reference/SummarySeq.html">SummarySeq()</a></code>):
nearly 60% of individuals have a genotyped parent assigned, in line with
the 40% non-genotyped parents we simulated.</li>
</ul>
<p>In addition, you will see several messages, including about the
initial and post-parentage total LL (log10-likelihood). This is the
probability of observing the genotype data, given the (current)
pedigree; initially it is assumed that all individuals are unrelated.
This number is negative, and gets closer to zero when the pedigree
explains the data better.</p>
<p>The result is a list, with the following elements:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">ParOUT</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Specs"            "ErrM"             "args.AP"          "Snps-LowCallRate"</span></span>
<span><span class="co">##  [5] "DupLifeHistID"    "NoLH"             "AgePriors"        "LifeHist"        </span></span>
<span><span class="co">##  [9] "PedigreePar"      "TotLikPar"        "AgePriorExtra"    "LifeHistPar"</span></span></code></pre>
<p>which are explained in detail in the helpfile (<code><a href="../reference/sequoia.html">?sequoia</a></code>)
and the main vignette.</p>
<p>you find the assigned parents in list element
<code>PedigreePar</code>:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         id    dam   sire LLRdam LLRsire LLRpair OHdam OHsire MEpair</span></span>
<span><span class="co">## 915 b05187 a04045   &lt;NA&gt;   7.16      NA      NA     0     NA     NA</span></span>
<span><span class="co">## 916 a05188 a04045   &lt;NA&gt;   7.37      NA      NA     0     NA     NA</span></span>
<span><span class="co">## 917 a05189   &lt;NA&gt; b04177     NA    3.32      NA    NA      0     NA</span></span>
<span><span class="co">## 918 b05190   &lt;NA&gt; b04177     NA    3.53      NA    NA      0     NA</span></span>
<span><span class="co">## 919 b05191   &lt;NA&gt; b04177     NA    3.50      NA    NA      0     NA</span></span>
<span><span class="co">## 920 b05192   &lt;NA&gt; b04177     NA    4.34      NA    NA      0     NA</span></span></code></pre>
<p>we can compare these to the true parents, in the original pedigree
from which we simulated the genotype data:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chk_par</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">Ped_HSg5</span>, Ped2 <span class="op">=</span> <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-h-1.png" width="576"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chk_par</span><span class="op">$</span><span class="va">Counts</span><span class="op">[</span><span class="st">"GG"</span>,,<span class="op">]</span>   </span></code></pre></div>
<pre><code><span><span class="co">##           parent</span></span>
<span><span class="co">## class      dam sire</span></span>
<span><span class="co">##   Total    537  544</span></span>
<span><span class="co">##   Match    513  523</span></span>
<span><span class="co">##   Mismatch   1    1</span></span>
<span><span class="co">##   P1only    23   20</span></span>
<span><span class="co">##   P2only     0    0</span></span></code></pre>
<p>Here ‘GG’ stands for <em>G</em>enotyped offspring, <em>G</em>enotyped
parent.</p>
</div>
<div class="section level3">
<h3 id="pedigree-reconstruction">Pedigree reconstruction<a class="anchor" aria-label="anchor" href="#pedigree-reconstruction"></a>
</h3>
<p>Then, we can run full pedigree reconstruction. By default, this
re-runs the parentage assignment.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT_HSg5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>, LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>, Module <span class="op">=</span> <span class="st">"ped"</span><span class="op">)</span></span></code></pre></div>
<p>If you don’t want to re-run parentage assignment (e.g. because it
took quite a bit of time), or if you have specified several non-default
parameter values you want to use again, , you can provide the old output
as input.</p>
<p>Re-used will be:</p>
<ul>
<li>parameter settings (<code>$Specs</code>)</li>
<li>genotyping error structure (<code>$ErrM</code>)</li>
<li>Lifehistory data (<code>$LifeHist</code>)</li>
<li>Age prior, based on the age distribution of assigned parents
(<code>$AgePriors</code>)</li>
</ul>
<p>The last is generated by <code><a href="../reference/MakeAgePrior.html">MakeAgePrior()</a></code>, which has
detected that all parent-offspring pairs have an age difference of <span class="math inline">\(1\)</span>, and all siblings an age difference of
<span class="math inline">\(0\)</span>, i.e. that generations do not
overlap. This information will be used during the full pedigree
reconstruction.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ParOUT</span><span class="op">$</span><span class="va">AgePriors</span></span></code></pre></div>
<pre><code><span><span class="co">##   M P FS MS PS</span></span>
<span><span class="co">## 0 0 0  1  1  1</span></span>
<span><span class="co">## 1 1 1  0  0  0</span></span>
<span><span class="co">## 2 0 0  0  0  0</span></span></code></pre>
<p>So run <code><a href="../reference/sequoia.html">sequoia()</a></code> with the old output as input,</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT_HSg5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,</span>
<span>                  SeqList <span class="op">=</span> <span class="va">ParOUT</span>,</span>
<span>                  Module <span class="op">=</span> <span class="st">"ped"</span><span class="op">)</span></span></code></pre></div>
<p>Full pedigree reconstruction will take at least a few minutes to run
on this fairly simple dataset with one thousand individuals. It may take
a few hours on larger or more complex datasets, and/or if there is much
ambiguity about relationships due to a low number of SNPs, genotyping
errors, and missing data.</p>
<p>The output from this example pedigree is included with the
package:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">SeqOUT_HSg5</span><span class="op">)</span></span></code></pre></div>
<p>You will get several messages:</p>
<ul>
<li>From <code><a href="../reference/CheckGeno.html">CheckGeno()</a></code> to inform you that the genotype data
is OK<br>
</li>
<li>To inform you which parts of <code>SeqList</code> (the output from
the old <code><a href="../reference/sequoia.html">sequoia()</a></code> run) are being re-used<br>
</li>
<li>At the end of each iteration, to update you on the total
log10-likelihood (LL), the time it took for that iteration, and the
number of dams &amp; sires assigned to real, genotyped (non-dummy)
individuals. The total LL and number of assigned parents should plateau
(but may ‘wobble’ a bit), and then the algorithm will finish.</li>
</ul>
<p>Again you will see some plots:</p>
<ul>
<li>Histograms of the minor allele frequency and missingness<br>
</li>
<li>Heatmap of an extended version of the ageprior, including all
grandparents and avuncular (aunt/uncle) relationships
(<code>$AgePriorExtra</code>). These are derived from the ageprior for
parents and siblings, which hasn’t changed since just after parentage
assignment<br>
</li>
<li>an update of the assignment rates plot, now including dummy
parents.</li>
</ul>
<p>For additional plots and a few tables to inspect the pedigree, you
can use <code><a href="../reference/SummarySeq.html">SummarySeq()</a></code>:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">summary_seq1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">SeqOUT_HSg5</span>, Panels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sibships"</span>, <span class="st">"D.parents"</span>, <span class="st">"LLR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-k-1.png" width="576"><img src="quick_start_example1_files/figure-html/example1-k-2.png" width="576"><img src="quick_start_example1_files/figure-html/example1-k-3.png" width="576"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">summary_seq1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "PedSummary"  "ParentCount" "GPCount"     "SibSize"</span></span></code></pre>
<p>And again we can also compare the results to the true parents:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">Ped_HSg5</span>, Ped2 <span class="op">=</span> <span class="va">SeqOUT_HSg5</span><span class="op">$</span><span class="va">Pedigree</span><span class="op">)</span></span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-l-1.png" width="576"></p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chk</span><span class="op">$</span><span class="va">Counts</span></span></code></pre></div>
<pre><code><span><span class="co">## , , parent = dam</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     class</span></span>
<span><span class="co">## cat  Total Match Mismatch P1only P2only</span></span>
<span><span class="co">##   GG   537   537        0      0      0</span></span>
<span><span class="co">##   GD   357   357        0      0      0</span></span>
<span><span class="co">##   GT   894   894        0      0      0</span></span>
<span><span class="co">##   DG    39    39        0      0      0</span></span>
<span><span class="co">##   DD    27    27        0      0      0</span></span>
<span><span class="co">##   DT    66    66        0      0      0</span></span>
<span><span class="co">##   TT   960   960        0      0      0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## , , parent = sire</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     class</span></span>
<span><span class="co">## cat  Total Match Mismatch P1only P2only</span></span>
<span><span class="co">##   GG   543   543        0      0      0</span></span>
<span><span class="co">##   GD   351   351        0      0      0</span></span>
<span><span class="co">##   GT   894   894        0      0      0</span></span>
<span><span class="co">##   DG    33    33        0      0      0</span></span>
<span><span class="co">##   DD    33    33        0      0      0</span></span>
<span><span class="co">##   DT    66    66        0      0      0</span></span>
<span><span class="co">##   TT   960   960        0      0      0</span></span></code></pre>
<p>See <code><a href="../reference/PedCompare.html">?PedCompare</a></code> for a more interesting example with some
mismatches.</p>
<p>If you wish to count e.g. the number of full sibling pairs that are
assigned as full siblings, paternal half siblings, maternal
halfsiblings, or unrelated, or similar comparisons, use
<code><a href="../reference/ComparePairs.html">ComparePairs()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="save-results">Save results<a class="anchor" aria-label="anchor" href="#save-results"></a>
</h3>
<p>Lastly, you often wish to save the results to file. You can do this
as an <code>Rdata</code> file, which can contain any number of R
objects. You can later retrieve these in R using <code>load</code>, so
that you can later resume where you left of. The disadvantage is that
you cannot open the <code>Rdata</code> files outside of R (as far as I
am aware). Therefore, <code>sequoia</code> also includes a function to
write all output to a set of plain text files in a specified folder.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">SeqOUT_HSg5</span>, <span class="va">Geno</span>, file<span class="op">=</span><span class="st">"Sequoia_output_date.RData"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/writeSeq.html">writeSeq</a></span><span class="op">(</span>SeqList <span class="op">=</span> <span class="va">SeqOUT_HSg5</span>, GenoM <span class="op">=</span> <span class="va">Geno</span>, folder <span class="op">=</span> <span class="st">"Sequoia-OUT"</span><span class="op">)</span></span></code></pre></div>
<!---
## Real data
For a template which can be adapted to analyse real genotype data, see [quick-start example 2](./quick_start_example2.html) in the main vignette. 
--->
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
