<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Report generated on griffin example data.  
">
<title>Sequoia report • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sequoia report">
<meta property="og:description" content="Report generated on griffin example data.  
">
<meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo_s.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<meta name="twitter:card" content="summary_large_image">
<meta name="twitter:creator" content="@HuismanJisca">
<meta name="twitter:site" content="@HuismanJisca">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.5.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href=".././index.html">
    <span class="fas fa fas fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/sequoia.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a>
    <a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    <a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a>
    <a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a>
    <a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-help">Help</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-help">
    <a class="dropdown-item" href="../articles/help.html">Need help?</a>
    <a class="dropdown-item" href="../articles/pedigree_consultancy.html">Consultancy</a>
    <a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="nav-link" href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa fas fa-envelope"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JiscaH/sequoia">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca">
    <span class="fab fa fab fa-mastodon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/HuismanJisca">
    <span class="fab fa fab fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="sequoia_report_files/kePrint-0.0.1/kePrint.js"></script><link href="sequoia_report_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Sequoia report</h1>
            
      
      
      <div class="d-none name"><code>sequoia_report.Rmd</code></div>
    </div>

    
    
<hr>
<p><em>This report was created with the various griffin example files
included with the package. The markdown file can be found at [<a href="https://github.com/JiscaH/sequoiaExtra" class="external-link uri">https://github.com/JiscaH/sequoiaExtra</a>] .</em></p>
<hr>
<p>This pedigree was reconstructed using the R package
<code>sequoia</code>.</p>
<p>The R package <code>sequoia</code> reconstructs the pedigree in two
steps: first genotyped parents are assigned to genotyped offspring
(‘parentage assignment’), followed by clustering of siblings sharing a
non-genotyped parent and assignment of grandparents (‘full pedigree
reconstruction’).</p>
<hr>
<div class="section level2">
<h2 id="comments">Comments<a class="anchor" aria-label="anchor" href="#comments"></a>
</h2>
<p>This is a overview of the pedigree reconstruction for Griffins born
2001-2010.</p>
<hr>
</div>
<div class="section level2">
<h2 id="input-summary">Input summary<a class="anchor" aria-label="anchor" href="#input-summary"></a>
</h2>
<div class="section level3">
<h3 id="parameter-settings">Parameter settings<a class="anchor" aria-label="anchor" href="#parameter-settings"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">Specs</span><span class="op">)</span><span class="op">)</span>, booktabs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">kable_styling</span><span class="op">(</span>full_width <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
Specs
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
NumberIndivGenotyped
</td>
<td style="text-align:left;">
142
</td>
</tr>
<tr>
<td style="text-align:left;">
NumberSnps
</td>
<td style="text-align:left;">
400
</td>
</tr>
<tr>
<td style="text-align:left;">
GenotypingErrorRate
</td>
<td style="text-align:left;">
1e-04
</td>
</tr>
<tr>
<td style="text-align:left;">
MaxMismatchDUP
</td>
<td style="text-align:left;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
MaxMismatchOH
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
MaxMismatchME
</td>
<td style="text-align:left;">
2
</td>
</tr>
<tr>
<td style="text-align:left;">
Tfilter
</td>
<td style="text-align:left;">
-2
</td>
</tr>
<tr>
<td style="text-align:left;">
Tassign
</td>
<td style="text-align:left;">
0.5
</td>
</tr>
<tr>
<td style="text-align:left;">
nAgeClasses
</td>
<td style="text-align:left;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
MaxSibshipSize
</td>
<td style="text-align:left;">
100
</td>
</tr>
<tr>
<td style="text-align:left;">
Module
</td>
<td style="text-align:left;">
ped
</td>
</tr>
<tr>
<td style="text-align:left;">
MaxSibIter
</td>
<td style="text-align:left;">
42
</td>
</tr>
<tr>
<td style="text-align:left;">
DummyPrefixFemale
</td>
<td style="text-align:left;">
F
</td>
</tr>
<tr>
<td style="text-align:left;">
DummyPrefixMale
</td>
<td style="text-align:left;">
M
</td>
</tr>
<tr>
<td style="text-align:left;">
Complexity
</td>
<td style="text-align:left;">
full
</td>
</tr>
<tr>
<td style="text-align:left;">
UseAge
</td>
<td style="text-align:left;">
yes
</td>
</tr>
<tr>
<td style="text-align:left;">
FindMaybeRel
</td>
<td style="text-align:left;">
FALSE
</td>
</tr>
<tr>
<td style="text-align:left;">
CalcLLR
</td>
<td style="text-align:left;">
TRUE
</td>
</tr>
<tr>
<td style="text-align:left;">
ErrFlavour
</td>
<td style="text-align:left;">
version2.0
</td>
</tr>
<tr>
<td style="text-align:left;">
SequoiaVersion
</td>
<td style="text-align:left;">
2.1.0
</td>
</tr>
<tr>
<td style="text-align:left;">
TimeStart
</td>
<td style="text-align:left;">
2020-10-03 14:10:53
</td>
</tr>
<tr>
<td style="text-align:left;">
TimeEnd
</td>
<td style="text-align:left;">
2020-10-03 14:11:10
</td>
</tr>
</tbody>
</table>
<p>Genotyping error: probability that actual genotype ‘act’ (rows) is
observed as genotype ‘obs’ (columns).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">ErrM</span>,</span>
<span>      caption <span class="op">=</span> <span class="st">'Presumed genotyping error rate'</span>, booktabs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">kable_styling</span><span class="op">(</span>full_width <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<caption>
Presumed genotyping error rate
</caption>
<tbody>
<tr>
<td style="text-align:right;">
0.99990
</td>
<td style="text-align:right;">
0.0001
</td>
<td style="text-align:right;">
0.00000
</td>
</tr>
<tr>
<td style="text-align:right;">
0.00005
</td>
<td style="text-align:right;">
0.9999
</td>
<td style="text-align:right;">
0.00005
</td>
</tr>
<tr>
<td style="text-align:right;">
0.00000
</td>
<td style="text-align:right;">
0.0001
</td>
<td style="text-align:right;">
0.99990
</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: this error matrix can be fully customised, see ?ErrToM</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="genetic-data">Genetic data<a class="anchor" aria-label="anchor" href="#genetic-data"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">GenoM</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="../reference/SnpStats.html">SnpStats</a></span><span class="op">(</span><span class="va">GenoM</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'No genotype matrix provided for this report. \n'</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/SnpStats-1.png" width="90%"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note1: SNPs with missingness &gt;=90% (scored for &lt;10% of individuals) are </span></span>
<span><span class="co">#       automatically excluded.</span></span>
<span><span class="co"># Note2: Higher MAF is better. Monomorphic SNPs are automatically excluded.</span></span>
<span><span class="co"># Note3: Very strong departure from HWE may affect pedigree reconstruction. </span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="life-history-data">Life history data<a class="anchor" aria-label="anchor" href="#life-history-data"></a>
</h3>
<p>In sequoia, sex is coded as 1=female, 2=male, 3=unknown,
4=hermaphrodite.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span>Sex <span class="op">=</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">LifeHist</span><span class="op">$</span><span class="va">Sex</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Sex</span></span>
<span><span class="co">##   1   2 </span></span>
<span><span class="co">##  95 105</span></span></code></pre>
<p>Unknown birth years may hinder pedigree reconstruction. Among others,
they are used to determine which individual is the parent, and which the
offspring in genetically identified parent-offspring pairs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LH</span> <span class="op">&lt;-</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">LifeHist</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="st">'Birth Year'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">LH</span><span class="op">$</span><span class="va">BirthYear</span> <span class="op">&lt;</span><span class="fl">0</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>                            labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'known'</span>, <span class="st">'missing'</span><span class="op">)</span><span class="op">)</span>,</span>
<span>      <span class="st">'min/max Birth Year'</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">LH</span><span class="op">$</span><span class="va">BY.min</span> <span class="op">&lt;</span><span class="fl">0</span> <span class="op">&amp;</span> <span class="va">LH</span><span class="op">$</span><span class="va">BY.max</span> <span class="op">&lt;</span><span class="fl">0</span>, </span>
<span>                                    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">FALSE</span>, <span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>                                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'known'</span>, <span class="st">'missing'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           min/max Birth Year</span></span>
<span><span class="co">## Birth Year known missing</span></span>
<span><span class="co">##    known       0     200</span></span>
<span><span class="co">##    missing     0       0</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">LH</span><span class="op">$</span><span class="va">BirthYear</span>, </span>
<span>     breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">LH</span><span class="op">$</span><span class="va">BirthYear</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">:</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">LH</span><span class="op">$</span><span class="va">BirthYear</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span><span class="fl">.5</span>,</span>
<span>     main <span class="op">=</span> <span class="st">'Distribution of birth years'</span>, xlab<span class="op">=</span><span class="st">''</span><span class="op">)</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/LH-BY-1.png" width="70%"></p>
</div>
<div class="section level3">
<h3 id="age-distribution-prior">Age distribution prior<a class="anchor" aria-label="anchor" href="#age-distribution-prior"></a>
</h3>
<p>The ‘age prior’ specifies the minimum and maximum age of parents, and
the age difference distribution between siblings.</p>
<p>By default, for parentage assignment a flat prior is used with
maximum parental age equal to the largest age difference between
genotyped individuals. The maximum age for dams and sires can also be
specified, as can discrete versus overlapping generations. The
distribution can also be fully customised.</p>
<p>This age difference distribution is updated after parentage
assignment and before full pedigree reconstruction (see further).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">GenoM</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">LH_a</span> <span class="op">&lt;-</span> <span class="va">LH</span><span class="op">[</span><span class="va">LH</span><span class="op">$</span><span class="va">ID</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">GenoM</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">LH_a</span> <span class="op">&lt;-</span> <span class="va">LH</span></span>
<span><span class="op">}</span></span>
<span><span class="va">AP_IN</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">MakeAgePrior</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Pedigree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                      LifeHistData <span class="op">=</span> <span class="va">LH_a</span><span class="op">)</span>,</span>
<span>                                 <span class="va">SeqOUT</span><span class="op">$</span><span class="va">args.AP</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Ageprior: Flat 0/1, overlapping generations, MaxAgeParent = 10,10</span></span></code></pre>
<p><img src="sequoia_report_files/figure-html/AP-in-1.png" width="70%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Note: Details about and help on the age prior can be found in </span></span>
<span><span class="co">#       vignette("Sequoia - Age 'Prior'")</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">args.AP</span></span></code></pre></div>
<pre><code><span><span class="co">## NULL</span></span></code></pre>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="output-summary">Output summary<a class="anchor" aria-label="anchor" href="#output-summary"></a>
</h2>
<div class="section level3">
<h3 id="pedigree-summary">Pedigree summary<a class="anchor" aria-label="anchor" href="#pedigree-summary"></a>
</h3>
<p>Sibling clusters sharing a non-genotyped parent are assigned a
‘dummy’ parent. Via grandparent assignment to sibling clusters, parents
are assigned to these dummy individuals. These grandparents may be both
genotyped or dummy individuals.</p>
<p>Identifying the real non-genotyped individual corresponding to each
dummy individual may not always be possible, but the <a href="https://jiscah.github.io/articles/pedcompare_example.html"><code>sequoia</code>
website</a> offers some suggestions when candidates are known.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sumry</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">SeqOUT</span>, Plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/PlotSeqSum.html">PlotSeqSum</a></span><span class="op">(</span><span class="va">sumry</span>, <span class="va">SeqOUT</span><span class="op">$</span><span class="va">Pedigree</span>, Panels<span class="op">=</span><span class="st">'all'</span>, ask<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/sumry-plots-1.png" width="100%"><img src="sequoia_report_files/figure-html/sumry-plots-2.png" width="100%"><img src="sequoia_report_files/figure-html/sumry-plots-3.png" width="100%"><img src="sequoia_report_files/figure-html/sumry-plots-4.png" width="100%"><img src="sequoia_report_files/figure-html/sumry-plots-5.png" width="100%"></p>
<p>The distributions of the number of opposing homozygous loci and other
Mendelian errors in parent-offspring pairs and parent-parent-offspring
trios give a rough impression of the genotyping error rate. These would
always be zero in absence of any genotyping errors (and in absence of
pedigree errors).</p>
<p>The distributions of the log10 likelihood ratios give a rough
impression of the power of the genetic data to distinguish between
different types of relationships and resolve the pedigree. Note that
this is the likelihood ratio between the assigned parent being the
parent versus it being another type of close relative, such as a full
sibling of the focal individual or the true parent. It is <em>not</em>
relative to other candidate parents of the focal individual.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">kable</span><span class="op">(</span><span class="va">sumry</span><span class="op">$</span><span class="va">PedSummary</span>, booktabs<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">kable_styling</span><span class="op">(</span>full_width <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<table class="table table" style="width: auto !important; margin-left: auto; margin-right: auto;">
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:right;">
All
</th>
<th style="text-align:right;">
SNPd
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
records
</td>
<td style="text-align:right;">
170
</td>
<td style="text-align:right;">
142
</td>
</tr>
<tr>
<td style="text-align:left;">
maternities
</td>
<td style="text-align:right;">
125
</td>
<td style="text-align:right;">
65
</td>
</tr>
<tr>
<td style="text-align:left;">
paternities
</td>
<td style="text-align:right;">
122
</td>
<td style="text-align:right;">
79
</td>
</tr>
<tr>
<td style="text-align:left;">
full sibs
</td>
<td style="text-align:right;">
5
</td>
<td style="text-align:right;">
5
</td>
</tr>
<tr>
<td style="text-align:left;">
maternal half sib
</td>
<td style="text-align:right;">
131
</td>
<td style="text-align:right;">
87
</td>
</tr>
<tr>
<td style="text-align:left;">
paternal half sibs
</td>
<td style="text-align:right;">
114
</td>
<td style="text-align:right;">
77
</td>
</tr>
<tr>
<td style="text-align:left;">
maternal grandmothers
</td>
<td style="text-align:right;">
91
</td>
<td style="text-align:right;">
77
</td>
</tr>
<tr>
<td style="text-align:left;">
maternal grandfathers
</td>
<td style="text-align:right;">
90
</td>
<td style="text-align:right;">
80
</td>
</tr>
<tr>
<td style="text-align:left;">
paternal grandmothers
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
60
</td>
</tr>
<tr>
<td style="text-align:left;">
paternal grandfathers
</td>
<td style="text-align:right;">
73
</td>
<td style="text-align:right;">
61
</td>
</tr>
<tr>
<td style="text-align:left;">
maximum pedigree depth
</td>
<td style="text-align:right;">
8
</td>
<td style="text-align:right;">
7
</td>
</tr>
<tr>
<td style="text-align:left;">
founders
</td>
<td style="text-align:right;">
36
</td>
<td style="text-align:right;">
39
</td>
</tr>
</tbody>
</table>
<p>The size, depth and interconnectedness of the pedigree affect the
power with which the pedigree can be used in subsequent analyses, such
as heritability estimates.</p>
</div>
<div class="section level3">
<h3 id="age-distribution">Age distribution<a class="anchor" aria-label="anchor" href="#age-distribution"></a>
</h3>
<p>After assignment of genotyped parents to genotyped offspring, the
reconstructed pedigree is combined with the provided birth year
information to estimate the distribution of age differences among
mother-offspring, father-offspring, and sibling pairs. This age
distribution is then used during further pedigree reconstruction.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/PlotAgePrior.html">PlotAgePrior</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">AgePriors</span><span class="op">)</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/age-1.png" width="70%"></p>
<div class="section level4">
<h4 id="estimated-birth-years">Estimated birth years<a class="anchor" aria-label="anchor" href="#estimated-birth-years"></a>
</h4>
<p>When a birth year is unknown, it is estimated from combining the
above distribution of parental ages with any known birth years of its
assigned parents and offspring.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LH_new</span> <span class="op">&lt;-</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">LifeHistSib</span></span>
<span><span class="va">LH_new</span><span class="op">[</span><span class="va">LH_new</span><span class="op">$</span><span class="va">BirthYear</span> <span class="op">&lt;</span> <span class="fl">0</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] id        Sex       BirthYear BY.min    BY.max    Sexx      BY.est   </span></span>
<span><span class="co">## [8] BY.lo     BY.hi    </span></span>
<span><span class="co">## &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="inferred-sex">Inferred sex<a class="anchor" aria-label="anchor" href="#inferred-sex"></a>
</h4>
<p>When the sex of an individual is missing from the input, it may be
inferred during pedigree reconstruction when this individual forms a
complementary parent pair with an individual of known sex (in species
without hermaphrodites).</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LH_new</span><span class="op">[</span><span class="va">LH_new</span><span class="op">$</span><span class="va">Sex</span> <span class="op">==</span> <span class="fl">3</span> <span class="op">&amp;</span> <span class="va">LH_new</span><span class="op">$</span><span class="va">Sexx</span> <span class="op">!=</span> <span class="fl">3</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] id        Sex       BirthYear BY.min    BY.max    Sexx      BY.est   </span></span>
<span><span class="co">## [8] BY.lo     BY.hi    </span></span>
<span><span class="co">## &lt;0 rows&gt; (or 0-length row.names)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="non-assigned-likely-relatives">Non-assigned likely relatives<a class="anchor" aria-label="anchor" href="#non-assigned-likely-relatives"></a>
</h3>
<p>The <code>sequoia</code> algorithm is rather conservative when making
assignments; it sequentially ‘grows’ the pedigree, and tries to avoid a
snowball effect of assignment errors. So, when no parent or siblings are
assigned, this does not necessarily mean these are not present in the
dataset.</p>
<p>The R package includes a separate function to identify pairs of
likely relatives that have not been assigned, which may be due to a
variety of reasons.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">Maybe</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">MaybePO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetRelM.html">GetRelM</a></span><span class="op">(</span>Pairs<span class="op">=</span><span class="va">Maybe</span><span class="op">$</span><span class="va">MaybePar</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">MaybePO</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="st">'MaybeRel'</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">Maybe</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">MaybeM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetRelM.html">GetRelM</a></span><span class="op">(</span>Pairs<span class="op">=</span><span class="va">Maybe</span><span class="op">$</span><span class="va">MaybeRel</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="../reference/PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">MaybeM</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">GenoM</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">Maybe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetMaybeRel.html">GetMaybeRel</a></span><span class="op">(</span><span class="va">GenoM</span>, SeqList <span class="op">=</span> <span class="va">SeqOUT</span>, Module <span class="op">=</span> <span class="st">'ped'</span>, quiet<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="va">MaybeM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/GetRelM.html">GetRelM</a></span><span class="op">(</span>Pairs<span class="op">=</span><span class="va">Maybe</span><span class="op">$</span><span class="va">MaybeRel</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">MaybeM</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"No 'maybe relatives' provided for this report. \n"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/mayberel-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="confidence-probabilities">Confidence probabilities<a class="anchor" aria-label="anchor" href="#confidence-probabilities"></a>
</h3>
<p>The assignment accuracy is estimated from simulations. These
simulations make several simplifying assumptions, and these numbers are
therefore a lower bound for the confidence probability.</p>
<p>The probability is not estimated for each individual separately.
Instead, they are grouped across a few categories, based on whether the
parent is dam or sire, genotyped or a dummy, and whether there is a
co-parent or not. It is also separated for genotyped versus dummy focal
individuals.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">Conf</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">kable</span><span class="op">(</span><span class="va">Conf</span><span class="op">$</span><span class="va">ConfProb</span>,</span>
<span>        caption <span class="op">=</span> <span class="st">'parent-pair confidence, per category (Genotyped/Dummy/None)'</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'No confidence probabilities provided for this report. \n'</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<table class="table">
<caption>
parent-pair confidence, per category (Genotyped/Dummy/None)
</caption>
<thead><tr>
<th style="text-align:left;">
</th>
<th style="text-align:left;">
id.cat
</th>
<th style="text-align:left;">
dam.cat
</th>
<th style="text-align:left;">
sire.cat
</th>
<th style="text-align:right;">
dam.conf
</th>
<th style="text-align:right;">
sire.conf
</th>
<th style="text-align:right;">
pair.conf
</th>
<th style="text-align:right;">
N
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
14
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
568
</td>
</tr>
<tr>
<td style="text-align:left;">
13
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
277
</td>
</tr>
<tr>
<td style="text-align:left;">
15
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
227
</td>
</tr>
<tr>
<td style="text-align:left;">
11
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.9841270
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
0.9841270
</td>
<td style="text-align:right;">
252
</td>
</tr>
<tr>
<td style="text-align:left;">
10
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
88
</td>
</tr>
<tr>
<td style="text-align:left;">
12
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
49
</td>
</tr>
<tr>
<td style="text-align:left;">
17
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
197
</td>
</tr>
<tr>
<td style="text-align:left;">
16
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
38
</td>
</tr>
<tr>
<td style="text-align:left;">
18
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
804
</td>
</tr>
<tr>
<td style="text-align:left;">
5
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.9914530
</td>
<td style="text-align:right;">
0.9914530
</td>
<td style="text-align:right;">
0.9914530
</td>
<td style="text-align:right;">
117
</td>
</tr>
<tr>
<td style="text-align:left;">
4
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
6
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
34
</td>
</tr>
<tr>
<td style="text-align:left;">
2
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
0.8888889
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
0.8888889
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
1
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
NaN
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
3
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
0.7777778
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
9
</td>
</tr>
<tr>
<td style="text-align:left;">
8
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
G
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
0.9705882
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
34
</td>
</tr>
<tr>
<td style="text-align:left;">
7
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
1.0000000
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
3
</td>
</tr>
<tr>
<td style="text-align:left;">
9
</td>
<td style="text-align:left;">
D
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:left;">
X
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
NA
</td>
<td style="text-align:right;">
100
</td>
</tr>
</tbody>
</table>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="further-details">Further details<a class="anchor" aria-label="anchor" href="#further-details"></a>
</h2>
<div class="section level3">
<h3 id="likelihood-curve">Likelihood curve<a class="anchor" aria-label="anchor" href="#likelihood-curve"></a>
</h3>
<p>The total likelihood is the probability of observing the genetic
data, given the inferred pedigree and the presumed genotyping error
rate. The initial likelihood is the probability if all genotyped
individuals were unrelated and sampled from a large population in
Hardy-Weinberg Equilibrium. The likelihood increases during pedigree
reconstruction, and pedigree reconstruction is terminated when the total
likelihood asymptotes.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nIt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>par <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">TotLikPar</span><span class="op">)</span>,</span>
<span>         ped <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">SeqOUT</span><span class="op">$</span><span class="va">TotLikSib</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">SeqOUT</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">nIt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="va">TotLikPar</span>, type<span class="op">=</span><span class="st">"b"</span>, lwd<span class="op">=</span><span class="fl">2</span>, col<span class="op">=</span><span class="st">"forestgreen"</span>,</span>
<span>                  xlim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">nIt</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span>, xlab<span class="op">=</span><span class="st">"Iteration (ped)"</span>, xaxt<span class="op">=</span><span class="st">'n'</span>, cex.lab<span class="op">=</span><span class="fl">1.2</span>,</span>
<span>                  ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">TotLikPar</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">TotLikSib</span><span class="op">)</span><span class="op">)</span>, ylab<span class="op">=</span><span class="st">"Total log-likelihood"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">SeqOUT</span>, <span class="fu"><a href="https://rdrr.io/r/graphics/lines.html" class="external-link">lines</a></span><span class="op">(</span><span class="op">(</span><span class="va">nIt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">:</span><span class="va">nIt</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="va">TotLikSib</span>, type<span class="op">=</span><span class="st">"b"</span>, lwd<span class="op">=</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="fl">1</span>, labels <span class="op">=</span> <span class="st">'HWE'</span>, cex.axis<span class="op">=</span><span class="fl">1.2</span>, col.axis<span class="op">=</span><span class="st">'darkgrey'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="op">(</span><span class="va">nIt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">2</span>, labels <span class="op">=</span> <span class="st">'par'</span>, lwd.ticks<span class="op">=</span><span class="fl">0</span>, col.axis<span class="op">=</span><span class="st">'forestgreen'</span>, </span>
<span>     cex.axis<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/axis.html" class="external-link">axis</a></span><span class="op">(</span><span class="fl">1</span>, at<span class="op">=</span><span class="op">(</span><span class="va">nIt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">:</span><span class="op">(</span><span class="va">nIt</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">+</span><span class="va">nIt</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>, labels<span class="op">=</span><span class="fl">0</span><span class="op">:</span><span class="va">nIt</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, cex.axis<span class="op">=</span><span class="fl">1.2</span><span class="op">)</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/totlik-1.png" width="70%"></p>
</div>
<div class="section level3">
<h3 id="genotyping-error-estimate">Genotyping error estimate<a class="anchor" aria-label="anchor" href="#genotyping-error-estimate"></a>
</h3>
<p>This estimate is based on Mendelian inconsistencies between assigned
parent-offspring pairs. It is a ballpark estimate of the genotyping
error for each SNP, as not all genotyping errors result in such
inconsistencies, and true parent-offspring pairs with many
inconsistencies are less likely to be assigned. On the other hand, a
single genotyping error in an individual with many offspring may lead to
a high number of Mendelian inconsistencies.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="va">GenoM</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'NA'</span>, <span class="cn">NA</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">sstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="../reference/SnpStats.html">SnpStats</a></span><span class="op">(</span><span class="va">GenoM</span>, Pedigree <span class="op">=</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">Pedigree</span>,</span>
<span>             ErrFlavour <span class="op">=</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">Specs</span><span class="op">$</span><span class="va">ErrFlavour</span>, Plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">sstats</span><span class="op">[</span>,<span class="st">'Err.hat'</span><span class="op">]</span>, breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sstats</span><span class="op">)</span><span class="op">/</span><span class="fl">5</span>, </span>
<span>       main <span class="op">=</span> <span class="st">'Estimated genotyping error'</span>, xlab<span class="op">=</span><span class="st">''</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">'No genotype matrix provided for this report. \n'</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="sequoia_report_files/figure-html/SnpStats-2-1.png" width="70%"></p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
