<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate assignment probabilities — CalcParentProbs • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate assignment probabilities — CalcParentProbs"><meta name="description" content="For each assigned offspring-parent pair, calculate the
 probability they are parent-offspring vs otherwise related. Probabilities
 are scaled to sum to one across all possible* relationships between the pair
 or trio; see Details."><meta property="og:description" content="For each assigned offspring-parent pair, calculate the
 probability they are parent-offspring vs otherwise related. Probabilities
 are scaled to sum to one across all possible* relationships between the pair
 or trio; see Details."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate assignment probabilities</h1>

      <div class="d-none name"><code>CalcParentProbs.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>For each assigned offspring-parent pair, calculate the
 probability they are parent-offspring vs otherwise related. Probabilities
 are scaled to sum to one across all possible* relationships between the pair
 or trio; see Details.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CalcParentProbs</span><span class="op">(</span>Pedigree <span class="op">=</span> <span class="cn">NULL</span>, GenoM <span class="op">=</span> <span class="cn">NULL</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span>, nCores <span class="op">=</span> <span class="fl">1</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#arg-pedigree"></a></dt>
<dd><p>dataframe with columns id-dam-sire. By default, any
non-genotyped individuals are 'dummified'; use <code>Module='par'</code> to
ignore them.</p></dd>


<dt id="arg-genom">GenoM<a class="anchor" aria-label="anchor" href="#arg-genom"></a></dt>
<dd><p>numeric matrix with genotype data: One row per individual,
one column per SNP, coded as 0, 1, 2, missing values as a negative number
or NA. You can reformat data with <code><a href="GenoConvert.html">GenoConvert</a></code>, or use other
packages to get it into a genlight object and then use <code>as.matrix</code>.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>logical, suppress messages. No progress is printed when &gt;1 core
is used.</p></dd>


<dt id="arg-ncores">nCores<a class="anchor" aria-label="anchor" href="#arg-ncores"></a></dt>
<dd><p>number of computer cores to use. If <code>2</code> or <code>4</code>,
package <span class="pkg">parallel</span> is used (other values are not applicable).</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to <code><a href="CalcPairLL.html">CalcPairLL</a></code>, such as
the genotyping error rate <code>Err</code>, age information in
<code>LifeHistData</code> and <code>AgePrior</code>, or <code>InclDup</code> to include the
probability that the two samples are duplicates.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>the <code>Pedigree</code> dataframe with the three applicable columns
renamed to id-dam-sire, and 7 additional columns:</p>
<dl><dt>Probdam</dt>
<dd><p>Probability that individual in dam column is the maternal
  parent, rather than otherwise related (LL(PO)/sum(LL))</p></dd>

<dt>Probsire</dt>
<dd><p>Analogous for sire</p></dd>

<dt>Probpair</dt>
<dd><p>Probability for id-dam-sire trio. Approximated as the minimum
  of dam conditional on sire and sire conditional on dam, thus not including
  e.g. both being siblings (those other configurations are considered by
  sequoia during pedigree reconstruction, but can (currently) not be accessed
  directly)</p></dd>

<dt>dam_alt, sire_alt</dt>
<dd><p>Most likely alternative (not PO) relationship
 between id-dam and id-sire, respectively</p></dd>

<dt>Probdam_alt, Probsire_alt</dt>
<dd><p>Probability of most likely alternative
 relationship</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The returned probabilities are calculated from the likelihoods used
  throughout the rest of this package, by scaling them to sum to one across
  all possible relationships. For <code>Complex='simp'</code> these are
  PO=parent-offspring, FS=full siblings, HS=half siblings, GP=grand-parental,
  FA=full avuncular, HA=third degree relatives (incl half avuncular), and
  U=unrelated. For <code>Complex='full'</code> there are numerous double
  relationship considered (PO &amp; HS, HS &amp; HA, etc), making both numerator and
  denominator in the scaling step less unambiguous, and the returned
  probabilities an approximation.</p>
<p>The likelihoods are calculated by calling <code><a href="CalcPairLL.html">CalcPairLL</a></code> once or
 twice for each id-dam and id-sire pair: once not conditioning on the
 co-parent, and once conditional on the co-parent, if any. For genotyped
 individuals this is done with <code>focal='PO'</code>, and for dummy individuals
 with <code>focal='GP'</code>.</p>
<p>For relationships between a genotyped and a dummy individual, it may only be
 possible to determine that the genotyped individual is a second degree
 relative (GP, HS, or FA) to the dummy's offspring. This then results in a
 probability of at most 0.33, even when the two are indeed parent and
 offspring.</p>
<p>See <code><a href="CalcPairLL.html">CalcPairLL</a></code> and the vignettes for further details.</p>
<p>Note that for large pedigrees this function can be fairly slow, especially
when using <code><a href="CalcPairLL.html">CalcPairLL</a></code>'s default <code>Module='ped'</code> and
<code>Complex='full'</code>.</p>
<p>Subsetting the genotype data may give different results, as the likelihoods
 and thus the probabilities depend on the allele frequencies in the sample.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CalcPairLL.html">CalcPairLL</a></code>, <code><a href="LLtoProb.html">LLtoProb</a></code></p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">test_ped</span> <span class="op">&lt;-</span> <span class="va">Ped_griffin</span><span class="op">[</span><span class="fl">21</span><span class="op">:</span><span class="fl">25</span>,<span class="op">]</span></span></span>
<span class="r-in"><span><span class="co"># add an incorrect sire to illustrate</span></span></span>
<span class="r-in"><span><span class="va">test_ped</span><span class="op">$</span><span class="va">sire</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">test_ped</span><span class="op">$</span><span class="va">sire</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">test_ped</span><span class="op">$</span><span class="va">sire</span><span class="op">[</span><span class="fl">5</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">'i057_2003_M'</span></span></span>
<span class="r-in"><span><span class="va">Ped_with_probs</span> <span class="op">&lt;-</span> <span class="fu">CalcParentProbs</span><span class="op">(</span><span class="va">test_ped</span>, <span class="va">Geno_griffin</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> `GenoM` and `Pedigree` share few common individuals</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calculating LL for  dam_solo</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 145 individuals, 5 dams and 4 sires</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calculating LL for  sire_solo</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 145 individuals, 5 dams and 4 sires</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calculating LL for  dam_paired</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 145 individuals, 5 dams and 4 sires</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>  </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Calculating LL for  sire_paired</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 145 individuals, 5 dams and 4 sires</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">Ped_with_probs</span>, digits<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>            id         dam        sire birthyear Probdam dam_alt Probdam_alt</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 i021_2002_M i001_2001_F i020_2001_M      2002       1      FS     2.4e-06</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 i022_2002_F i015_2001_F i006_2001_M      2002       1      GP     2.8e-09</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 i023_2002_F i007_2001_F i019_2001_M      2002       1      FS     7.9e-07</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 i024_2002_M i013_2001_F        &lt;NA&gt;      2002      NA    &lt;NA&gt;          NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 i025_2002_M i014_2001_F i057_2003_M      2002       1      GP     2.3e-08</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Probsire sire_alt Probsire_alt Probpair</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    1e+00       HS      2.4e-08  1.0e+00</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    1e+00       GP      9.3e-08  1.0e+00</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3       NA     &lt;NA&gt;           NA       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4       NA     &lt;NA&gt;           NA       NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    6e-34       HS      2.5e-01  6.7e-61</span>
<span class="r-in"><span><span class="co"># Any non-genotyped non-'dummifiable' individuals are automatically skipped</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># To get likelihoods for 'all' relationships, not just probabilities for</span></span></span>
<span class="r-in"><span><span class="co"># PO &amp; (next-)most-likely:</span></span></span>
<span class="r-in"><span><span class="va">LL_sire_single</span> <span class="op">&lt;-</span> <span class="fu"><a href="CalcPairLL.html">CalcPairLL</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  Pairs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>id1<span class="op">=</span><span class="va">test_ped</span><span class="op">$</span><span class="va">id</span>,</span></span>
<span class="r-in"><span>                     id2<span class="op">=</span><span class="va">test_ped</span><span class="op">$</span><span class="va">sire</span>,</span></span>
<span class="r-in"><span>                     dropPar1<span class="op">=</span><span class="st">'both'</span>, <span class="co"># drop both -&gt; id2 as single parent</span></span></span>
<span class="r-in"><span>                     focal<span class="op">=</span><span class="st">'PO'</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  Pedigree <span class="op">=</span> <span class="va">Ped_griffin</span>,   <span class="co"># pedigree to condition on</span></span></span>
<span class="r-in"><span>  GenoM <span class="op">=</span> <span class="va">Geno_griffin</span>, Plot<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 200 individuals, 167 dams and 163 sires</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #BBBB00;">!</span> Assuming columns 1 and 2 of Pairs are 'ID1' and 'ID2'</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

