<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Count opposing homozygous SNPs between pairs of individuals — CountOH • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Count opposing homozygous SNPs between pairs of individuals — CountOH"><meta name="description" content="Quick identification of likely parents, as the number of
  opposing homozygous (OH) SNPs is expected to be zero for parent- offspring
  pairs in absence of genotyping errors, and greater than zero for all other
  pairs."><meta property="og:description" content="Quick identification of likely parents, as the number of
  opposing homozygous (OH) SNPs is expected to be zero for parent- offspring
  pairs in absence of genotyping errors, and greater than zero for all other
  pairs."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.2.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Count opposing homozygous SNPs between pairs of individuals</h1>

      <div class="d-none name"><code>CountOH.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Quick identification of likely parents, as the number of
  opposing homozygous (OH) SNPs is expected to be zero for parent- offspring
  pairs in absence of genotyping errors, and greater than zero for all other
  pairs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">CountOH</span><span class="op">(</span>x <span class="op">=</span> <span class="cn">NULL</span>, ID2 <span class="op">=</span> <span class="cn">NULL</span>, GenoM <span class="op">=</span> <span class="cn">NULL</span>, max_OH <span class="op">=</span> <span class="op">-</span><span class="fl">1</span>, quiet <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>Either a matrix, dataframe or similar where the first two columns
are individual IDs, or a vector with IDs. In the second case, you may
provide <code>ID2</code>, and the output will be an ID1 x ID2 matrix; else the
output will be an ID1 x ID1 matrix. Non-genotyped individuals are included
in the results with all <code>NA</code>'s, and a warning.</p></dd>


<dt id="arg-id-">ID2<a class="anchor" aria-label="anchor" href="#arg-id-"></a></dt>
<dd><p>optional second vector with IDs</p></dd>


<dt id="arg-genom">GenoM<a class="anchor" aria-label="anchor" href="#arg-genom"></a></dt>
<dd><p>numeric matrix with genotype data: One row per individual, one
column per SNP, coded as 0, 1, 2, missing values as a negative number or
NA. Row names must be individual IDs, column names are ignored. You can
reformat data with <code><a href="GenoConvert.html">GenoConvert</a></code>, or use other packages to get
it into a genlight object and then use <code>as.matrix</code>.</p></dd>


<dt id="arg-max-oh">max_OH<a class="anchor" aria-label="anchor" href="#arg-max-oh"></a></dt>
<dd><p>stop counting OH's for a pair if this value is reached, to
reduce computation time. Ignored if negative value or equal to total number
of SNPs.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>suppress messages</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Counting the number of opposing homozygous (OH) SNPs is much faster
than calculating likelihoods, and does not rely on an estimated genotyping
error rate. It can therefore be useful during quality control, or to help
figure out problems when assignment rate with <code><a href="sequoia.html">sequoia</a></code> is
 lower than expected.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CalcPairLL.html">CalcPairLL</a></code> to calculate likelihoods for pairs,
 <code><a href="CalcOHLLR.html">CalcOHLLR</a></code> to calculate OH for a pedigree,
 <code><a href="CalcMaxMismatch.html">CalcMaxMismatch</a></code> for calculation of the maximum OH used by
 <code><a href="sequoia.html">sequoia</a></code> to filter potential parent-offspring pairs.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">offspring_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">LH_HSg5</span>, <span class="va">ID</span><span class="op">[</span><span class="va">BirthYear</span><span class="op">==</span><span class="fl">2001</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">candidate_father_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">LH_HSg5</span>, <span class="va">ID</span><span class="op">[</span><span class="va">BirthYear</span><span class="op">==</span><span class="fl">2000</span> <span class="op">&amp;</span> <span class="va">Sex</span><span class="op">==</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">OH_matrix</span> <span class="op">&lt;-</span> <span class="fu">CountOH</span><span class="op">(</span><span class="va">offspring_ids</span>, <span class="va">candidate_father_ids</span>, GenoM<span class="op">=</span><span class="va">Geno_HSg5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Some IDs are not among rownames of GenoM</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">OH_matrix</span><span class="op">)</span>, breaks<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">40</span><span class="op">)</span><span class="op">-</span><span class="fl">.5</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="CountOH-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># with high quality SNP data, there is often a clear separation in OH counts</span></span></span>
<span class="r-in"><span><span class="co"># between parent-offspring pairs (here: OH&lt;3) and others (here: OH&gt;4).</span></span></span>
<span class="r-in"><span><span class="co"># BUT: non-PO close relatives may have very low OH counts by chance,</span></span></span>
<span class="r-in"><span><span class="co"># and true PO pairs may have fairly high OH counts due to genotyping errors.</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

