<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Convert Genotyping Error Rates from per-allele to per-locus — Err_RADseq • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Convert Genotyping Error Rates from per-allele to per-locus — Err_RADseq"><meta name="description" content="Convert per-allele genotyping rates at homozygous (E0) and
  heterozygous (E1) sites to a length-3 vector with per-locus error rates
  hom|hom, het|hom, hom|het."><meta property="og:description" content="Convert per-allele genotyping rates at homozygous (E0) and
  heterozygous (E1) sites to a length-3 vector with per-locus error rates
  hom|hom, het|hom, hom|het."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Convert Genotyping Error Rates from per-allele to per-locus</h1>

      <div class="d-none name"><code>Err_RADseq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convert per-allele genotyping rates at homozygous (E0) and
  heterozygous (E1) sites to a length-3 vector with per-locus error rates
  hom|hom, het|hom, hom|het.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Err_RADseq</span><span class="op">(</span>E0 <span class="op">=</span> <span class="fl">0.005</span>, E1 <span class="op">=</span> <span class="fl">0.05</span>, Return <span class="op">=</span> <span class="st">"vector"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-e-">E0<a class="anchor" aria-label="anchor" href="#arg-e-"></a></dt>
<dd><p>per-allele genotyping rates at homozygous sites</p></dd>


<dt id="arg-e-">E1<a class="anchor" aria-label="anchor" href="#arg-e-"></a></dt>
<dd><p>per-allele genotyping rates at heterozygous sites</p></dd>


<dt id="arg-return">Return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>output format, 'vector' (default) or 'matrix'</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Depending on <code>Return</code>, either:</p><ul><li><p><code>'vector'</code>: a length 3 vector, with the probabilities (observed
     given actual) hom|other hom, het|hom, and hom|het.</p></li>
<li><p><code>'matrix'</code>: a 3x3 matrix, with probabilities of observed
     genotypes (columns) conditional on actual (rows)</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Estimation of per-allele genotyping rates is described in Bresadola
  et al (2020) - 'Estimating and accounting for genotyping errors in RAD-seq
  experiments', MER. The error model implemented here is identical to that in
  Table 1 of that paper, and the default values are also taken from that
  paper.</p>
<p>For further information on how the sequoia package handles genotyping errors,
 see <code><a href="ErrToM.html">ErrToM</a></code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Compare with default error pattern (SNP chip based) :</span></span></span>
<span class="r-in"><span><span class="fu">Err_RADseq</span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.001</span>, E1<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.000001 0.001998 0.047500</span>
<span class="r-in"><span><span class="fu"><a href="ErrToM.html">ErrToM</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fl">0.05</span><span class="op">)</span><span class="op">*</span><span class="fl">2</span>, Return<span class="op">=</span><span class="st">'vector'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] 0.00225625 0.09274375 0.04750000</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># usage in sequoia() and other functions:</span></span></span>
<span class="r-in"><span><span class="va">Err_low</span> <span class="op">&lt;-</span> <span class="fu">Err_RADseq</span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.002</span>, E1<span class="op">=</span><span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">Err_high</span> <span class="op">&lt;-</span> <span class="fu">Err_RADseq</span><span class="op">(</span>E0<span class="op">=</span><span class="fl">0.01</span>, E1<span class="op">=</span><span class="fl">0.15</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span> <span class="va">SeqOUT_lowErr</span> <span class="op">&lt;-</span> <span class="fu"><a href="sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoM</span>, <span class="va">LHdata</span>, Err<span class="op">=</span><span class="va">Err_low</span><span class="op">)</span></span></span>
<span class="r-in"><span> <span class="va">SeqOUT_highErr</span> <span class="op">&lt;-</span> <span class="fu"><a href="sequoia.html">sequoia</a></span><span class="op">(</span><span class="va">GenoM</span>, <span class="va">LHdata</span>, Err<span class="op">=</span><span class="va">Err_high</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># also usable for confidence estimates, and to explore potential consequences</span></span></span>
<span class="r-in"><span><span class="co"># of the actual genotyping error rate being much higher/lower than assumed</span></span></span>
<span class="r-in"><span><span class="va">EC</span> <span class="op">&lt;-</span> <span class="fu"><a href="EstConf.html">EstConf</a></span><span class="op">(</span><span class="va">best_pedigree</span>, <span class="va">LHdata</span>, args.sim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>SnpError<span class="op">=</span><span class="va">Err_high</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>         args.seq<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Err<span class="op">=</span><span class="va">Err_low</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

