<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate Genotyping Error Matrix — ErrToM • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate Genotyping Error Matrix — ErrToM"><meta name="description" content="Make a vector or matrix specifying the genotyping error
  pattern, or a function to generate such a vector/matrix from a single
  value Err.
with the probabilities of observed genotypes
  (columns) conditional on actual genotypes (rows), or return a function to
  generate such matrices (using a single value Err as input to that
 function)."><meta property="og:description" content="Make a vector or matrix specifying the genotyping error
  pattern, or a function to generate such a vector/matrix from a single
  value Err.
with the probabilities of observed genotypes
  (columns) conditional on actual genotypes (rows), or return a function to
  generate such matrices (using a single value Err as input to that
 function)."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Generate Genotyping Error Matrix</h1>

      <div class="d-none name"><code>ErrToM.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Make a vector or matrix specifying the genotyping error
  pattern, or a function to generate such a vector/matrix from a single
  value Err.</p>
<p>with the probabilities of observed genotypes
  (columns) conditional on actual genotypes (rows), or return a function to
  generate such matrices (using a single value Err as input to that
 function).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ErrToM</span><span class="op">(</span>Err <span class="op">=</span> <span class="cn">NA</span>, flavour <span class="op">=</span> <span class="st">"version2.9"</span>, Return <span class="op">=</span> <span class="st">"matrix"</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-err">Err<a class="anchor" aria-label="anchor" href="#arg-err"></a></dt>
<dd><p>estimated genotyping error rate, as a single number, or 3x3 or 4x4
matrix, or length 3 vector. If a single number, an error model is used that
aims to deal with scoring errors typical for SNP arrays. If a matrix, this
should be the probability of observed genotype (columns) conditional on
actual genotype (rows). Each row must therefore sum to 1. If
<code>Return='function'</code>, this may be <code>NA</code>. If a vector, these are the
probabilities (observed given actual) hom|other hom, het|hom, and hom|het.</p></dd>


<dt id="arg-flavour">flavour<a class="anchor" aria-label="anchor" href="#arg-flavour"></a></dt>
<dd><p>vector-generating or matrix-generating function, or one of
'version2.9', 'version2.0', 'version1.3' (='SNPchip'), 'version1.1'
(='version111'), referring to the sequoia version in which it was used as
default. Only used if <code>Err</code> is a single number.</p></dd>


<dt id="arg-return">Return<a class="anchor" aria-label="anchor" href="#arg-return"></a></dt>
<dd><p>output, 'matrix' (default), 'vector', 'function'
(matrix-generating), or 'v_function' (vector-generating)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Depending on <code>Return</code>, either:</p><ul><li><p><code>'matrix'</code>: a 3x3 matrix, with probabilities of observed genotypes
   (columns) conditional on actual (rows)</p></li>
<li><p><code>'function'</code>: a function taking a single value <code>Err</code> as input, and
   generating a 3x3 matrix</p></li>
<li><p><code>'vector'</code>: a length 3 vector, with the probabilities (observed given
     actual) hom|other hom, het|hom, and hom|het.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>By default (<code>flavour</code> = "version2.9"), <code>Err</code> is
  interpreted as a locus-level error rate (rather than allele-level), and
  equals the probability that an actual heterozygote is observed as either
  homozygote (i.e., the probability that it is observed as AA = probability
  that observed as aa = <code>Err</code>/2). The probability that one homozygote is
  observed as the other is (<code>Err</code>/2\()^2\).</p>
<p>The inbuilt 'flavours' correspond to the presumed and simulated error
structures, which have changed with sequoia versions. The most appropriate
error structure will depend on the genotyping platform; 'version0.9' and
'version1.1' were inspired by SNP array genotyping while 'version1.3' and
'version2.0' are intended to be more general.</p>
<p>This function, and throughout the package, it is assumed that the two alleles
\(A\) and \(a\) are equivalent. Thus, using notation \(P\)(observed
genotype |actual genotype), that \(P(AA|aa) = P(aa|AA)\),
\(P(aa|Aa)=P(AA|Aa)\), and \(P(aA|aa)=P(aA|AA)\).</p>
<table class="table table"><tr><td><strong>version</strong></td><td><strong>hom|hom</strong></td><td><strong>het|hom</strong></td><td><strong>hom|het</strong></td></tr><tr><td><strong>2.9</strong></td><td>\((E/2)^2\)</td><td>\(E-(E/2)^2\)</td><td>\(E/2\)</td></tr><tr><td><strong>2.0</strong></td><td>\((E/2)^2\)</td><td>\(E(1-E/2)\)</td><td>\(E/2\)</td></tr><tr><td><strong>1.3</strong></td><td>\((E/2)^2\)</td><td>\(E\)</td><td>\(E/2\)</td></tr><tr><td><strong>1.1</strong></td><td>\(E/2\)</td><td>\(E/2\)</td><td>\(E/2\)</td></tr><tr><td><strong>0.9</strong></td><td>\(0\)</td><td>\(E\)</td><td>\(E/2\)</td></tr></table><p>or in matrix form, Pr(observed genotype (columns) | actual genotype (rows)):</p>
<p><em>version2.9:</em></p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E\)</td><td>\(E -(E/2)^2\)</td><td>\((E/2)^2\)</td></tr><tr><td><strong>1</strong></td><td>\(E/2\)</td><td>\(1-E\)</td><td>\(E/2\)</td></tr><tr><td><strong>2</strong></td><td>\((E/2)^2\)</td><td>\(E -(E/2)^2\)</td><td>\(1-E\)</td></tr></table><p><em>version2.0:</em></p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\((1-E/2)^2\)</td><td>\(E(1-E/2)\)</td><td>\((E/2)^2\)</td></tr><tr><td><strong>1</strong></td><td>\(E/2\)</td><td>\(1-E\)</td><td>\(E/2\)</td></tr><tr><td><strong>2</strong></td><td>\((E/2)^2\)</td><td>\(E(1-E/2)\)</td><td>\((1-E/2)^2\)</td></tr></table><p><em>version1.3</em></p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E-(E/2)^2\)</td><td>\(E\)</td><td>\((E/2)^2\)</td></tr><tr><td><strong>1</strong></td><td>\(E/2\)</td><td>\(1-E\)</td><td>\(E/2\)</td></tr><tr><td><strong>2</strong></td><td>\((E/2)^2\)</td><td>\(E\)</td><td>\(1-E-(E/2)^2\)</td></tr></table><p><em>version1.1</em></p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E\)</td><td>\(E/2\)</td><td>\(E/2\)</td></tr><tr><td><strong>1</strong></td><td>\(E/2\)</td><td>\(1-E\)</td><td>\(E/2\)</td></tr><tr><td><strong>2</strong></td><td>\(E/2\)</td><td>\(E/2\)</td><td>\(1-E\)</td></tr></table><p><em>version0.9</em> (not recommended)</p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E\)</td><td>\(E\)</td><td>\(0\)</td></tr><tr><td><strong>1</strong></td><td>\(E/2\)</td><td>\(1-E\)</td><td>\(E/2\)</td></tr><tr><td><strong>2</strong></td><td>\(0\)</td><td>\(E\)</td><td>\(1-E\)</td></tr></table><p>When <code>Err</code> is a length 3 vector, or if <code>Return = 'vector'</code> these
 are the following probabilities:</p><ul><li><p>hom|hom: an actual homozygote is observed as the other homozygote
  (\(E_1\))</p></li>
<li><p>het|hom: an actual homozygote is observed as heterozygote (\(E_2\))</p></li>
<li><p>hom|het: an actual heterozygote is observed as homozygote (\(E_3\))</p></li>
</ul><p>and Pr(observed genotype (columns) | actual genotype (rows)) is then:</p><table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E_1-E_2\)</td><td>\(E_2\)</td><td>\(E_1\)</td></tr><tr><td><strong>1</strong></td><td>\(E_3\)</td><td>\(1-2E_3\)</td><td>\(E_3\)</td></tr><tr><td><strong>2</strong></td><td>\(E_1\)</td><td>\(E_2\)</td><td>\(1-E_1-E_2\)</td></tr></table><p>When the SNPs are scored via sequencing (e.g. RADseq or DArTseq), the 3rd
 error rate (hom|het) is typically considerably higher than the other two,
 while for SNP arrays it tends to be similar to P(het|hom).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">ErM</span> <span class="op">&lt;-</span> <span class="fu">ErrToM</span><span class="op">(</span>Err <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ErM</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    obs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> act        0        1        2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0 0.950000 0.049375 0.000625</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1 0.025000 0.950000 0.025000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2 0.000625 0.049375 0.950000</span>
<span class="r-in"><span><span class="fu">ErrToM</span><span class="op">(</span><span class="va">ErM</span>, Return <span class="op">=</span> <span class="st">'vector'</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  hom|hom  het|hom  hom|het </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.000625 0.049375 0.025000 </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># use error matrix from Whalen, Gorjanc &amp; Hickey 2018</span></span></span>
<span class="r-in"><span><span class="va">funE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">E</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="va">E</span><span class="op">*</span><span class="fl">3</span><span class="op">/</span><span class="fl">4</span>, <span class="va">E</span><span class="op">/</span><span class="fl">2</span>, <span class="va">E</span><span class="op">/</span><span class="fl">4</span>,</span></span>
<span class="r-in"><span>          <span class="va">E</span><span class="op">/</span><span class="fl">4</span>, <span class="fl">1</span><span class="op">-</span><span class="fl">2</span><span class="op">*</span><span class="va">E</span><span class="op">/</span><span class="fl">4</span>, <span class="va">E</span><span class="op">/</span><span class="fl">4</span>,</span></span>
<span class="r-in"><span>          <span class="va">E</span><span class="op">/</span><span class="fl">4</span>, <span class="va">E</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">1</span><span class="op">-</span><span class="va">E</span><span class="op">*</span><span class="fl">3</span><span class="op">/</span><span class="fl">4</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>          <span class="fl">3</span>,<span class="fl">3</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>  <span class="op">}</span></span></span>
<span class="r-in"><span><span class="fu">ErrToM</span><span class="op">(</span>Err <span class="op">=</span> <span class="fl">0.05</span>, flavour <span class="op">=</span> <span class="va">funE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>        [,1]  [,2]   [,3]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1,] 0.9625 0.025 0.0125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [2,] 0.0125 0.975 0.0125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [3,] 0.0125 0.025 0.9625</span>
<span class="r-in"><span><span class="co"># equivalent to:</span></span></span>
<span class="r-in"><span><span class="fu">ErrToM</span><span class="op">(</span>Err <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span><span class="op">/</span><span class="fl">4</span>, <span class="fl">0.05</span><span class="op">/</span><span class="fl">2</span>, <span class="fl">0.05</span><span class="op">/</span><span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>    obs</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> act      0     1      2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   0 0.9625 0.025 0.0125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   1 0.0125 0.975 0.0125</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   2 0.0125 0.025 0.9625</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

