<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Confidence Probabilities — EstConf • sequoia</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Confidence Probabilities — EstConf" />
<meta property="og:description" content="Estimate confidence probabilities ('backward') and assignment
  error rates ('forward') per category (genotyped/dummy) by repeatedly
  simulating genotype data from a reference pedigree using
  SimGeno, reconstruction a pedigree from this using
  sequoia, and counting the number of mismatches using
  PedCompare." />
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png" />
<meta property="og:image:alt" content="sequoia hexlogo" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href=".././index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_main/book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/performance_stats.html">Stats</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Confidence Probabilities</h1>
    
    <div class="hidden name"><code>EstConf.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Estimate confidence probabilities ('backward') and assignment
  error rates ('forward') per category (genotyped/dummy) by repeatedly
  simulating genotype data from a reference pedigree using
  <code><a href='SimGeno.html'>SimGeno</a></code>, reconstruction a pedigree from this using
  <code><a href='sequoia.html'>sequoia</a></code>, and counting the number of mismatches using
  <code><a href='PedCompare.html'>PedCompare</a></code>.</p>
    </div>

    <pre class="usage"><span class='fu'>EstConf</span><span class='op'>(</span>
  Pedigree <span class='op'>=</span> <span class='cn'>NULL</span>,
  LifeHistData <span class='op'>=</span> <span class='cn'>NULL</span>,
  args.sim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>nSnp <span class='op'>=</span> <span class='fl'>400</span>, SnpError <span class='op'>=</span> <span class='fl'>0.001</span>, ParMis <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.4</span>, <span class='fl'>0.4</span><span class='op'>)</span><span class='op'>)</span>,
  args.seq <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Module <span class='op'>=</span> <span class='st'>"ped"</span>, Err <span class='op'>=</span> <span class='fl'>0.001</span>, Tassign <span class='op'>=</span> <span class='fl'>0.5</span>, CalcLLR <span class='op'>=</span> <span class='cn'>FALSE</span><span class='op'>)</span>,
  nSim <span class='op'>=</span> <span class='fl'>10</span>,
  nCores <span class='op'>=</span> <span class='fl'>1</span>,
  quiet <span class='op'>=</span> <span class='cn'>TRUE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>Pedigree</th>
      <td><p>reference pedigree from which to simulate, dataframe with
columns id-dam-sire. Additional columns are ignored.</p></td>
    </tr>
    <tr>
      <th>LifeHistData</th>
      <td><p>dataframe with id, sex (1=female, 2=male, 3=unknown), and
birth year.</p></td>
    </tr>
    <tr>
      <th>args.sim</th>
      <td><p>list of arguments to pass to <code><a href='SimGeno.html'>SimGeno</a></code>, such as
<code>nSnp</code> (number of SNPs), <code>SnpError</code> (genotyping error rate) and
<code>ParMis</code> (proportion of non-genotyped parents). Set to <code>NULL</code> to
use all default values.</p></td>
    </tr>
    <tr>
      <th>args.seq</th>
      <td><p>list of arguments to pass to <code><a href='sequoia.html'>sequoia</a></code>, such as
<code>Module</code> ('par' or 'ped'), <code>Err</code> (assumed genotyping error rate),
and <code>Complex</code>. May include (part of) <code>SeqList</code>, a list of sequoia
output (i.e. as a list-within-a-list). Set to <code>NULL</code> to use all
default values.</p></td>
    </tr>
    <tr>
      <th>nSim</th>
      <td><p>number of iterations of simulate - reconstruct - compare to
perform, i.e. number of simulated datasets.</p></td>
    </tr>
    <tr>
      <th>nCores</th>
      <td><p>number of computer cores to use. If <code>&gt;1</code>, package
<span class="pkg">parallel</span> is used. Set to NULL to use all but one of the available
cores, as detected by <code><a href='https://rdrr.io/r/parallel/detectCores.html'>parallel::detectCores()</a></code> (using all cores tends
to freeze up your computer).</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>suppress messages. <code>TRUE</code> runs <code>SimGeno</code> and
<code>sequoia</code> quietly, <code>'very'</code> also suppresses other messages and
the iteration counter when <code>nCores=1</code> (there is no iteration counter
when <code>nCores&gt;1</code>).</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list, with elements:</p>
<dt>ConfProb</dt><dd><p>See below</p></dd>
  <dt>PedErrors</dt><dd><p>See below</p></dd>
  <dt>Pedigree.reference</dt><dd><p>the pedigree from which data was simulated</p></dd>
  <dt>LifeHistData</dt><dd><p></p></dd>
  <dt>Pedigree.inferred</dt><dd><p>a list with for each iteration the inferred
    pedigree based on the simulated data</p></dd>
  <dt>SimSNPd</dt><dd><p>a list with for each iteration the IDs of the individuals
    simulated to have been genotyped</p></dd>
  <dt>PedComp.fwd</dt><dd><p><code>Counts</code> from the 'forward' <code>PedCompare</code>,
    from which <code>PedErrors</code> is calculated</p></dd>
  <dt>RunParams</dt><dd><p>a list with the call to <code>EstConf</code>, as well as
  the default parameter values for <code>SimGeno</code>, and <code>sequoia</code>.</p></dd>
  <dt>RunTime</dt><dd><p><code>sequoia</code> runtime per simulation in seconds, as
    measured by <code><a href='https://rdrr.io/r/base/system.time.html'>system.time</a>()['elapsed']</code>.</p></dd>

Dataframe ConfProb has 7 columns:
<dt>id.cat, dam.cat, sire.cat</dt><dd><p>Category of the focal individual, dam, and
  sire, in the pedigree inferred based on the simulated data. Coded as
  G=genotyped, D=dummy, X=none</p></dd>
<dt>dam.conf</dt><dd><p>Probability that the dam is correct, given the categories of
  the assigned dam and sire (ignoring whether or not the sire is correct)</p></dd>
<dt>sire.conf</dt><dd><p>as <code>dam.conf</code>, for the sire</p></dd>
<dt>pair.conf</dt><dd><p>Probability that both dam and sire are correct, given their
  categories</p></dd>
<dt>N</dt><dd><p>Number of individuals per category-combination, across all
  <code>nSim</code> iterations</p></dd>

Array PedErrors has three dimensions:
<dt>class</dt><dd><p><ul>
<li><p><code>FalseNeg</code>(atives): could have been assigned but was not
(individual + parent both genotyped or dummyfiable; P1only in
<code>PedCompare</code>).</p></li>
<li><p><code>FalsePos</code>(itives): no parent in reference pedigree, but
one was assigned based on the simulated data (P2only)</p></li>
<li><p><code>Mismatch</code>: different parents between the pedigrees</p></li>
</ul></p></dd>
<dt>cat</dt><dd><p>Category of individual + parent, as a two-letter code where the
  first letter indicates the focal individual and the second the parent;
  G=Genotyped, D=Dummy, T=Total</p></dd>
<dt>parent</dt><dd><p>dam or sire</p></dd>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The confidence probability is taken as the number of correct
  (matching) assignments, divided by all assignments made in the
  <em>observed</em> (inferred-from-simulated) pedigree. In contrast, the false
  negative &amp; false positive assignment rates are proportions of the number of
  parents in the <em>true</em> (reference) pedigree. Each rate is calculated
  separatedly for dams &amp; sires, and separately for each category
  (<strong>G</strong>enotyped/<strong>D</strong>ummy(fiable)/<strong>X</strong> (none)) of
  individual, parent and co-parent.</p>
<p>This function does not know which individuals in the actual <code>Pedigree</code>
 are genotyped, so the confidence probabilities need to be added to the
 <code>Pedigree</code> as shown in the example at the bottom.</p>
<p>A confidence of \(1\) means all assignments on simulated data were correct for
 that category-combination. It should be interpreted as (and perhaps modified
 to) \(&gt; 1 - 1/N\), where sample size <code>N</code> is given in the last column
 of the <code>ConfProb</code> and <code>PedErrors</code> dataframes in the output. The
 same applies for a false negative/positive rate of \(0\) (i.e. to be
 interpreted as \(&lt; 1/N\)).</p>
    <h2 class="hasAnchor" id="assumptions"><a class="anchor" href="#assumptions"></a>Assumptions</h2>

    

<p>Because the actual true pedigree is (typically) unknown, the provided
  reference pedigree is used as a stand-in and assumed to be the true
  pedigree, with unrelated founders. It is also assumed that the probability
  to be genotyped is equal for all parents; in each iteration, a new random
  set of parents (proportion set by <code>ParMis</code>) is mimicked to be
  non-genotyped. In addition, SNPs are assumed to segregate independently.</p>
    <h2 class="hasAnchor" id="object-size"><a class="anchor" href="#object-size"></a>Object size</h2>

    

<p>The size in Kb of the returned list can become pretty big, as each of the
  inferred pedigrees is included. When running <code>EstConf</code> many times for
  a range of parameter values, it may be prudent to save the required summary
  statistics for each run rather than the full output.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='SimGeno.html'>SimGeno</a>, <a href='sequoia.html'>sequoia</a>, <a href='PedCompare.html'>PedCompare</a></code>.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># \donttest{</span>
<span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>Ped_HSg5</span>, <span class='va'>LH_HSg5</span>, package<span class='op'>=</span><span class='st'>"sequoia"</span><span class='op'>)</span>

<span class='co'>## Example A: parentage assignment only</span>
<span class='va'>conf.A</span> <span class='op'>&lt;-</span> <span class='fu'>EstConf</span><span class='op'>(</span>Pedigree <span class='op'>=</span> <span class='va'>Ped_HSg5</span>, LifeHistData <span class='op'>=</span> <span class='va'>LH_HSg5</span>,
   args.sim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>nSnp <span class='op'>=</span> <span class='fl'>100</span>, SnpError <span class='op'>=</span> <span class='fl'>5e-3</span>, ParMis<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span>,
   args.seq <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Module<span class='op'>=</span><span class='st'>"par"</span>, Err<span class='op'>=</span><span class='fl'>1e-3</span>, Tassign<span class='op'>=</span><span class='fl'>0.5</span><span class='op'>)</span>, nSim <span class='op'>=</span> <span class='fl'>3</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Simulating parentage assignment only ...</span></div><div class='output co'>#&gt; i= 1 	 19:19:06 
#&gt; i= 2 	 19:19:11 
#&gt; i= 3 	 19:19:17 </div><div class='input'>
<span class='co'># parent-pair confidence, per category:</span>
<span class='va'>conf.A</span><span class='op'>$</span><span class='va'>ConfProb</span>
</div><div class='output co'>#&gt;   id.cat dam.cat sire.cat dam.conf sire.conf pair.conf    N
#&gt; 1      G       G        G        1    0.9991    0.9991 1074
#&gt; 2      G       G        X        0        NA        NA  981
#&gt; 3      G       X        G       NA    0.0000        NA  291
#&gt; 4      G       X        X       NA        NA        NA  462</div><div class='input'>
<span class='co'># calculate (correct) assignment rates (ignores co-parent)</span>
<span class='fl'>1</span> <span class='op'>-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>conf.A</span><span class='op'>$</span><span class='va'>PedErrors</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>1</span>,<span class='fl'>3</span><span class='op'>)</span>, <span class='va'>sum</span>, na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt;     parent
#&gt; cat        dam      sire
#&gt;   GG 0.9372347 0.9787446
#&gt;   GD 0.0000000 0.0000000
#&gt;   GT 0.7506371 0.4899128
#&gt;   DG 0.0000000 0.0000000
#&gt;   DD 0.0000000 0.0000000
#&gt;   DT 0.0000000 0.0000000
#&gt;   TT 0.7107639 0.4638889</div><div class='input'>
<span class='co'>## Example B: with sibship clustering, based on sequoia inferred pedigree</span>
<span class='va'>RealGenotypes</span> <span class='op'>&lt;-</span> <span class='fu'><a href='SimGeno.html'>SimGeno</a></span><span class='op'>(</span>Ped <span class='op'>=</span> <span class='va'>Ped_HSg5</span>, nSnp <span class='op'>=</span> <span class='fl'>100</span>,
                         ParMis<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.19</span>,<span class='fl'>0.53</span><span class='op'>)</span>, SnpError <span class='op'>=</span> <span class='fl'>6e-3</span><span class='op'>)</span>
<span class='va'>SeqOUT</span> <span class='op'>&lt;-</span> <span class='fu'><a href='sequoia.html'>sequoia</a></span><span class='op'>(</span>GenoM <span class='op'>=</span> <span class='va'>RealGenotypes</span>,
                  LifeHistData <span class='op'>=</span> <span class='va'>LH_HSg5</span>,
                  Err<span class='op'>=</span><span class='fl'>5e-3</span>, Module<span class='op'>=</span><span class='st'>"ped"</span>,
                  quiet<span class='op'>=</span><span class='cn'>TRUE</span>, Plot<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>

<span class='va'>conf.B</span> <span class='op'>&lt;-</span> <span class='fu'>EstConf</span><span class='op'>(</span>Pedigree <span class='op'>=</span> <span class='va'>SeqOUT</span><span class='op'>$</span><span class='va'>Pedigree</span>,
              LifeHistData <span class='op'>=</span> <span class='va'>LH_HSg5</span>,
               args.sim <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>nSnp <span class='op'>=</span> <span class='fl'>100</span>, SnpError <span class='op'>=</span> <span class='fl'>5e-3</span>,
                               ParMis<span class='op'>=</span><span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>0.2</span>, <span class='fl'>0.5</span><span class='op'>)</span><span class='op'>)</span>,
              args.seq <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span>Err<span class='op'>=</span><span class='fl'>5e-3</span>, Module<span class='op'>=</span><span class='st'>"ped"</span><span class='op'>)</span>,
              nSim <span class='op'>=</span> <span class='fl'>2</span>, nCores<span class='op'>=</span><span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Simulating full pedigree reconstruction ...</span></div><div class='output co'>#&gt; <span class='message'>Using 2 out of 6 cores</span></div><div class='input'><span class='va'>conf.B</span><span class='op'>$</span><span class='va'>ConfProb</span>
</div><div class='output co'>#&gt;    id.cat dam.cat sire.cat dam.conf sire.conf pair.conf   N
#&gt; 1       G       G        G    0.989     0.980     0.979 663
#&gt; 2       G       G        D    1.000     0.914     0.914 748
#&gt; 3       G       G        X    1.000        NA        NA  12
#&gt; 4       G       D        G    0.878     0.928     0.878 180
#&gt; 5       G       D        D    0.934     0.890     0.835 182
#&gt; 6       G       D        X    0.750        NA        NA   4
#&gt; 7       G       X        G       NA     1.000        NA   7
#&gt; 8       G       X        D       NA     1.000        NA   6
#&gt; 9       G       X        X       NA        NA        NA  72
#&gt; 10      D       G        G    1.000     1.000     1.000  34
#&gt; 11      D       G        D    1.000     0.951     0.951  48
#&gt; 12      D       G        X    0.000        NA        NA   7
#&gt; 13      D       D        G    0.778     0.852     0.778  15
#&gt; 14      D       D        D    0.754     0.877     0.754  13
#&gt; 15      D       D        X      NaN        NA        NA   0
#&gt; 16      D       X        G       NA     0.000        NA   1
#&gt; 17      D       X        D       NA     0.000        NA   1
#&gt; 18      D       X        X       NA        NA        NA  30</div><div class='input'>
<span class='va'>Ped.withConf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='getAssignCat.html'>getAssignCat</a></span><span class='op'>(</span>Pedigree <span class='op'>=</span> <span class='va'>SeqOUT</span><span class='op'>$</span><span class='va'>Pedigree</span>,
                             SNPd <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/colnames.html'>rownames</a></span><span class='op'>(</span><span class='va'>RealGenotypes</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>Ped.withConf</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/merge.html'>merge</a></span><span class='op'>(</span><span class='va'>Ped.withConf</span>, <span class='va'>conf.B</span><span class='op'>$</span><span class='va'>ConfProb</span>, all.x<span class='op'>=</span><span class='cn'>TRUE</span>, sort<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>Ped.withConf</span> <span class='op'>&lt;-</span> <span class='va'>Ped.withConf</span><span class='op'>[</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"id"</span>,<span class='st'>"dam"</span>,<span class='st'>"sire"</span>, <span class='st'>"dam.conf"</span>, <span class='st'>"sire.conf"</span>,
                                 <span class='st'>"id.cat"</span>, <span class='st'>"dam.cat"</span>, <span class='st'>"sire.cat"</span><span class='op'>)</span><span class='op'>]</span>
<span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>Ped.withConf</span><span class='op'>[</span><span class='va'>Ped.withConf</span><span class='op'>$</span><span class='va'>dam.cat</span><span class='op'>==</span><span class='st'>"G"</span>, <span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt;        id    dam  sire dam.conf sire.conf id.cat dam.cat sire.cat
#&gt; 33 a01002 a00014 M0019        1     0.914      G       G        D
#&gt; 34 a01001 a00014 M0019        1     0.914      G       G        D
#&gt; 35 b01006 a00013 M0034        1     0.914      G       G        D
#&gt; 36 b01007 a00013 M0034        1     0.914      G       G        D
#&gt; 37 a01008 a00013 M0034        1     0.914      G       G        D
#&gt; 38 b01009 a00008 M0015        1     0.914      G       G        D</div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>Ped.withConf</span><span class='op'>[</span><span class='va'>Ped.withConf</span><span class='op'>$</span><span class='va'>dam.cat</span><span class='op'>==</span><span class='st'>"D"</span>, <span class='op'>]</span><span class='op'>)</span>
</div><div class='output co'>#&gt;         id   dam  sire dam.conf sire.conf id.cat dam.cat sire.cat
#&gt; 406 b02186 F0016 M0035    0.934      0.89      G       D        D
#&gt; 407 b03020 F0013 M0024    0.934      0.89      G       D        D
#&gt; 408 b02092 F0003 M0009    0.934      0.89      G       D        D
#&gt; 409 a05058 F0023 M0011    0.934      0.89      G       D        D
#&gt; 410 b02128 F0017 M0036    0.934      0.89      G       D        D
#&gt; 411 b04097 F0024 M0016    0.934      0.89      G       D        D</div><div class='input'>

<span class='co'>## P(actual FS | inferred as FS) etc.</span>
<span class='va'>PairL</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/list.html'>list</a></span><span class='op'>(</span><span class='op'>)</span>
<span class='kw'>for</span> <span class='op'>(</span><span class='va'>i</span> <span class='kw'>in</span> <span class='fl'>1</span><span class='op'>:</span><span class='fu'><a href='https://rdrr.io/r/base/length.html'>length</a></span><span class='op'>(</span><span class='va'>conf.A</span><span class='op'>$</span><span class='va'>Pedigree.inferred</span><span class='op'>)</span><span class='op'>)</span> <span class='op'>{</span>  <span class='co'># nSim</span>
  <span class='fu'><a href='https://rdrr.io/r/base/cat.html'>cat</a></span><span class='op'>(</span><span class='va'>i</span>, <span class='st'>"\t"</span><span class='op'>)</span>
  <span class='va'>PairL</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span> <span class='op'>&lt;-</span> <span class='fu'><a href='ComparePairs.html'>ComparePairs</a></span><span class='op'>(</span><span class='va'>conf.A</span><span class='op'>$</span><span class='va'>Pedigree.reference</span>,
                             <span class='va'>conf.A</span><span class='op'>$</span><span class='va'>Pedigree.inferred</span><span class='op'>[[</span><span class='va'>i</span><span class='op'>]</span><span class='op'>]</span>,
                             GenBack<span class='op'>=</span><span class='fl'>1</span>, patmat<span class='op'>=</span><span class='cn'>TRUE</span>, ExcludeDummies <span class='op'>=</span> <span class='cn'>TRUE</span>,
                             Return<span class='op'>=</span><span class='st'>"Counts"</span><span class='op'>)</span>
<span class='op'>}</span>
</div><div class='output co'>#&gt; 1 	2 	3 	</div><div class='input'><span class='co'># P(actual relationship (Ped1) | inferred relationship (Ped2))</span>
<span class='va'>PairA</span> <span class='op'>&lt;-</span> <span class='fu'>plyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/plyr/man/laply.html'>laply</a></span><span class='op'>(</span><span class='va'>PairL</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>M</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sweep.html'>sweep</a></span><span class='op'>(</span><span class='va'>M</span>, <span class='fl'>2</span>, <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>colSums</a></span><span class='op'>(</span><span class='va'>M</span><span class='op'>)</span>, <span class='st'>"/"</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>PairRel.prop</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/apply.html'>apply</a></span><span class='op'>(</span><span class='va'>PairA</span>, <span class='fl'>2</span><span class='op'>:</span><span class='fl'>3</span>, <span class='va'>mean</span>, na.rm<span class='op'>=</span><span class='cn'>TRUE</span><span class='op'>)</span>  <span class='co'># mean across simulations</span>
<span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span><span class='op'>(</span><span class='va'>PairRel.prop</span>, <span class='fl'>2</span><span class='op'>)</span>
</div><div class='output co'>#&gt;      Ped2
#&gt; Ped1  M    P O FS  MHS  PHS    U    X
#&gt;   M   1 0.00 0  0 0.00 0.00 0.00 0.00
#&gt;   P   0 0.98 0  0 0.00 0.00 0.00 0.01
#&gt;   FS  0 0.00 0  1 0.30 0.08 0.00 0.00
#&gt;   MHS 0 0.00 0  0 0.69 0.00 0.00 0.00
#&gt;   PHS 0 0.00 0  0 0.00 0.91 0.00 0.01
#&gt;   U   0 0.02 1  0 0.01 0.01 0.99 0.98
#&gt;   X   0 0.00 0  0 0.00 0.00 0.00 0.00</div><div class='input'><span class='co'>#' # or: P(inferred relationship | actual relationship)</span>
<span class='va'>PairA2</span> <span class='op'>&lt;-</span> <span class='fu'>plyr</span><span class='fu'>::</span><span class='fu'><a href='https://rdrr.io/pkg/plyr/man/laply.html'>laply</a></span><span class='op'>(</span><span class='va'>PairL</span>, <span class='kw'>function</span><span class='op'>(</span><span class='va'>M</span><span class='op'>)</span> <span class='fu'><a href='https://rdrr.io/r/base/sweep.html'>sweep</a></span><span class='op'>(</span><span class='va'>M</span>, <span class='fl'>1</span>, <span class='fu'><a href='https://rdrr.io/r/base/colSums.html'>rowSums</a></span><span class='op'>(</span><span class='va'>M</span><span class='op'>)</span>, <span class='st'>"/"</span><span class='op'>)</span><span class='op'>)</span>
<span class='co'># }</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


