<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Estimate genotyping error rate (REMOVED; will be re-implemented) — EstEr • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate genotyping error rate (REMOVED; will be re-implemented) — EstEr"><meta name="description" content="Estimate the genotyping error rates in SNP data, based on a
  pedigree and/or duplicates. Estimates probabilities (observed given
  actual) hom|other hom, het|hom, and hom|het. THESE ARE APPROXIMATE VALUES!"><meta property="og:description" content="Estimate the genotyping error rates in SNP data, based on a
  pedigree and/or duplicates. Estimates probabilities (observed given
  actual) hom|other hom, het|hom, and hom|het. THESE ARE APPROXIMATE VALUES!"><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Estimate genotyping error rate (REMOVED; will be re-implemented)</h1>

      <div class="d-none name"><code>EstEr.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Estimate the genotyping error rates in SNP data, based on a
  pedigree and/or duplicates. Estimates probabilities (observed given
  actual) hom|other hom, het|hom, and hom|het. THESE ARE APPROXIMATE VALUES!</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">EstEr</span><span class="op">(</span></span>
<span>  <span class="va">GenoM</span>,</span>
<span>  <span class="va">Pedigree</span>,</span>
<span>  Duplicates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Er_start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.05</span>, <span class="fl">0.05</span>, <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>  perSNP <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genom">GenoM<a class="anchor" aria-label="anchor" href="#arg-genom"></a></dt>
<dd><p>Genotype matrix</p></dd>


<dt id="arg-pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#arg-pedigree"></a></dt>
<dd><p>data.frame with columns id - dam - sire</p></dd>


<dt id="arg-duplicates">Duplicates<a class="anchor" aria-label="anchor" href="#arg-duplicates"></a></dt>
<dd><p>matrix or data.frame with 2 columns, id1 &amp; id2</p></dd>


<dt id="arg-er-start">Er_start<a class="anchor" aria-label="anchor" href="#arg-er-start"></a></dt>
<dd><p>vector of length 3 with starting values for <code>optim</code>.</p></dd>


<dt id="arg-persnp">perSNP<a class="anchor" aria-label="anchor" href="#arg-persnp"></a></dt>
<dd><p>logical, estimate error rate per SNP. WARNING not very
precise, use only as an approximate indicator! Try on simulated data first,
e.g. with <code><a href="SimGeno.html">SimGeno</a></code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>vector of length 3 with estimated genotyping error rates: the
 probabilities that</p><ul><li><p>hom|hom: an actual homozygote is observed as the other homozygote</p></li>
<li><p>het|hom: an actual homozygote is observed as heterozygote</p></li>
<li><p>hom|het: an actual heterozygote is observed as homozygote</p></li>
</ul><p>These are three independent parameters, that define the genotyping error
 matrix (see <code><a href="ErrToM.html">ErrToM</a></code>) as follows:</p>
<table class="table table"><tr><td></td><td><strong>0</strong></td><td><strong>1</strong></td><td><strong>2</strong></td></tr><tr><td><strong>0</strong></td><td>\(1-E_1-E_2\)</td><td>\(E_2\)</td><td>\(E_1\)</td></tr><tr><td><strong>1</strong></td><td>\(E_3\)</td><td>\(1-2E_3\)</td><td>\(E_3\)</td></tr><tr><td><strong>2</strong></td><td>\(E_1\)</td><td>\(E_2\)</td><td>\(1-E_1-E_2\)</td></tr></table><p>Note that for <code>optim</code> a lower bound of 1e-6 and upper bound of 0.499
 are used; if these values are returned this should be interpreted as
 'inestimably small' and 'inestimably large', respectively. PLEASE DO NOT USE
 THESE VALUES AS INPUT IN SUBSEQUENT ANALYSIS BUT SUBSITUTE BY A SENSIBLE
 VALUE!!</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The result should be interpreted as approximate, ballpark estimates!
The estimated error rates from a pedigree will not be as accurate as from
duplicate samples. Errors in individuals without parents or offspring will
not be counted, and errors in individuals with only few offspring may not be
noted either. Deviation of genotype frequencies among founders from
Hardy-Weinberg equilibrium may wrongly be attributed to genotyping errors.
Last but not least, any pedigree errors will result in higher estimated
genotyping errors.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">GenoX</span> <span class="op">&lt;-</span> <span class="fu"><a href="SimGeno.html">SimGeno</a></span><span class="op">(</span><span class="va">Ped_griffin</span>, nSnp<span class="op">=</span><span class="fl">400</span>, SnpError<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.01</span>,<span class="fl">0.07</span>, <span class="fl">0.1</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                ParMis<span class="op">=</span><span class="fl">0.1</span>, CallRate<span class="op">=</span><span class="fl">0.9</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># EstEr(GenoM=GenoX, Pedigree=Ped_griffin)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

