<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Convert genotype data in various formats to sequoia's
  1-column-per-marker format or Colony's 2-columns-per-marker format."><title>Convert Genotype Data — GenoConvert • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Convert Genotype Data — GenoConvert"><meta property="og:description" content="Convert genotype data in various formats to sequoia's
  1-column-per-marker format or Colony's 2-columns-per-marker format."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo_s.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.7.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href=".././index.html">
    <span class="fas fa fas fa-home"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/sequoia.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Functions</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a>
    <a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    <a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a>
    <a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a>
    <a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Change log</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-help">Help</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-help">
    <a class="dropdown-item" href="../articles/help.html">Need help?</a>
    <a class="dropdown-item" href="../articles/pedigree_consultancy.html">Consultancy</a>
    <a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="nav-link" href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa fas fa-envelope"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/JiscaH/sequoia">
    <span class="fab fa fab fa-github"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca">
    <span class="fab fa fab fa-mastodon"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://twitter.com/HuismanJisca">
    <span class="fab fa fab fa-twitter"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Convert Genotype Data</h1>
      
      <div class="d-none name"><code>GenoConvert.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convert genotype data in various formats to sequoia's
  1-column-per-marker format or Colony's 2-columns-per-marker format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GenoConvert</span><span class="op">(</span></span>
<span>  InData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  InFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  InFormat <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>  OutFile <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  OutFormat <span class="op">=</span> <span class="st">"seq"</span>,</span>
<span>  Missing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-9"</span>, <span class="st">"??"</span>, <span class="st">"?"</span>, <span class="st">"NA"</span>, <span class="st">"NULL"</span>, <span class="st">"-1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span><span class="op">)</span><span class="op">[</span><span class="va">InFormat</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>,</span>
<span>    <span class="st">"ped"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"\t"</span>, <span class="st">","</span>, <span class="st">";"</span><span class="op">)</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  IDcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  FIDcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  FIDsep <span class="op">=</span> <span class="st">"__"</span>,</span>
<span>  dropcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>InData</dt>
<dd><p>dataframe or matrix with genotypes to be converted.</p></dd>


<dt>InFile</dt>
<dd><p>character string with name of genotype file to be converted.</p></dd>


<dt>InFormat</dt>
<dd><p>One of 'single', 'double', 'col', 'ped', 'raw', or 'seq', see
Details.</p></dd>


<dt>OutFile</dt>
<dd><p>character string with name of converted file. If NA, return
matrix with genotypes in console (default); if NULL, write to
'GenoForSequoia.txt' in current working directory.</p></dd>


<dt>OutFormat</dt>
<dd><p>as <code>InFormat</code>; only 'seq' and 'col' are implemented.</p></dd>


<dt>Missing</dt>
<dd><p>vector with symbols interpreted as missing data. '0' is
missing data for InFormats 'col' and 'ped' only.</p></dd>


<dt>sep</dt>
<dd><p>vector with field separator strings that will be tried on
<code>InFile</code>. The <code>OutFile</code> separator uses the
<code><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></code> default, i.e. one blank space.</p></dd>


<dt>header</dt>
<dd><p>a logical value indicating whether the file contains a header
as its first line. If NA (default), set to TRUE for 'raw', and FALSE
otherwise.</p></dd>


<dt>IDcol</dt>
<dd><p>number giving the column with individual IDs; 0 indicates the
rownames (for InData only). If NA (default), set to 2 for InFormat 'raw'
and 'ped', and otherwise to 1 for InFile and 0 (rownames) for InData,
except when InData has a column labeled 'ID'.</p></dd>


<dt>FIDcol</dt>
<dd><p>column with the family IDs, if any are wished to
be used. This is column 1 for InFormat 'raw' and 'seq', but those are by
default not used.</p></dd>


<dt>FIDsep</dt>
<dd><p>string used to paste FID and IID together into a composite-ID
(value passed to <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></code>'s <code>collapse</code>). This joining can
be reversed using <code><a href="PedStripFID.html">PedStripFID</a></code>.</p></dd>


<dt>dropcol</dt>
<dd><p>columns to exclude from the output data, on top of IDcol and
FIDcol (which become rownames). When NA, defaults to columns 3-6 for
InFormat 'raw' and 'seq'. Can also be used to drop some SNPs, see example
below on how to do this for the 2-columns-per-SNP input formats.</p></dd>


<dt>quiet</dt>
<dd><p>suppress messages and warnings.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>A genotype matrix in the specified output format. If 'OutFile' is
  specified, the matrix is written to this file and nothing is returned
  inside R. When converting to 0/1/2 format, 2 is the homozygote for the
  minor allele, and 0 the homozygote for the major allele.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The first two arguments are interchangeable, and can be given
  unnamed. The first argument is assumed to be a file name if it is of class
  'character' and length 1, and to be the genetic data if it is a matrix or
  dataframe.</p>
    </div>
    <div class="section level2">
    <h2 id="input-formats">Input formats<a class="anchor" aria-label="anchor" href="#input-formats"></a></h2>
    

<p>The following formats can be specified by <code>InFormat</code>:</p><dl><dt>seq</dt>
<dd><p>(sequoia) genotypes are coded as 0, 1, 2, missing as \(-9\),
  in 1 column per marker. Column 1 contains IDs, there is no header row.</p></dd>

  <dt>raw</dt>
<dd><p>(PLINK) genotypes are coded as 0, 1, 2, missing as NA, in 1
  column per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored), and there is a header row. This is produced by PLINK's option
  --recodeA</p></dd>

  <dt>ped</dt>
<dd><p>(PLINK) genotypes are coded as A, C, T, G, missing as 0, in 2
  columns per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored).</p></dd>

  <dt>col</dt>
<dd><p>(Colony) genotypes are coded as numeric values, missing as 0, in
  2 columns per marker. Column 1 contains IDs.</p></dd>

  <dt>single</dt>
<dd><p>1 column per marker, otherwise unspecified</p></dd>

  <dt>double</dt>
<dd><p>2 columns per marker, otherwise unspecified</p></dd>

  
</dl><p>For each <code>InFormat</code>, its default values for <code>Missing, header,
 IDcol, FIDcol</code>, and <code>dropcol</code> can be overruled by specifying the
 corresponding input parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="error-messages">Error messages<a class="anchor" aria-label="anchor" href="#error-messages"></a></h2>
    

<p>Occasionally when reading in a file <code>GenoConvert</code> may give an error
  that 'rows have unequal length'. GenoConvert makes use of
  <code><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></code> and <code><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></code>, which is much faster
  than <code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></code> for large datafiles, but also more sensitive
  to unusual line endings, unusual end-of-file characters, or invisible
  characters (spaces or tabs) after the end of some lines. In these cases,
  try to read the data from file using read.table or read.csv, and then use
  <code>GenoConvert</code> on this dataframe or matrix, see example.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CheckGeno.html">CheckGeno</a>, <a href="SnpStats.html">SnpStats</a>, <a href="LHConvert.html">LHConvert</a></code>.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Jisca Huisman, <a href="mailto:jisca.huisman@gmail.com">jisca.huisman@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Requires PLINK installed &amp; in system PATH:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># tinker with window size, window overlap and VIF to get a set of</span></span></span>
<span class="r-in"><span><span class="co"># 400 - 800 markers (100-200 enough for just parentage):</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cmd"</span>, input <span class="op">=</span> <span class="st">"plink --file mydata --indep 50 5 2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cmd"</span>, input <span class="op">=</span> <span class="st">"plink --file mydata --extract plink.prune.in</span></span></span>
<span class="r-in"><span><span class="st">  --recodeA --out PlinkOUT"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span>InFile <span class="op">=</span> <span class="st">"PlinkOUT.raw"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># save time on file conversion next time:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">GenoM</span>, file<span class="op">=</span><span class="st">"Geno_sequoia.txt"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, col.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"Geno_sequoia.txt"</span>, row.names<span class="op">=</span><span class="fl">1</span>, header<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># drop some SNPs, e.g. after a warning of &gt;2 alleles:</span></span></span>
<span class="r-in"><span><span class="va">dropSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">68</span>,<span class="fl">101</span>,<span class="fl">128</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span><span class="va">ColonyFile</span>, InFormat <span class="op">=</span> <span class="st">"col"</span>,</span></span>
<span class="r-in"><span>                     dropcol <span class="op">=</span> <span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">dropSNP</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">*</span><span class="va">dropSNP</span><span class="op">)</span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># circumvent a 'rows have unequal length' error:</span></span></span>
<span class="r-in"><span><span class="va">GenoTmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"mydata.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span>InData<span class="op">=</span><span class="va">GenoTmp</span>, InFormat<span class="op">=</span><span class="st">"single"</span>, IDcol<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

