<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Convert Genotype Data — GenoConvert • sequoia</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Convert Genotype Data — GenoConvert" />
<meta property="og:description" content="Convert genotype data in various formats to sequoia's
  1-column-per-marker format or Colony's 2-columns-per-marker format." />
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png" />
<meta property="og:image:alt" content="sequoia hexlogo" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href=".././index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_main/book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/performance_stats.html">Stats</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Convert Genotype Data</h1>
    
    <div class="hidden name"><code>GenoConvert.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Convert genotype data in various formats to sequoia's
  1-column-per-marker format or Colony's 2-columns-per-marker format.</p>
    </div>

    <pre class="usage"><span class='fu'>GenoConvert</span><span class='op'>(</span>
  InData <span class='op'>=</span> <span class='cn'>NULL</span>,
  InFile <span class='op'>=</span> <span class='cn'>NULL</span>,
  InFormat <span class='op'>=</span> <span class='st'>"raw"</span>,
  OutFile <span class='op'>=</span> <span class='cn'>NA</span>,
  OutFormat <span class='op'>=</span> <span class='st'>"seq"</span>,
  Missing <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"-9"</span>, <span class='st'>"??"</span>, <span class='st'>"?"</span>, <span class='st'>"NA"</span>, <span class='st'>"NULL"</span>, <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"0"</span><span class='op'>)</span><span class='op'>[</span><span class='va'>InFormat</span> <span class='op'>%in%</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"col"</span>, <span class='st'>"ped"</span><span class='op'>)</span><span class='op'>]</span><span class='op'>)</span>,
  sep <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>" "</span>, <span class='st'>"\t"</span>, <span class='st'>","</span>, <span class='st'>";"</span><span class='op'>)</span>,
  header <span class='op'>=</span> <span class='cn'>NA</span>,
  IDcol <span class='op'>=</span> <span class='cn'>NA</span>,
  FIDcol <span class='op'>=</span> <span class='cn'>NA</span>,
  FIDsep <span class='op'>=</span> <span class='st'>"__"</span>,
  dropcol <span class='op'>=</span> <span class='cn'>NA</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>InData</th>
      <td><p>dataframe or matrix with genotypes to be converted.</p></td>
    </tr>
    <tr>
      <th>InFile</th>
      <td><p>character string with name of genotype file to be converted.</p></td>
    </tr>
    <tr>
      <th>InFormat</th>
      <td><p>One of 'single', 'double', 'col', 'ped', 'raw', or 'seq', see
Details.</p></td>
    </tr>
    <tr>
      <th>OutFile</th>
      <td><p>character string with name of converted file. If NA, return
matrix with genotypes in console (default); if NULL, write to
'GenoForSequoia.txt' in current working directory.</p></td>
    </tr>
    <tr>
      <th>OutFormat</th>
      <td><p>as <code>InFormat</code>; only 'seq' and 'col' are implemented.</p></td>
    </tr>
    <tr>
      <th>Missing</th>
      <td><p>vector with symbols interpreted as missing data. '0' is
missing data for InFormats 'col' and 'ped' only.</p></td>
    </tr>
    <tr>
      <th>sep</th>
      <td><p>vector with field separator strings that will be tried on
<code>InFile</code>. The <code>OutFile</code> separator uses the
<code><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></code> default, i.e. one blank space.</p></td>
    </tr>
    <tr>
      <th>header</th>
      <td><p>a logical value indicating whether the file contains a header
as its first line. If NA (default), set to TRUE for 'raw', and FALSE
otherwise.</p></td>
    </tr>
    <tr>
      <th>IDcol</th>
      <td><p>number giving the column with individual IDs; 0 indicates the
rownames (for InData only). If NA (default), set to 2 for InFormat 'raw'
and 'ped', and otherwise to 1 for InFile and 0 (rownames) for InData,
except when InData has a column labeled 'ID'.</p></td>
    </tr>
    <tr>
      <th>FIDcol</th>
      <td><p>column with the family IDs, if any are wished to
be used. This is column 1 for InFormat 'raw' and 'seq', but those are by
default not used.</p></td>
    </tr>
    <tr>
      <th>FIDsep</th>
      <td><p>string used to paste FID and IID together into a composite-ID
(value passed to <code><a href='https://rdrr.io/r/base/paste.html'>paste</a></code>'s <code>collapse</code>). This joining can
be reversed using <code><a href='PedStripFID.html'>PedStripFID</a></code>.</p></td>
    </tr>
    <tr>
      <th>dropcol</th>
      <td><p>columns to exclude from the output data, on top of IDcol and
FIDcol (which become rownames). When NA, defaults to columns 3-6 for
InFormat 'raw' and 'seq'. Can also be used to drop some SNPs, see example
below on how to do this for the 2-columns-per-SNP input formats.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>suppress messages and warnings.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A genotype matrix in the specified output format. If 'OutFile' is
  specified, the matrix is written to this file and nothing is returned
  inside R. When converting to 0/1/2 format, 2 is the homozygote for the
  minor allele, and 0 the homozygote for the major allele.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>The first two arguments are interchangeable, and can be given
  unnamed. The first argument is assumed to be a file name if it is of class
  'character' and length 1, and to be the genetic data if it is a matrix or
  dataframe.</p>
    <h2 class="hasAnchor" id="input-formats"><a class="anchor" href="#input-formats"></a>Input formats</h2>

    

<p>The following formats can be specified by <code>InFormat</code>:</p><dl>
  <dt>seq</dt><dd><p>(sequoia) genotypes are coded as 0, 1, 2, missing as \(-9\),
  in 1 column per marker. Column 1 contains IDs, there is no header row.</p></dd>
  <dt>raw</dt><dd><p>(PLINK) genotypes are coded as 0, 1, 2, missing as NA, in 1
  column per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored), and there is a header row. This is produced by PLINK's option
  --recodeA</p></dd>
  <dt>ped</dt><dd><p>(PLINK) genotypes are coded as A, C, T, G, missing as 0, in 2
  columns per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored).</p></dd>
  <dt>col</dt><dd><p>(Colony) genotypes are coded as numeric values, missing as 0, in
  2 columns per marker. Column 1 contains IDs.</p></dd>
  <dt>single</dt><dd><p>1 column per marker, otherwise unspecified</p></dd>
  <dt>double</dt><dd><p>2 columns per marker, otherwise unspecified</p></dd>
  
</dl><p>For each <code>InFormat</code>, its default values for <code>Missing, header,
 IDcol, FIDcol</code>, and <code>dropcol</code> can be overruled by specifying the
 corresponding input parameters.</p>
    <h2 class="hasAnchor" id="error-messages"><a class="anchor" href="#error-messages"></a>Error messages</h2>

    

<p>Occasionally when reading in a file <code>GenoConvert</code> may give an error
  that 'rows have unequal length'. GenoConvert makes use of
  <code><a href='https://rdrr.io/r/base/readLines.html'>readLines</a></code> and <code><a href='https://rdrr.io/r/base/strsplit.html'>strsplit</a></code>, which is much faster
  than <code><a href='https://rdrr.io/r/utils/read.table.html'>read.table</a></code> for large datafiles, but also more sensitive
  to unusual line endings, unusual end-of-file characters, or invisible
  characters (spaces or tabs) after the end of some lines. In these cases,
  try to read the data from file using read.table or read.csv, and then use
  <code>GenoConvert</code> on this dataframe or matrix, see example.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='CheckGeno.html'>CheckGeno</a>, <a href='SnpStats.html'>SnpStats</a>, <a href='LHConvert.html'>LHConvert</a></code>.</p></div>
    <h2 class="hasAnchor" id="author"><a class="anchor" href="#author"></a>Author</h2>

    <p>Jisca Huisman, <a href='mailto:jisca.huisman@gmail.com'>jisca.huisman@gmail.com</a></p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> <span class='op'>(</span><span class='cn'>FALSE</span><span class='op'>)</span> <span class='op'>{</span>
<span class='co'># Requires PLINK installed &amp; in system PATH:</span>

<span class='co'># tinker with window size, window overlap and VIF to get a set of</span>
<span class='co'># 400 - 800 markers (100-200 enough for just parentage):</span>
<span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span><span class='op'>(</span><span class='st'>"cmd"</span>, input <span class='op'>=</span> <span class='st'>"plink --file mydata --indep 50 5 2"</span><span class='op'>)</span>
<span class='fu'><a href='https://rdrr.io/r/base/system.html'>system</a></span><span class='op'>(</span><span class='st'>"cmd"</span>, input <span class='op'>=</span> <span class='st'>"plink --file mydata --extract plink.prune.in
  --recodeA --out PlinkOUT"</span><span class='op'>)</span>

<span class='va'>GenoM</span> <span class='op'>&lt;-</span> <span class='fu'>GenoConvert</span><span class='op'>(</span>InFile <span class='op'>=</span> <span class='st'>"PlinkOUT.raw"</span><span class='op'>)</span>

<span class='co'># save time on file conversion next time:</span>
<span class='fu'><a href='https://rdrr.io/r/utils/write.table.html'>write.table</a></span><span class='op'>(</span><span class='va'>GenoM</span>, file<span class='op'>=</span><span class='st'>"Geno_sequoia.txt"</span>, quote<span class='op'>=</span><span class='cn'>FALSE</span>, col.names<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>GenoM</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.table</a></span><span class='op'>(</span><span class='st'>"Geno_sequoia.txt"</span>, row.names<span class='op'>=</span><span class='fl'>1</span>, header<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span><span class='op'>)</span>

<span class='co'># drop some SNPs, e.g. after a warning of &gt;2 alleles:</span>
<span class='va'>dropSNP</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>5</span>,<span class='fl'>68</span>,<span class='fl'>101</span>,<span class='fl'>128</span><span class='op'>)</span>
<span class='va'>GenoM</span> <span class='op'>&lt;-</span> <span class='fu'>GenoConvert</span><span class='op'>(</span><span class='va'>ColonyFile</span>, InFormat <span class='op'>=</span> <span class='st'>"col"</span>,
                     dropcol <span class='op'>=</span> <span class='fl'>1</span> <span class='op'>+</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='fl'>2</span><span class='op'>*</span><span class='va'>dropSNP</span><span class='op'>-</span><span class='fl'>1</span>, <span class='fl'>2</span><span class='op'>*</span><span class='va'>dropSNP</span><span class='op'>)</span> <span class='op'>)</span>

<span class='co'># circumvent a 'rows have unequal length' error:</span>
<span class='va'>GenoTmp</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/matrix.html'>as.matrix</a></span><span class='op'>(</span><span class='fu'><a href='https://rdrr.io/r/utils/read.table.html'>read.table</a></span><span class='op'>(</span><span class='st'>"mydata.txt"</span>, header<span class='op'>=</span><span class='cn'>TRUE</span>, row.names<span class='op'>=</span><span class='fl'>1</span><span class='op'>)</span><span class='op'>)</span>
<span class='va'>GenoM</span> <span class='op'>&lt;-</span> <span class='fu'>GenoConvert</span><span class='op'>(</span>InData<span class='op'>=</span><span class='va'>GenoTmp</span>, InFormat<span class='op'>=</span><span class='st'>"single"</span>, IDcol<span class='op'>=</span><span class='fl'>0</span><span class='op'>)</span>
<span class='op'>}</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


