<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Convert Genotype Data — GenoConvert • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Convert Genotype Data — GenoConvert"><meta name="description" content="Convert genotype data in various formats to sequoia's
  1-column-per-marker format, PLINK's ped format, or Colony's
  2-columns-per-marker format."><meta property="og:description" content="Convert genotype data in various formats to sequoia's
  1-column-per-marker format, PLINK's ped format, or Colony's
  2-columns-per-marker format."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Convert Genotype Data</h1>

      <div class="d-none name"><code>GenoConvert.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Convert genotype data in various formats to sequoia's
  1-column-per-marker format, PLINK's ped format, or Colony's
  2-columns-per-marker format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GenoConvert</span><span class="op">(</span></span>
<span>  InData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  InFile <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  InFormat <span class="op">=</span> <span class="st">"raw"</span>,</span>
<span>  OutFile <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  OutFormat <span class="op">=</span> <span class="st">"seq"</span>,</span>
<span>  Missing <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-9"</span>, <span class="st">"NA"</span>, <span class="st">"??"</span>, <span class="st">"?"</span>, <span class="st">"NULL"</span>, <span class="st">"-1"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span><span class="op">)</span><span class="op">[</span><span class="va">InFormat</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col"</span>,</span>
<span>    <span class="st">"ped"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,</span>
<span>  sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">" "</span>, <span class="st">"\t"</span>, <span class="st">","</span>, <span class="st">";"</span><span class="op">)</span>,</span>
<span>  header <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  IDcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  FIDcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  FIDsep <span class="op">=</span> <span class="st">"__"</span>,</span>
<span>  dropcol <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-indata">InData<a class="anchor" aria-label="anchor" href="#arg-indata"></a></dt>
<dd><p>dataframe, matrix or
<code><a href="https://rdrr.io/pkg/adegenet/man/genlight.html" class="external-link">genlight</a></code> object with genotypes to be
converted.</p></dd>


<dt id="arg-infile">InFile<a class="anchor" aria-label="anchor" href="#arg-infile"></a></dt>
<dd><p>character string with name of genotype file to be converted.</p></dd>


<dt id="arg-informat">InFormat<a class="anchor" aria-label="anchor" href="#arg-informat"></a></dt>
<dd><p>One of 'seq' (sequoia), 'ped' (PLINK .ped file), 'col'
(COLONY), 'raw' (PLINK –recodeA), 'vcf' (requires library <code>{vcfR}</code>),
'single' (1 column per SNP), or 'double' (2 columns per SNP); see Details.</p></dd>


<dt id="arg-outfile">OutFile<a class="anchor" aria-label="anchor" href="#arg-outfile"></a></dt>
<dd><p>character string with name of converted file. If NA, return
matrix with genotypes in console (default); if NULL, write to
'GenoForSequoia.txt' in current working directory.</p></dd>


<dt id="arg-outformat">OutFormat<a class="anchor" aria-label="anchor" href="#arg-outformat"></a></dt>
<dd><p>as <code>InFormat</code>; only 'seq', 'col', and 'ped' are
implemented. For 'ped' also a sham .map file is created, so that the file
can be read by PLINK. Only for 'ped' are extensions .ped &amp; .map added to
the specified OutFile filename.</p></dd>


<dt id="arg-missing">Missing<a class="anchor" aria-label="anchor" href="#arg-missing"></a></dt>
<dd><p>vector with symbols interpreted as missing data. '0' is
missing data for <code>InFormat</code>s 'col' and 'ped' only.</p></dd>


<dt id="arg-sep">sep<a class="anchor" aria-label="anchor" href="#arg-sep"></a></dt>
<dd><p>vector with field separator strings that will be tried on
<code>InFile</code>. Ignored if package <span class="pkg">data.table</span> is present or if
<code>InFormat='vcf'</code>. The <code>OutFile</code> separator uses the
<code><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></code> default, i.e. one blank space.</p></dd>


<dt id="arg-header">header<a class="anchor" aria-label="anchor" href="#arg-header"></a></dt>
<dd><p>a logical value indicating whether the file contains a header
as its first line. If NA (default), set to TRUE for 'raw', and FALSE
otherwise.</p></dd>


<dt id="arg-idcol">IDcol<a class="anchor" aria-label="anchor" href="#arg-idcol"></a></dt>
<dd><p>number giving the column with individual IDs; 0 indicates the
rownames (for InData only). If NA (default), set to 2 for InFormat 'raw'
and 'ped', and otherwise to 1 for InFile and 0 (rownames) for InData,
except when InData has a column labeled 'ID'.</p></dd>


<dt id="arg-fidcol">FIDcol<a class="anchor" aria-label="anchor" href="#arg-fidcol"></a></dt>
<dd><p>column with the family IDs, if any are wished to
be used. This is column 1 for InFormat 'raw' and 'seq', but those are by
default not used.</p></dd>


<dt id="arg-fidsep">FIDsep<a class="anchor" aria-label="anchor" href="#arg-fidsep"></a></dt>
<dd><p>string used to paste FID and IID together into a composite-ID
(value passed to <code><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></code>'s <code>collapse</code>). This joining can
be reversed using <code><a href="PedStripFID.html">PedStripFID</a></code>.</p></dd>


<dt id="arg-dropcol">dropcol<a class="anchor" aria-label="anchor" href="#arg-dropcol"></a></dt>
<dd><p>columns to exclude from the output data, on top of IDcol and
FIDcol (which become rownames). When NA, defaults to columns 3-6 for
InFormat 'raw' and 'seq'. Can also be used to drop some SNPs, see example
below on how to do this for the 2-columns-per-SNP input formats.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>suppress messages and warnings.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A genotype matrix in the specified output format; the default sequoia
  format ('seq') has 1 column per SNP coded in 0/1/2 format (major homozygote
  /heterozygote /minor homozygote) with -9 for missing values, sample IDs in
  row names and SNP names in column names. If 'OutFile' is specified, the
  matrix is written to this file and nothing is returned inside R.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The first two arguments are interchangeable, and can be given
  unnamed. The first argument is assumed to be a file name if it is of class
  'character' and length 1, and to be the genetic data if it is a matrix or
  dataframe.</p>
<p>If package <span class="pkg">data.table</span> is detected, <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></code>
  is used to read in the data from file. Otherwise, a combination of
  <code><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></code> and <code><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></code> is used.</p>
    </div>
    <div class="section level2">
    <h2 id="input-formats">Input formats<a class="anchor" aria-label="anchor" href="#input-formats"></a></h2>


<p>The following formats can be specified by <code>InFormat</code>:</p><dl><dt>seq</dt>
<dd><p>(sequoia) genotypes are coded as 0, 1, 2, missing as \(-9\) (in input
 any negative number or NA are OK),
  in 1 column per marker. Column 1 contains IDs, there is no header row.</p></dd>

  <dt>ped</dt>
<dd><p>(PLINK) genotypes are coded as A, C, T, G, missing as 0, in 2
  columns per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored). If an associated .map file exists, SNP names will be read from
  there.</p></dd>

  <dt>raw</dt>
<dd><p>(PLINK) genotypes are coded as 0, 1, 2, missing as NA, in 1
  column per marker. The first 6 columns are descriptive (1:FID, 2:IID, 3 to
  6 ignored), and there is a header row. This is produced by PLINK's option
  –recodeA</p></dd>

  <dt>col</dt>
<dd><p>(Colony) genotypes are coded as numeric values, missing as 0, in
  2 columns per marker. Column 1 contains IDs.</p></dd>

  <dt>vcf</dt>
<dd><p>(VCF) genotypes are coded as '0/0','0/1','1/1', variable number
    of header rows followed by 1 row per SNP, with various columns of
    metadata followed by 1 column per individual. Requires package
    <span class="pkg">vcfR</span>.</p></dd>

  <dt>single</dt>
<dd><p>1 column per marker, otherwise unspecified</p></dd>

  <dt>double</dt>
<dd><p>2 columns per marker, otherwise unspecified</p></dd>


</dl><p>For each <code>InFormat</code>, its default values for <code>Missing, header,
 IDcol, FIDcol</code>, and <code>dropcol</code> can be overruled by specifying the
 corresponding input parameters.</p>
    </div>
    <div class="section level2">
    <h2 id="error-messages">Error messages<a class="anchor" aria-label="anchor" href="#error-messages"></a></h2>


<p>Occasionally when reading in a file <code>GenoConvert</code> may give an error
  that 'rows have unequal length'. <code>GenoConvert</code> makes use of
  <code><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></code> and <code><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></code>, which is much faster
  than <code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></code> for large datafiles, but also more sensitive
  to unusual line endings, unusual end-of-file characters, or invisible
  characters (spaces or tabs) after the end of some lines. In these cases,
  try to read the data from file using read.table or read.csv, and then use
  <code>GenoConvert</code> on this dataframe or matrix, see example.</p>
<p>Any warnings generated by <code><a href="CheckGeno.html">CheckGeno</a></code> regarding SNPs scored for
 few individuals and/or individuals scored for few SNPs etc. are only for
 your information; none are excluded from <code>GenoConvert</code>'s output, but
 these SNPs and/or individuals will be excluded during pre-processing of the
 data in any of the other functions in this package.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="CheckGeno.html">CheckGeno</a>, <a href="SnpStats.html">SnpStats</a>, <a href="LHConvert.html">LHConvert</a></code>.</p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Jisca Huisman, <a href="mailto:jisca.huisman@gmail.com">jisca.huisman@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Requires PLINK installed &amp; in system PATH:</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># tinker with window size, window overlap and VIF to get a set of</span></span></span>
<span class="r-in"><span><span class="co"># 400 - 800 markers (100-200 enough for just parentage):</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cmd"</span>, input <span class="op">=</span> <span class="st">"plink --file mydata --indep 50 5 2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"cmd"</span>, input <span class="op">=</span> <span class="st">"plink --file mydata --extract plink.prune.in</span></span></span>
<span class="r-in"><span><span class="st">  --recodeA --out PlinkOUT"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span>InFile <span class="op">=</span> <span class="st">"PlinkOUT.raw"</span>, InFormat<span class="op">=</span><span class="st">'raw'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># which is the same as</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span><span class="va">PlinkOUT.raw</span>, InFormat<span class="op">=</span><span class="st">'single'</span>,</span></span>
<span class="r-in"><span>                    IDcol<span class="op">=</span><span class="fl">2</span>, dropcol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># (but it will complain that InFormat='single' is not consistent with .raw</span></span></span>
<span class="r-in"><span><span class="co"># file extension)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># save time on file conversion next time:</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">GenoM</span>, file<span class="op">=</span><span class="st">"Geno_sequoia.txt"</span>, quote<span class="op">=</span><span class="cn">FALSE</span>, col.names<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"Geno_sequoia.txt"</span>, row.names<span class="op">=</span><span class="fl">1</span>, header<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># drop some SNPs, e.g. after a warning of &gt;2 alleles:</span></span></span>
<span class="r-in"><span><span class="va">dropSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">68</span>,<span class="fl">101</span>,<span class="fl">128</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span><span class="va">ColonyFile</span>, InFormat <span class="op">=</span> <span class="st">"col"</span>,</span></span>
<span class="r-in"><span>                     dropcol <span class="op">=</span> <span class="fl">1</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">dropSNP</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">*</span><span class="va">dropSNP</span><span class="op">)</span> <span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># circumvent a 'rows have unequal length' error:</span></span></span>
<span class="r-in"><span><span class="va">GenoTmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"mydata.txt"</span>, header<span class="op">=</span><span class="cn">TRUE</span>, row.names<span class="op">=</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">GenoM</span> <span class="op">&lt;-</span> <span class="fu">GenoConvert</span><span class="op">(</span>InData<span class="op">=</span><span class="va">GenoTmp</span>, InFormat<span class="op">=</span><span class="st">"single"</span>, IDcol<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># can also write to file, e.g. simulated genotypes:</span></span></span>
<span class="r-in"><span><span class="fu">GenoConvert</span><span class="op">(</span><span class="va">Geno_A</span>, InFormat<span class="op">=</span><span class="st">'seq'</span>, OutFormat<span class="op">=</span><span class="st">'ped'</span>, OutFile <span class="op">=</span> <span class="va">sim_genotypes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

