<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Find Putative Relatives — GetMaybeRel • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find Putative Relatives — GetMaybeRel"><meta name="description" content="Identify pairs of individuals likely to be related, but not
  assigned as such in the provided pedigree."><meta property="og:description" content="Identify pairs of individuals likely to be related, but not
  assigned as such in the provided pedigree."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Find Putative Relatives</h1>

      <div class="d-none name"><code>GetMaybeRel.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Identify pairs of individuals likely to be related, but not
  assigned as such in the provided pedigree.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">GetMaybeRel</span><span class="op">(</span></span>
<span>  GenoM <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  SeqList <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Pedigree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LifeHistData <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  AgePrior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Module <span class="op">=</span> <span class="st">"par"</span>,</span>
<span>  Complex <span class="op">=</span> <span class="st">"full"</span>,</span>
<span>  Herm <span class="op">=</span> <span class="st">"no"</span>,</span>
<span>  Err <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  ErrFlavour <span class="op">=</span> <span class="st">"version2.9"</span>,</span>
<span>  Tassign <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  Tfilter <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>,</span>
<span>  MaxPairs <span class="op">=</span> <span class="fl">7</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">GenoM</span><span class="op">)</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genom">GenoM<a class="anchor" aria-label="anchor" href="#arg-genom"></a></dt>
<dd><p>numeric matrix with genotype data: One row per individual,
one column per SNP, coded as 0, 1, 2, missing values as a negative number
or NA. You can reformat data with <code><a href="GenoConvert.html">GenoConvert</a></code>, or use other
packages to get it into a genlight object and then use <code>as.matrix</code>.</p></dd>


<dt id="arg-seqlist">SeqList<a class="anchor" aria-label="anchor" href="#arg-seqlist"></a></dt>
<dd><p>list with output from <code><a href="sequoia.html">sequoia</a></code>.
<code>SeqList$Pedigree</code> is used if present, and <code>SeqList$PedigreePar</code>
otherwise, and overrides the input parameter <code>Pedigree</code>. If 'Specs' is
present, its elements override all input parameters with the same name. The
list elements  `LifeHist', `AgePriors', and `ErrM' are also used if
present, and similarly override the corresponding input parameters.</p></dd>


<dt id="arg-pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#arg-pedigree"></a></dt>
<dd><p>dataframe with id - dam - sire in columns 1-3. May include
non-genotyped individuals, which will be treated as dummy individuals. When
provided, all likelihoods (and thus all maybe-relatives) are conditional on
this pedigree. Note: <code>SeqList$Pedigree</code> or <code>SeqList$PedigreePar</code> take
precedent (for this function only).</p></dd>


<dt id="arg-lifehistdata">LifeHistData<a class="anchor" aria-label="anchor" href="#arg-lifehistdata"></a></dt>
<dd><p>data.frame with up to 6 columns:</p><dl><dt>ID</dt>
<dd><p>max. 30 characters long</p></dd>

 <dt>Sex</dt>
<dd><p>1 = female, 2 = male, 3 = unknown, 4 = hermaphrodite,
           other numbers or NA = unknown</p></dd>

<dt>BirthYear </dt>
<dd><p>birth or hatching year, integer, with missing values as NA
  or any negative number.</p></dd>

<dt>BY.min</dt>
<dd><p>minimum birth year, only used if BirthYear is missing</p></dd>

<dt>BY.max</dt>
<dd><p>maximum birth year, only used if BirthYear is missing</p></dd>

<dt>Year.last</dt>
<dd><p>Last year in which individual could have had offspring. Can
  e.g. in mammals be the year before death for females, and year after death
  for males.</p></dd>

</dl><p>"Birth year" may be in any arbitrary discrete time unit relevant to the
species (day, month, decade), as long as parents are never born in the same
time unit as their offspring, and only integers are used. Individuals do not
need to be in the same order as in `GenoM', nor do all genotyped individuals
need to be included.</p></dd>


<dt id="arg-ageprior">AgePrior<a class="anchor" aria-label="anchor" href="#arg-ageprior"></a></dt>
<dd><p>Agepriors matrix, as generated by <code><a href="MakeAgePrior.html">MakeAgePrior</a></code>
and included in the <code><a href="sequoia.html">sequoia</a></code> output. Affects which
relationships are considered possible (only those where \(P(A|R) / P(A)
&gt; 0\)).</p></dd>


<dt id="arg-module">Module<a class="anchor" aria-label="anchor" href="#arg-module"></a></dt>
<dd><p>type of relatives to check for. One of</p><dl><dt>par</dt>
<dd><p>parent - offspring pairs</p></dd>

  <dt>ped</dt>
<dd><p>all first and second degree relatives</p></dd>


</dl><p>When 'par', all pairs are returned that are more likely parent-offspring
than unrelated, potentially including pairs that are even more likely to
be otherwise related.</p></dd>


<dt id="arg-complex">Complex<a class="anchor" aria-label="anchor" href="#arg-complex"></a></dt>
<dd><p>Breeding system complexity. Either "full" (default), "simp"
(simplified, no explicit consideration of inbred relationships), "mono"
(monogamous).</p></dd>


<dt id="arg-herm">Herm<a class="anchor" aria-label="anchor" href="#arg-herm"></a></dt>
<dd><p>Hermaphrodites, either "no", "A" (distinguish between dam and
sire role, default if at least 1 individual with sex=4), or "B" (no
distinction between dam and sire role). Both of the latter deal with
selfing.</p></dd>


<dt id="arg-err">Err<a class="anchor" aria-label="anchor" href="#arg-err"></a></dt>
<dd><p>assumed per-locus genotyping error rate, as a single number, or a
length 3 vector with P(hom|hom), P(het|hom), P(hom|het), or a 3x3 matrix.
See details below. The error rate is presumed constant across SNPs, and
missingness is presumed random with respect to actual genotype. Using
<code>Err</code> &gt;5% is not recommended, and <code>Err</code> &gt;10% strongly
discouraged. See <code><a href="Err_RADseq.html">Err_RADseq</a></code> to convert per-allele rates at
homozygous and heterozygous sites to the required length-3 vector, and
<code><a href="ErrToM.html">ErrToM</a></code> for further genotyping error details.</p></dd>


<dt id="arg-errflavour">ErrFlavour<a class="anchor" aria-label="anchor" href="#arg-errflavour"></a></dt>
<dd><p>function that takes <code>Err</code> (single number) as input,
and returns a length 3 vector or 3x3 matrix, or choose from inbuilt options
'version2.9', 'version2.0', 'version1.3', or 'version1.1', referring to the
sequoia version in which they were the default. Ignored if <code>Err</code> is a
vector or matrix.</p></dd>


<dt id="arg-tassign">Tassign<a class="anchor" aria-label="anchor" href="#arg-tassign"></a></dt>
<dd><p>minimum LLR required for acceptance of proposed relationship,
relative to next most likely relationship. Higher values result in more
conservative assignments. Must be zero or positive.</p></dd>


<dt id="arg-tfilter">Tfilter<a class="anchor" aria-label="anchor" href="#arg-tfilter"></a></dt>
<dd><p>threshold log10-likelihood ratio (LLR) between a proposed
relationship versus unrelated, to select candidate relatives. Typically a
negative value, related to the fact that unconditional likelihoods are
calculated during the filtering steps. More negative values may decrease
non-assignment, but will increase computational time.</p></dd>


<dt id="arg-maxpairs">MaxPairs<a class="anchor" aria-label="anchor" href="#arg-maxpairs"></a></dt>
<dd><p>the maximum number of putative pairs to return.</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>logical, suppress messages.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list with</p>
<dl><dt>MaybePar</dt>
<dd><p>A dataframe with non-assigned likely parent-offspring pairs,
 with columns:</p><ul><li><p>ID1</p></li>
<li><p>ID2</p></li>
<li><p>TopRel: the most likely relationship, using abbreviations listed
        below</p></li>
<li><p>LLR: Log10-Likelihood Ratio between most likely and next most
       likely relationship</p></li>
<li><p>OH: Number of loci at which the two individuals are opposite
       homozygotes</p></li>
<li><p>BirthYear1: Birth year of ID1 (copied from LifeHistData)</p></li>
<li><p>BirthYear2</p></li>
<li><p>AgeDif: Age difference; BirthYear1 - BirthYear2</p></li>
<li><p>Sex1: Sex of ID1 (copied from LifeHistData)</p></li>
<li><p>Sex2</p></li>
<li><p>SnpdBoth: Number of loci at which the two individuals are both
       successfully genotyped</p></li>
</ul></dd>

<dt>MaybeRel</dt>
<dd><p>A dataframe with non-assigned likely pairs of relatives,
with columns identical to <code>MaybePar</code></p></dd>

<dt>MaybeTrio</dt>
<dd><p>A dataframe with non-assigned parent-parent-offspring
    trios, with columns:</p><ul><li><p>ID</p></li>
<li><p>parent1</p></li>
<li><p>parent2</p></li>
<li><p>TopRel: the most likely relationship, using abbreviations listed
        below</p></li>
<li><p>LLRparent1: Log10-Likelihood Ratio between parent1 being a
       parent of ID vs the next most likely relationship between the pair,
       ignoring parent2</p></li>
<li><p>LLRparent2: as LLRparent1</p></li>
<li><p>LLRpair: LLR for the parental pair, versus the next most likely
       configuration between the three individuals (with one or neither
       parent assigned)</p></li>
<li><p>OHparent1: Number of loci at which ID and parent1 are opposite
       homozygotes</p></li>
<li><p>OHparent2: as OHparent1</p></li>
<li><p>MEpair: Number of Mendelian errors between the offspring and the
   parent pair, includes OH as well as e.g. parents being opposing
   homozygotes, but the offspring not being a heterozygote. The offspring
   being OH with both parents is counted as 2 errors.</p></li>
<li><p>SNPd.id.parent1: Number of loci at which ID and parent1 are both
       successfully genotyped</p></li>
<li><p>SNPd.id.parent2: as SNPd.id.parent1</p></li>
</ul></dd>

</dl><p>The following categories are used in column 'TopRel', indicating the most
likely relationship category:</p>
<dl><dt>PO</dt>
<dd><p>Parent-Offspring</p></dd>

<dt>FS</dt>
<dd><p>Full Siblings</p></dd>

<dt>HS</dt>
<dd><p>Half Siblings</p></dd>

<dt>GP</dt>
<dd><p>GrandParent - grand-offspring</p></dd>

<dt>FA</dt>
<dd><p>Full Avuncular (aunt/uncle)</p></dd>

<dt>2nd</dt>
<dd><p>2nd degree relatives, not enough information to distinguish
  between HS,GP and FA</p></dd>

<dt>Q</dt>
<dd><p>Unclear, but probably 1st, 2nd or 3rd degree relatives</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>When <code>Module="par"</code>, the age difference of the putative pair is
  temporarily set to NA so that genetic parent-offspring pairs declared to be
  born in the same year may be discovered. When <code>Module="ped"</code>, only
  relationships possible given the age difference, if known from the
  LifeHistData, are considered.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="sequoia.html">sequoia</a></code> to identify likely pairs of duplicate
  genotypes and for pedigree reconstruction; <code><a href="GetRelM.html">GetRelM</a></code> to
  identify all pairs of relatives in a pedigree; <code><a href="CalcPairLL.html">CalcPairLL</a></code> for
  the likelihoods underlying the LLR.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># without conditioning on pedigree</span></span></span>
<span class="r-in"><span><span class="va">MaybeRel_griffin</span> <span class="op">&lt;-</span> <span class="fu">GetMaybeRel</span><span class="op">(</span>GenoM<span class="op">=</span><span class="va">Geno_griffin</span>, Err<span class="op">=</span><span class="fl">0.001</span>, Module<span class="op">=</span><span class="st">'par'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">MaybeRel_griffin</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "MaybePar"  "MaybeTrio"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># conditioning on pedigree</span></span></span>
<span class="r-in"><span><span class="va">MaybePO</span> <span class="op">&lt;-</span> <span class="fu">GetMaybeRel</span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno_griffin</span>, SeqList <span class="op">=</span> <span class="va">SeqOUT_griffin</span>,</span></span>
<span class="r-in"><span>                      Module <span class="op">=</span> <span class="st">'par'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Searching for non-assigned parent-offspring pairs ... (Module = par)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> using Pedigree in SeqList</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> using LifeHist in SeqList</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> using AgePriors in SeqList</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Genotype matrix looks OK! There are  142  individuals and  400  SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 167 individuals, 122 dams and 116 sires</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> settings in SeqList$Specs will overrule input parameters</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Counting opposing homozygous loci between all individuals ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Checking for non-assigned Parent-Offspring pairs ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0   10  20  30  40  50  60  70  80  90  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  |   |   |   |   |   |   |   |   |   |   |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ****************************************</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Found 0 likely parent-offspring pairs, and 0, other non-assigned pairs of possible relatives</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">MaybePO</span><span class="op">$</span><span class="va">MaybePar</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># instead of providing the entire SeqList, one may specify the relevant</span></span></span>
<span class="r-in"><span><span class="co"># elements separately</span></span></span>
<span class="r-in"><span><span class="va">Maybe</span> <span class="op">&lt;-</span> <span class="fu">GetMaybeRel</span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno_griffin</span>,</span></span>
<span class="r-in"><span>                     Pedigree <span class="op">=</span> <span class="va">SeqOUT_griffin</span><span class="op">$</span><span class="va">PedigreePar</span>,</span></span>
<span class="r-in"><span>                     LifeHistData <span class="op">=</span> <span class="va">LH_griffin</span>,</span></span>
<span class="r-in"><span>                     Err<span class="op">=</span><span class="fl">0.0001</span>, Complex <span class="op">=</span> <span class="st">"full"</span>,</span></span>
<span class="r-in"><span>                     Module <span class="op">=</span> <span class="st">"ped"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Searching for non-assigned relative pairs ... (Module = ped)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Genotype matrix looks OK! There are  142  individuals and  400  SNPs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Conditioning on pedigree with 142 individuals, 65 dams and 79 sires</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Ageprior: Pedigree-based, overlapping generations, smoothed, MaxAgeParent = 5,5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Transferring input pedigree ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Counting opposing homozygous loci between all individuals ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Checking for non-assigned relatives ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  0   10  20  30  40  50  60  70  80  90  100% </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  |   |   |   |   |   |   |   |   |   |   |</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ****************************************</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Found 0 likely parent-offspring pairs, and 129, other non-assigned pairs of possible relatives</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">Maybe</span><span class="op">$</span><span class="va">MaybeRel</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           ID1         ID2 TopRel   LLR OH BirthYear1 BirthYear2 AgeDif Sex1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 i081_2005_F i083_2005_M     FS  5.42 16       2005       2005      0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 i165_2009_F i175_2009_M     FS  5.34 17       2009       2009      0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 i131_2007_F i133_2007_F     FS  5.09  8       2007       2007      0    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 i160_2008_F i130_2007_F     FS  2.25 11       2008       2007      1    1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 i158_2008_M i133_2007_F     HS 17.36 10       2008       2007      1    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 i158_2008_M i130_2007_F     HS 16.15 21       2008       2007      1    2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Sex2 SNPdBoth</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    2      392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    2      392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    1      392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    1      392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    1      392</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    1      392</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># visualise results, turn dataframe into matrix first:</span></span></span>
<span class="r-in"><span><span class="va">MaybeM</span> <span class="op">&lt;-</span> <span class="fu"><a href="GetRelM.html">GetRelM</a></span><span class="op">(</span>Pairs <span class="op">=</span> <span class="va">Maybe</span><span class="op">$</span><span class="va">MaybeRel</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">MaybeM</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="GetMaybeRel-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co"># or combine with pedigree (note suffix '?')</span></span></span>
<span class="r-in"><span><span class="va">RelM</span> <span class="op">&lt;-</span> <span class="fu"><a href="GetRelM.html">GetRelM</a></span><span class="op">(</span>Pedigree <span class="op">=</span><span class="va">SeqOUT_griffin</span><span class="op">$</span><span class="va">PedigreePar</span>, Pairs <span class="op">=</span> <span class="va">Maybe</span><span class="op">$</span><span class="va">MaybeRel</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="PlotRelPairs.html">PlotRelPairs</a></span><span class="op">(</span><span class="va">RelM</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="GetMaybeRel-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

