<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Find Putative Relatives — GetMaybeRel • sequoia</title>

<!-- favicons -->
<link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png" />
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png" />
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png" />
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png" />

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Find Putative Relatives — GetMaybeRel" />
<meta property="og:description" content="Identify pairs of individuals likely to be related, but not
  assigned as such in the provided pedigree." />
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png" />
<meta property="og:image:alt" content="sequoia hexlogo" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href=".././index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/vignette_main/book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/performance_stats.html">Stats</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find Putative Relatives</h1>
    
    <div class="hidden name"><code>GetMaybeRel.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Identify pairs of individuals likely to be related, but not
  assigned as such in the provided pedigree.</p>
    </div>

    <pre class="usage"><span class='fu'>GetMaybeRel</span><span class='op'>(</span>
  GenoM <span class='op'>=</span> <span class='cn'>NULL</span>,
  SeqList <span class='op'>=</span> <span class='cn'>NULL</span>,
  Pedigree <span class='op'>=</span> <span class='cn'>NULL</span>,
  LifeHistData <span class='op'>=</span> <span class='cn'>NULL</span>,
  AgePrior <span class='op'>=</span> <span class='cn'>NULL</span>,
  ParSib <span class='op'>=</span> <span class='cn'>NULL</span>,
  Module <span class='op'>=</span> <span class='st'>"par"</span>,
  Complex <span class='op'>=</span> <span class='st'>"full"</span>,
  Herm <span class='op'>=</span> <span class='st'>"no"</span>,
  Err <span class='op'>=</span> <span class='fl'>1e-04</span>,
  ErrFlavour <span class='op'>=</span> <span class='st'>"version2.0"</span>,
  MaxMismatch <span class='op'>=</span> <span class='cn'>NA</span>,
  Tassign <span class='op'>=</span> <span class='fl'>0.5</span>,
  Tfilter <span class='op'>=</span> <span class='op'>-</span><span class='fl'>2</span>,
  MaxPairs <span class='op'>=</span> <span class='fl'>7</span> <span class='op'>*</span> <span class='fu'><a href='https://rdrr.io/r/base/nrow.html'>nrow</a></span><span class='op'>(</span><span class='va'>GenoM</span><span class='op'>)</span>,
  quiet <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>GenoM</th>
      <td><p>numeric matrix with genotype data: One row per individual, and
one column per SNP, coded as 0, 1, 2 or -9 (missing). See also
<code><a href='GenoConvert.html'>GenoConvert</a></code>.</p></td>
    </tr>
    <tr>
      <th>SeqList</th>
      <td><p>list with output from <code><a href='sequoia.html'>sequoia</a></code>.
<code>SeqList$Pedigree</code> is used if present, and <code>SeqList$PedigreePar</code>
otherwise, and overrides the input parameter <code>Pedigree</code>. If 'Specs' is
present, its elements override all input parameters with the same name. The
list elements  `LifeHist', `AgePriors', and `ErrM' are also used if
present, and similarly override the corresponding input parameters.</p></td>
    </tr>
    <tr>
      <th>Pedigree</th>
      <td><p>dataframe with id - dam - sire in columns 1-3. May include
non-genotyped individuals, which will be treated as dummy individuals. When
provided, all likelihoods (and thus all maybe-relatives) are conditional on
this pedigree. Note: <code>SeqList$Pedigree</code> or <code>SeqList$PedigreePar</code> take
precedent (for this function only).</p></td>
    </tr>
    <tr>
      <th>LifeHistData</th>
      <td><p>dataframe with 3 columns (optionally 5):</p><dl>
 <dt>ID</dt><dd><p>max. 30 characters long</p></dd>
 <dt>Sex</dt><dd><p>1 = female, 2 = male, 3 = unknown, 4 = hermaphrodite,
           other numbers or NA = unknown</p></dd>
<dt>BirthYear </dt><dd><p>birth or hatching year, integer, with missing values as NA
  or any negative value.</p></dd>
<dt>BY.min</dt><dd><p>minimum birth year, only used if BirthYear is missing</p></dd>
<dt>BY.max</dt><dd><p>maximum birth year, only used if BirthYear is missing</p></dd> 
</dl><p>If the species has multiple generations per year, use an integer coding such
that the candidate parents' `Birth year' is at least one smaller than their
putative offspring's. Column names are ignored, so ensure column order is ID
- sex - birth year (- BY.min - BY.max). Individuals do not need to be in the
same order as in `GenoM', nor do all genotyped individuals need to be
included.</p></td>
    </tr>
    <tr>
      <th>AgePrior</th>
      <td><p>Agepriors matrix, as generated by <code><a href='MakeAgePrior.html'>MakeAgePrior</a></code>
and included in the <code><a href='sequoia.html'>sequoia</a></code> output. Affects which
relationships are considered possible (only those where \(P(A|R) / P(A)
&gt; 0\)).</p></td>
    </tr>
    <tr>
      <th>ParSib</th>
      <td><p>either 'par' to check for putative parent-offspring pairs only,
or 'sib' to check for all types of first and second degree relatives. This
argument will be deprecated, please use <code>Module</code>.</p></td>
    </tr>
    <tr>
      <th>Module</th>
      <td><p>type of relatives to check for. One of</p><dl>
  <dt>par</dt><dd><p>parent - offspring pairs</p></dd>
  <dt>ped</dt><dd><p>all first and second degree relatives</p></dd>

</dl><p>When 'par', all pairs are returned that are more likely parent-offspring
than unrelated, potentially including pairs that are even more likely to
be otherwise related.</p></td>
    </tr>
    <tr>
      <th>Complex</th>
      <td><p>Breeding system complexity. Either "full" (default), "simp"
(simplified, no explicit consideration of inbred relationships), "mono"
(monogamous).</p></td>
    </tr>
    <tr>
      <th>Herm</th>
      <td><p>Hermaphrodites, either "no", "A" (distinguish between dam and
sire role, default if at least 1 individual with sex=4), or "B" (no
distinction between dam and sire role). Both of the latter deal with
selfing.</p></td>
    </tr>
    <tr>
      <th>Err</th>
      <td><p>estimated genotyping error rate, as a single number or 3x3 matrix.
Details below. The error rate is presumed constant across SNPs, and
missingness is presumed random with respect to actual genotype.</p></td>
    </tr>
    <tr>
      <th>ErrFlavour</th>
      <td><p>function that takes <code>Err</code> (single number) as input,
and returns a 3x3 matrix of observed (columns) conditional on actual (rows)
genotypes, or choose from inbuilt options 'version2.0', 'version1.3', or
'version1.1', referring to the sequoia version in which they were the
default. Ignored if <code>Err</code> is a matrix. See <code><a href='ErrToM.html'>ErrToM</a></code>.</p></td>
    </tr>
    <tr>
      <th>MaxMismatch</th>
      <td><p><strong>DEPRECATED AND IGNORED</strong>. Now calculated
automatically using <code><a href='CalcMaxMismatch.html'>CalcMaxMismatch</a></code>.</p></td>
    </tr>
    <tr>
      <th>Tassign</th>
      <td><p>minimum LLR required for acceptance of proposed relationship,
relative to next most likely relationship. Higher values result in more
conservative assignments. Must be zero or positive.</p></td>
    </tr>
    <tr>
      <th>Tfilter</th>
      <td><p>threshold log10-likelihood ratio (LLR) between a proposed
relationship versus unrelated, to select candidate relatives. Typically a
negative value, related to the fact that unconditional likelihoods are
calculated during the filtering steps. More negative values may decrease
non-assignment, but will increase computational time.</p></td>
    </tr>
    <tr>
      <th>MaxPairs</th>
      <td><p>the maximum number of putative pairs to return.</p></td>
    </tr>
    <tr>
      <th>quiet</th>
      <td><p>logical, suppress messages.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A list with</p>
<dt>MaybePar or MaybeRel</dt><dd><p>A dataframe with non-assigned likely
    relatives, with columns ID1 - ID2 - TopRel - LLR - OH - BirthYear1 -
    BirthYear2 - AgeDif - Sex1 - Sex2 - SNPdBoth</p></dd>
  <dt>MaybeTrio</dt><dd><p>A dataframe with non-assigned parent-parent-offspring
    trios, with columns id - parent1 - parent2 - LLRparent1 - LLRparent2 -
    LLRpair - OHparent1 - OHparent2 - MEpair - SNPd.id.parent1 -
    SNPd.id.parent2</p></dd>
The following categories are used in column 'TopRel', indicating the most
likely relationship category:
<dt>PO</dt><dd><p>Parent-Offspring</p></dd>
<dt>FS</dt><dd><p>Full Siblings</p></dd>
<dt>HS</dt><dd><p>Half Siblings</p></dd>
<dt>GP</dt><dd><p>GrandParent - grand-offspring</p></dd>
<dt>FA</dt><dd><p>Full Avuncular (aunt/uncle)</p></dd>
<dt>2nd</dt><dd><p>2nd degree relatives, not enough information to distinguish
  between HS,GP and FA</p></dd>
<dt>Q</dt><dd><p>Unclear, but probably 1st, 2nd or 3rd degree relatives</p></dd>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>When <code>Module="par"</code>, the age difference of the putative pair is
  temporarily set to NA so that genetic parent-offspring pairs declared to be
  born in the same year may be discovered. When <code>Module="ped"</code>, only
  relationships possible given the age difference, if known from the
  LifeHistData, are considered.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p><code><a href='sequoia.html'>sequoia</a></code> to identify likely pairs of duplicate
  genotypes and for pedigree reconstruction; <code><a href='GetRelM.html'>GetRelM</a></code> to
  identify all pairs of relatives in a pedigree; <code><a href='CalcPairLL.html'>CalcPairLL</a></code> for
  the likelihoods underlying the LLR.</p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># \donttest{</span>
<span class='va'>SeqOUT</span> <span class='op'>&lt;-</span> <span class='fu'><a href='sequoia.html'>sequoia</a></span><span class='op'>(</span>GenoM <span class='op'>=</span> <span class='va'>SimGeno_example</span>,
                  LifeHistData <span class='op'>=</span> <span class='va'>LH_HSg5</span>,
                  Module <span class='op'>=</span> <span class='st'>"par"</span>,
                  quiet<span class='op'>=</span><span class='cn'>TRUE</span>, Plot<span class='op'>=</span><span class='cn'>FALSE</span><span class='op'>)</span>
<span class='va'>MaybePO</span> <span class='op'>&lt;-</span> <span class='fu'>GetMaybeRel</span><span class='op'>(</span>GenoM <span class='op'>=</span> <span class='va'>SimGeno_example</span>,
                      SeqList <span class='op'>=</span> <span class='va'>SeqOUT</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Searching for non-assigned parent-offspring pairs ... (Module = par)</span></div><div class='output co'>#&gt; <span class='message'>using PedigreePar in SeqList</span></div><div class='output co'>#&gt; <span class='message'>using LifeHist in SeqList</span></div><div class='output co'>#&gt; <span class='message'>using AgePriors in SeqList</span></div><div class='output co'>#&gt; <span class='message'>Genotype matrix looks OK! There are  214  individuals and  200  SNPs.</span></div><div class='output co'>#&gt; <span class='message'>settings in SeqList$Specs will overrule input parameters</span></div><div class='output co'>#&gt; Counting opposing homozygous loci between all individuals ...
#&gt; Checking for non-assigned Parent-Offspring pairs ...
#&gt; Checking for Parent-Parent-Offspring trios ...</div><div class='output co'>#&gt; <span class='message'>Found 6 likely parent-offspring pairs, and 0 other non-assigned pairs of possible relatives</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>MaybePO</span><span class='op'>$</span><span class='va'>MaybePar</span><span class='op'>)</span>
</div><div class='output co'>#&gt;      ID1    ID2 TopRel  LLR OH BirthYear1 BirthYear2 AgeDif Sex1 Sex2 SNPdBoth
#&gt; 1 a01063 b00013     PO 3.63  1       2001       2000      1    1    2      199
#&gt; 2 a01166 b00002     PO 3.17  1       2001       2000      1    1    2      199
#&gt; 3 a01139 b00009     PO 2.52  1       2001       2000      1    1    2      194
#&gt; 4 a01167 b00002     PO 2.16  1       2001       2000      1    1    2      199
#&gt; 5 b01102 b00014     PO 1.75  1       2001       2000      1    2    2      197
#&gt; 6 a01114 b00010     PO 0.35  2       2001       2000      1    1    2      200</div><div class='input'>
<span class='co'># instead of providing the entire SeqList, one may</span>
<span class='co'># specify the relevant elements separately</span>
<span class='va'>Maybe</span> <span class='op'>&lt;-</span> <span class='fu'>GetMaybeRel</span><span class='op'>(</span>GenoM <span class='op'>=</span> <span class='va'>SimGeno_example</span>,
                     Pedigree <span class='op'>=</span> <span class='va'>SeqOUT</span><span class='op'>$</span><span class='va'>PedigreePar</span>,
                     LifeHistData <span class='op'>=</span> <span class='va'>LH_HSg5</span>,
                     Err<span class='op'>=</span><span class='fl'>0.0001</span>, Complex <span class='op'>=</span> <span class='st'>"full"</span>,
                     Module <span class='op'>=</span> <span class='st'>"ped"</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>Searching for non-assigned relative pairs ... (Module = ped)</span></div><div class='output co'>#&gt; <span class='message'>Genotype matrix looks OK! There are  214  individuals and  200  SNPs.</span></div><div class='output co'>#&gt; <span class='message'>Ageprior: Flat 0/1, discrete generations, MaxAgeParent = 1,1</span></div><div class='output co'>#&gt; Counting opposing homozygous loci between all individuals ...
#&gt; Checking for non-assigned relatives ...
#&gt; Checking for Parent-Parent-Offspring trios ...</div><div class='output co'>#&gt; <span class='message'>Found 16 likely parent-offspring pairs, and 256 other non-assigned pairs of possible relatives</span></div><div class='input'><span class='fu'><a href='https://rdrr.io/r/utils/head.html'>head</a></span><span class='op'>(</span><span class='va'>Maybe</span><span class='op'>$</span><span class='va'>MaybeRel</span><span class='op'>)</span>
</div><div class='output co'>#&gt;      ID1    ID2 TopRel  LLR OH BirthYear1 BirthYear2 AgeDif Sex1 Sex2 SNPdBoth
#&gt; 1 a01008 a00013     PO 7.31  2       2001       2000      1    1    1      197
#&gt; 2 b01075 a00013     PO 6.76  1       2001       2000      1    2    1      197
#&gt; 3 a01189 a00012     PO 4.30  2       2001       2000      1    1    1      199
#&gt; 4 a01063 b00013     PO 3.63  1       2001       2000      1    1    2      199
#&gt; 5 a01043 a00018     PO 3.44  1       2001       2000      1    1    1      197
#&gt; 6 a01166 b00002     PO 3.17  1       2001       2000      1    1    2      199</div><div class='input'>
<span class='co'># visualise results, turn dataframe into matrix first:</span>
<span class='va'>MaybeM</span> <span class='op'>&lt;-</span> <span class='fu'><a href='GetRelM.html'>GetRelM</a></span><span class='op'>(</span>Pairs<span class='op'>=</span><span class='va'>Maybe</span><span class='op'>$</span><span class='va'>MaybeRel</span><span class='op'>)</span>
<span class='fu'><a href='PlotRelPairs.html'>PlotRelPairs</a></span><span class='op'>(</span><span class='va'>MaybeM</span><span class='op'>)</span>
</div><div class='img'><img src='GetMaybeRel-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='co'># }</span>

</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


