# Fortran Wrapper for Pedigree Reconstruction

Call main Fortran part of sequoia, and convert its output to a list with
dataframes.

## Usage

``` r
SeqParSib(
  ParSib,
  FortPARAM,
  GenoM,
  LhIN,
  AgePriors,
  Parents,
  mtDif,
  DumPfx,
  quiet
)
```

## Arguments

- ParSib:

  either "par" to call parentage assignment, or "sib" to call the rest
  of the algorithm.

- FortPARAM:

  a named vector with parameter values, as generated by `ParamToSpecs`.

- GenoM:

  matrix with genotype data, size nInd x nSnp.

- LhIN:

  life history data: ID - sex - birth year.

- AgePriors:

  matrix with agepriors, size \`FortPARAM\["nAgeClasses"\]\` by 8.

- Parents:

  matrix with rownumbers of assigned parents, size nInd by 2.

- mtDif:

  matrix indicating whether individuals have definitely a different
  mitochondrial haplotype (1), or (possibly) the same (0). Size nInd x
  nInd.

- DumPfx:

  dummy prefixes

- quiet:

  suppress messages.

## Value

A list with

- PedigreePar or Pedigree:

  the pedigree

- DummyIDs:

  Info on dummies (not included if parentage-only)

- TotLikParents or TotLikSib:

  Total log-likelihood per iteration

- AgePriorExtra:

  Ageprior including columns for grandparental and avuncular
  relationships

- LifeHistPar or LifeHistSib:

  Includes sex and birthyear estimate inferred from the pedigree for
  individuals with initially unknown sex and/or birthyear

.

For a detailed description of the output see
[`sequoia`](https://jiscah.github.io/reference/sequoia.md).
