<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>SNP Summary Statistics — SnpStats • sequoia</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SNP Summary Statistics — SnpStats"><meta name="description" content="Estimate allele frequency (AF), missingness and Mendelian
errors per SNP."><meta property="og:description" content="Estimate allele frequency (AF), missingness and Mendelian
errors per SNP."><meta property="og:image" content="https://jiscah.github.io/reference/figures/sequoia_hexlogo.png"><meta property="og:image:alt" content="sequoia hexlogo"><meta name="twitter:card" content="summary_large_image"><meta name="twitter:creator" content="@HuismanJisca"><meta name="twitter:site" content="@HuismanJisca"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-dark" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">sequoia</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">3.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href=".././index.html" aria-label="home"><span class="fa fas fa-home"></span></a></li>
<li class="nav-item"><a class="nav-link" href="../articles/sequoia.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/vignette_main/book/background.html">Main vignette: User guide</a></li>
    <li><a class="dropdown-item" href="../articles/vignette_age/book/index.html">Age prior vignette</a></li>
    <li><a class="dropdown-item" href="../articles/sequoia_report.html">Mock report</a></li>
    <li><a class="dropdown-item" href="../articles/index.html">Examples &amp; More info on ...</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/performance_stats.html">Performance Stats</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Poster_sequoia_ESEB2017.pdf">poster</a></li>
    <li><a class="dropdown-item" href="../reference/figures/Slideshow_sequoia_background.pdf">slideshow with background info</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Change log</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-help" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Help</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-help"><li><a class="dropdown-item" href="../articles/help.html">Need help?</a></li>
    <li><a class="dropdown-item" href="../articles/want_to_help.html">Want to help?</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../mailto:jisca.huisman@gmail.com" aria-label="email"><span class="fa fas fa-envelope"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/JiscaH/sequoia" aria-label="github"><span class="fa fab fa-github"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://ecoevo.social/@HuismanJisca" aria-label="mastodon"><span class="fa fab fa-mastodon"></span></a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://www.linkedin.com/in/jisca-huisman-a4751153/" aria-label="linkedin"><span class="fa fab fa-linkedin-in"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>SNP Summary Statistics</h1>

      <div class="d-none name"><code>SnpStats.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Estimate allele frequency (AF), missingness and Mendelian
errors per SNP.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">SnpStats</span><span class="op">(</span></span>
<span>  <span class="va">GenoM</span>,</span>
<span>  Pedigree <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Duplicates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  Plot <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  quiet <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  calc_HWE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  <span class="va">ErrFlavour</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-genom">GenoM<a class="anchor" aria-label="anchor" href="#arg-genom"></a></dt>
<dd><p>genotype matrix, in sequoia's format: 1 column per SNP, 1 row
per individual, genotypes coded as 0/1/2/-9, and row names giving individual
IDs.</p></dd>


<dt id="arg-pedigree">Pedigree<a class="anchor" aria-label="anchor" href="#arg-pedigree"></a></dt>
<dd><p>dataframe with 3 columns: ID - parent1 - parent2.
Additional columns and non-genotyped individuals are ignored. Used to count
Mendelian errors per SNP and (poorly) estimate the error rate.</p></dd>


<dt id="arg-duplicates">Duplicates<a class="anchor" aria-label="anchor" href="#arg-duplicates"></a></dt>
<dd><p>dataframe with pairs of duplicated samples</p></dd>


<dt id="arg-plot">Plot<a class="anchor" aria-label="anchor" href="#arg-plot"></a></dt>
<dd><p>logical, show histograms of the results?</p></dd>


<dt id="arg-quiet">quiet<a class="anchor" aria-label="anchor" href="#arg-quiet"></a></dt>
<dd><p>logical, suppress messages?</p></dd>


<dt id="arg-calc-hwe">calc_HWE<a class="anchor" aria-label="anchor" href="#arg-calc-hwe"></a></dt>
<dd><p>logical, calculate chi-square test for Hardy-Weinberg
equilibrium? Can be relatively time consuming for large datasets.</p></dd>


<dt id="arg-errflavour">ErrFlavour<a class="anchor" aria-label="anchor" href="#arg-errflavour"></a></dt>
<dd><p>DEPRECATED AND IGNORED. Was used to estimate <code>Err.hat</code></p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A matrix with a number of rows equal to the number of SNPs
 (=number of columns of GenoM), and when no Pedigree is provided 2 columns:</p>
<dl><dt>AF</dt>
<dd><p>Allele frequency of the 'second allele' (the one for which the
  homozygote is coded 2)</p></dd>

<dt>Mis</dt>
<dd><p>Proportion of missing calls</p></dd>

<dt>HWE.p</dt>
<dd><p>p-value from chi-square test for Hardy-Weinberg equilibrium</p></dd>


</dl><p>When a Pedigree is provided, there are 8 additional columns:</p>
<dl><dt>n.dam, n.sire, n.pair</dt>
<dd><p>Number of dams, sires, parent-pairs successfully
  genotyped for the SNP</p></dd>

<dt>OHdam, OHsire</dt>
<dd><p>Count of number of opposing homozygous cases</p></dd>

<dt>MEpair</dt>
<dd><p>Count of Mendelian errors, includes opposing homozygous cases
  when only one parent is genotyped</p></dd>

<dt>n.dups, n.diff</dt>
<dd><p>Number of duplicate pairs successfully genotyped for
  the SNP; number of differences. The latter does not count cases where one
  duplicate is not successfully genotyped at the SNP</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Calculation of these summary statistics can be done in PLINK, and
  SNPs with low minor allele frequency or high missingness should be filtered
  out prior to pedigree reconstruction. This function is provided as an aid
  to inspect the relationship between AF, missingness and genotyping error to
  find a suitable combination of SNP filtering thresholds to use.</p>
<p>For pedigree reconstruction, SNPs with zero or one copies of the alternate
  allele in the dataset (MAF \(\le 1/2N\)) are considered fixed, and
  excluded.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="GenoConvert.html">GenoConvert</a></code> to convert from various data formats;
  <code><a href="CheckGeno.html">CheckGeno</a></code> to check the data is in valid format for sequoia
  and exclude monomorphic SNPs etc., <code><a href="CalcOHLLR.html">CalcOHLLR</a></code> to calculate OH
  &amp; ME per individual.</p></div>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">Genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="SimGeno.html">SimGeno</a></span><span class="op">(</span><span class="va">Ped_HSg5</span>, nSnp<span class="op">=</span><span class="fl">100</span>, CallRate <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">100</span>, <span class="fl">0.5</span>, <span class="fl">0.8</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                     SnpError <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">SnpStats</span><span class="op">(</span><span class="va">Genotypes</span><span class="op">)</span>   <span class="co"># only plots; data is returned invisibly</span></span></span>
<span class="r-plt img"><img src="SnpStats-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="va">SNPstats</span> <span class="op">&lt;-</span> <span class="fu">SnpStats</span><span class="op">(</span><span class="va">Genotypes</span>, Pedigree<span class="op">=</span><span class="va">Ped_HSg5</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

