<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Quick start example 1: Simulated data • sequoia</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/sandstone/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Quick start example 1: Simulated data">
<meta property="og:description" content="Install sequoia, simulate data, run parentage assignment and full pedigree reconstruction.
">
<meta property="og:image" content="/reference/figures/sequoia_hexlogo_s.png">
<meta property="og:image:alt" content="sequoia hexlogo">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">sequoia</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home"></span>
     
  </a>
</li>
<li>
  <a href="../articles/Get_started.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Manuals
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/vignette_main/_book/sec-Background.html">Main vignette: User guide</a>
    </li>
    <li>
      <a href="../articles/vignette_age/_book/index.html">Age prior vignette</a>
    </li>
    <li>
      <a href="../articles/index.html">Examples</a>
    </li>
    <li>
      <a href="../articles/index.html">More info on ...</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Change log</a>
</li>
<li>
  <a href="../articles/help.html">Help</a>
</li>
<li>
  <a href="../articles/_performance_stats.html">Stats</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/HuismanJisca">
    <span class="fab fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="../mailto:jisca.huisman@gmail.com">
    <span class="fas fa-envelope"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/JiscaH/sequoia">
    <span class="fab fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://www.linkedin.com/in/jisca-huisman-a4751153/">
    <span class="fab fa-linkedin-in"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="quick_start_example1_files/header-attrs-2.8/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Quick start example 1: Simulated data</h1>
            
      
      
      <div class="hidden name"><code>quick_start_example1.Rmd</code></div>

    </div>

    
    
<hr>
<div id="get-started" class="section level2">
<h2 class="hasAnchor">
<a href="#get-started" class="anchor"></a>Get started</h2>
<p><em>For a succinct version of this example where you can copy/paste all code in one go, see the main vignette</em></p>
<p>In this example genotypes are simulated from a fictional pedigree, which together with the associated life history data is provided with the package. This pedigree consists of 5 discrete generations with interconnected half-sib clusters (Pedigree II in the paper<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>).</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Install the package. This is only the required the first time, or if you wish to update</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"sequoia"</span><span class="op">)</span>   

<span class="co"># Load the package. This is required at the start of every new R session.</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://jiscah.github.io/">sequoia</a></span><span class="op">)</span> 

<span class="co"># Load the example pedigree and life history data</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Ped_HSg5</span>, <span class="va">LH_HSg5</span><span class="op">)</span>

<span class="co"># Take a look at the data structure</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">Ped_HSg5</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">LH_HSg5</span><span class="op">)</span>

<span class="co"># or, in Rstudio, view the full dataframe:</span>
<span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">Ped_HSg5</span><span class="op">)</span></code></pre></div>
</div>
<div id="simulate-genotypes" class="section level2">
<h2 class="hasAnchor">
<a href="#simulate-genotypes" class="anchor"></a>Simulate genotypes</h2>
<p>Simulate some genotype data to use for this try-out.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">Geno</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SimGeno.html">SimGeno</a></span><span class="op">(</span>Ped <span class="op">=</span> <span class="va">Ped_HSg5</span>, nSnp <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></code></pre></div>
<p>Function <code><a href="../reference/SimGeno.html">SimGeno()</a></code> can simulate data with a specific genotyping error rate, call rate and proportion of non-genotyped parents, see its helpfile (<code><a href="../reference/SimGeno.html">?SimGeno</a></code>) for details. Here we use the defaults; 40% of parents are presumed non-genotyped (their genotypes are removed from the dataset before <code><a href="../reference/SimGeno.html">SimGeno()</a></code> returns its result).</p>
</div>
<div id="parentage-assignment" class="section level2">
<h2 class="hasAnchor">
<a href="#parentage-assignment" class="anchor"></a>Parentage assignment</h2>
<p>Run sequoia with the genotype data, lifehistory data, and otherwise default values. It is often advisable to first only run parentage assignment, and check if the results are sensible and/or if any parameters need adjusting. Full pedigree reconstruction, including sibship clustering etc., is much more time consuming.</p>
<p>To only run a data check, duplicate check (always performed first) and parentage assignment, specify <code>Module = 'par'</code> (in older versions: <code>MaxSibIter=0</code>):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ParOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,
                  LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>,
                  Module <span class="op">=</span> <span class="st">'par'</span><span class="op">)</span></code></pre></div>
<pre><code>## Genotype matrix looks OK! There are  920  individuals and  200  SNPs.</code></pre>
<pre><code>## Ageprior: Flat 0/1, overlapping generations, MaxAgeParent = 6,6</code></pre>
<p><img src="quick_start_example1_files/figure-html/example1-e-1.png" width="576"></p>
<pre><code>## 
## ~~~ Parentage assignment ~~~</code></pre>
<pre><code>## Parentage ...
## Initial total LL :
## [1] -77382.2
## Post-parentage total LL :
## [1] -60410.37
## Estimating birth years ...
## Calculating parental LLR ...</code></pre>
<pre><code>## assigned 534 dams and 545 sires to 920 individuals</code></pre>
<p><img src="quick_start_example1_files/figure-html/example1-e-2.png" width="576"><img src="quick_start_example1_files/figure-html/example1-e-3.png" width="576"></p>
<!--- explain about ageprior messages & plots, and LL --->
<p>You will see several plots appearing:</p>
<ul>
<li>Histograms of the minor allele frequency and missingness, produced by <code><a href="../reference/SnpStats.html">SnpStats()</a></code>, which is called by <code><a href="../reference/CheckGeno.html">CheckGeno()</a></code> to make sure there are no monomorphic SNPs or SNPs with extremely high missingness in the dataset<br>
</li>
<li>Heatmap of the ageprior before parentage assignment, showing that an age difference of 0 years is disallowed (<span class="math inline">\(0\)</span>, black) for mothers (<em>M</em>) and fathers (<em>P</em>), but all other age differences are allowed (<span class="math inline">\(1\)</span>, pale green) for all relationships. The initial maximum age of parents is set to the maximum age difference found in the lifehistory data.<br>
</li>
<li>Barplots of assignment rate (produced by <code><a href="../reference/SummarySeq.html">SummarySeq()</a></code>): nearly 60% of individuals have a genotyped parent assigned, in line with the 40% non-genotyped parents we simulated.</li>
</ul>
<p>In addition, you will see several messages, including about the initial and post-parentage total LL (log10-likelihood). This is the probability of observing the genotype data, given the (current) pedigree; initially it is assumed that all individuals are unrelated. This number is negative, and gets closer to zero when the pedigree explains the data better.</p>
<p>The result is a list, with the following elements:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ParOUT</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "Specs"         "ErrM"          "AgePriors"     "LifeHist"     
##  [5] "DupLifeHistID" "NoLH"          "PedigreePar"   "TotLikPar"    
##  [9] "AgePriorExtra" "LifeHistPar"</code></pre>
<p>which are explained in detail in the helpfile (<code><a href="../reference/sequoia.html">?sequoia</a></code>) and the main vignette.</p>
<p>you find the assigned parents in list element <code>PedigreePar</code>:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></code></pre></div>
<pre><code>##         id    dam   sire LLRdam LLRsire LLRpair OHdam OHsire MEpair
## 915 b05187   &lt;NA&gt;   &lt;NA&gt;     NA      NA      NA    NA     NA     NA
## 916 a05188   &lt;NA&gt;   &lt;NA&gt;     NA      NA      NA    NA     NA     NA
## 917 a05189 a04006 b04177   7.80   11.77    9.24     0      0      0
## 918 b05190 a04006 b04177   7.17   10.22    8.45     0      0      0
## 919 b05191 a04006 b04177  11.01    8.02   10.40     0      0      0
## 920 b05192 a04006 b04177  10.08    9.66    9.71     0      0      0</code></pre>
<p>we can compare these to the true parents, in the original pedigree from which we simulated the genotype data:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">chk_par</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">Ped_HSg5</span>, Ped2 <span class="op">=</span> <span class="va">ParOUT</span><span class="op">$</span><span class="va">PedigreePar</span><span class="op">)</span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-h-1.png" width="576"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">chk_par</span><span class="op">$</span><span class="va">Counts</span><span class="op">[</span><span class="st">"GG"</span>,,<span class="op">]</span>   </code></pre></div>
<pre><code>##           parent
## class      dam sire
##   Total    536  545
##   Match    534  545
##   Mismatch   0    0
##   P1only     2    0
##   P2only     0    0</code></pre>
<p>Here ‘GG’ stands for <em>G</em>enotyped offspring, <em>G</em>enotyped parent.</p>
</div>
<div id="pedigree-reconstruction" class="section level2">
<h2 class="hasAnchor">
<a href="#pedigree-reconstruction" class="anchor"></a>Pedigree reconstruction</h2>
<p>Then, we can run full pedigree reconstruction. By default, this re-runs the parentage assignment.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>, LifeHistData <span class="op">=</span> <span class="va">LH_HSg5</span>, Module <span class="op">=</span> <span class="st">"ped"</span><span class="op">)</span></code></pre></div>
<p>If you don’t want to re-run parentage assignment (e.g. because it took quite a bit of time), or if you have specified several non-default parameter values you want to use again, , you can provide the old output as input.</p>
<p>Re-used will be:</p>
<ul>
<li>parameter settings (<code>$Specs</code>)</li>
<li>genotyping error structure (<code>$ErrM</code>)</li>
<li>Lifehistory data (<code>$LifeHist</code>)</li>
<li>Age prior, based on the age distribution of assigned parents (<code>$AgePriors</code>)</li>
</ul>
<p>The last is generated by <code><a href="../reference/MakeAgePrior.html">MakeAgePrior()</a></code>, which has detected that all parent-offspring pairs have an age difference of <span class="math inline">\(1\)</span>, and all siblings an age difference of <span class="math inline">\(0\)</span>, i.e. that generations do not overlap. This information will be used during the full pedigree reconstruction.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ParOUT</span><span class="op">$</span><span class="va">AgePriors</span></code></pre></div>
<pre><code>##   M P FS MS PS
## 0 0 0  1  1  1
## 1 1 1  0  0  0
## 2 0 0  0  0  0</code></pre>
<p>So run <code><a href="../reference/sequoia.html">sequoia()</a></code> with the old output as input,</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">SeqOUT</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sequoia.html">sequoia</a></span><span class="op">(</span>GenoM <span class="op">=</span> <span class="va">Geno</span>,
                  SeqList <span class="op">=</span> <span class="va">ParOUT</span>,
                  Module <span class="op">=</span> <span class="st">"ped"</span><span class="op">)</span></code></pre></div>
<pre><code>## Genotype matrix looks OK! There are  920  individuals and  200  SNPs.</code></pre>
<pre><code>## using LifeHistData in SeqList</code></pre>
<pre><code>## using AgePriors in SeqList</code></pre>
<pre><code>## using PedigreePar in SeqList</code></pre>
<pre><code>## 
## ~~~ Full pedigree reconstruction ~~~</code></pre>
<pre><code>## Sibships - Initial Total LL :
## [1] -60410.37
## Round 01 end, Total LogLik; time (sec):
## [1] -47756.052     17.875
## No. dams, sires for real indiv.:
## [1] 891 898
## Round 02 end, Total LogLik; time (sec):
## [1] -47037.238153      9.671875
## No. dams, sires for real indiv.:
## [1] 898 898
## Round 03 end, Total LogLik; time (sec):
## [1] -47028.68014      1.28125
## No. dams, sires for real indiv.:
## [1] 898 898
## Round 04 end, Total LogLik; time (sec):
## [1] -47028.56630      0.96875
## No. dams, sires for real indiv.:
## [1] 898 898
## Estimating birth years ...
## Calculating parental LLR ...</code></pre>
<pre><code>## assigned 960 dams and 960 sires to 920 + 80 individuals (real + dummy)</code></pre>
<p><img src="quick_start_example1_files/figure-html/example1-j-1.png" width="576"><img src="quick_start_example1_files/figure-html/example1-j-2.png" width="576"><img src="quick_start_example1_files/figure-html/example1-j-3.png" width="576"></p>
<p>Full pedigree reconstruction will take at least a few minutes to run on this fairly simple dataset with one thousand individuals. It may take a few hours on larger or more complex datasets, and/or if there is much ambiguity about relationships due to a low number of SNPs, genotyping errors, and missing data.</p>
<p>You will get several messages:</p>
<ul>
<li>From <code><a href="../reference/CheckGeno.html">CheckGeno()</a></code> to inform you that the genotype data is OK<br>
</li>
<li>To inform you which parts of <code>SeqList</code> (the output from the old <code><a href="../reference/sequoia.html">sequoia()</a></code> run) are being re-used<br>
</li>
<li>At the end of each iteration, to update you on the total log10-likelihood (LL), the time it took for that iteration, and the number of dams &amp; sires assigned to real, genotyped (non-dummy) individuals. The total LL and number of assigned parents should plateau (but may ‘wobble’ a bit), and then the algorithm will finish.</li>
</ul>
<p>Again you will see some plots:</p>
<ul>
<li>Histograms of the minor allele frequency and missingness<br>
</li>
<li>Heatmap of an extended version of the ageprior, including all grandparents and avuncular (aunt/uncle) relationships (<code>$AgePriorExtra</code>). These are derived from the ageprior for parents and siblings, which hasn’t changed since just after parentage assignment<br>
</li>
<li>an update of the assignment rates plot, now including dummy parents.</li>
</ul>
<p>For additional plots and a few tables to inspect the pedigree, you can use <code><a href="../reference/SummarySeq.html">SummarySeq()</a></code>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">summary_seq1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SummarySeq.html">SummarySeq</a></span><span class="op">(</span><span class="va">SeqOUT</span>, Panels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"sibships"</span>, <span class="st">"D.parents"</span>, <span class="st">"LLR"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-k-1.png" width="576"><img src="quick_start_example1_files/figure-html/example1-k-2.png" width="576"><img src="quick_start_example1_files/figure-html/example1-k-3.png" width="576"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">summary_seq1</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "PedSummary"  "ParentCount" "GPCount"     "SibSize"</code></pre>
<p>And again we can also compare the results to the true parents:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">chk</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PedCompare.html">PedCompare</a></span><span class="op">(</span>Ped1 <span class="op">=</span> <span class="va">Ped_HSg5</span>, Ped2 <span class="op">=</span> <span class="va">SeqOUT</span><span class="op">$</span><span class="va">Pedigree</span><span class="op">)</span></code></pre></div>
<p><img src="quick_start_example1_files/figure-html/example1-l-1.png" width="576"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">chk</span><span class="op">$</span><span class="va">Counts</span></code></pre></div>
<pre><code>## , , parent = dam
## 
##     class
## cat  Total Match Mismatch P1only P2only
##   GG   536   536        0      0      0
##   GD   362   362        0      0      0
##   GT   898   898        0      0      0
##   DG    40    40        0      0      0
##   DD    22    22        0      0      0
##   DT    62    62        0      0      0
##   TT   960   960        0      0      0
## 
## , , parent = sire
## 
##     class
## cat  Total Match Mismatch P1only P2only
##   GG   545   545        0      0      0
##   GD   353   353        0      0      0
##   GT   898   898        0      0      0
##   DG    31    31        0      0      0
##   DD    31    31        0      0      0
##   DT    62    62        0      0      0
##   TT   960   960        0      0      0</code></pre>
<p>See <code><a href="../reference/PedCompare.html">?PedCompare</a></code> for a more interesting example with some mismatches.</p>
<p>If you wish to count e.g. the number of full sibling pairs that are assigned as full siblings, paternal half siblings, maternal halfsiblings, or unrelated, or similar comparisons, use <code><a href="../reference/ComparePairs.html">ComparePairs()</a></code>.</p>
</div>
<div id="save-results" class="section level2">
<h2 class="hasAnchor">
<a href="#save-results" class="anchor"></a>Save results</h2>
<p>Lastly, you often wish to save the results to file. You can do this as an <code>Rdata</code> file, which can contain any number of R objects. You can later retrieve these in R using <code>load</code>, so that you can later resume where you left of. The disadvantage is that you cannot open the <code>Rdata</code> files outside of R (as far as I am aware). Therefore, <code>sequoia</code> also includes a function to write all output to a set of plain text files in a specified folder.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/save.html">save</a></span><span class="op">(</span><span class="va">SeqOUT</span>, <span class="va">Geno</span>, file<span class="op">=</span><span class="st">"Sequoia_output_date.RData"</span><span class="op">)</span>
<span class="fu"><a href="../reference/writeSeq.html">writeSeq</a></span><span class="op">(</span>SeqList <span class="op">=</span> <span class="va">SeqOUT</span>, GenoM <span class="op">=</span> <span class="va">Geno</span>, folder <span class="op">=</span> <span class="st">"Sequoia-OUT"</span><span class="op">)</span></code></pre></div>
<hr>
<div id="references" class="section level3">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h3>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Huisman, Jisca. “Pedigree reconstruction from SNP data: parentage assignment, sibship clustering and beyond.” Molecular ecology resources 17.5 (2017): 1009-1024.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jisca Huisman.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
